// @dvag/design-system/dist/collection/components/buttons/ds-button/ds-button.js
import { h, Host, } from "@stencil/core";
import { assertPropertyValue } from "../../../classes/assert";
import { ElementIdGenerator, verifyElementId, } from "../../../classes/element-id-handling";
/**
 * Displays a primary or secondary button.
 *
 * @path /Buttons
 * @name Button
 * @shortname Button
 * @icon minus-square
 * @stable
 *
 * @slot - Contains the label of the button. Use the `<ds-text>` component for that.
 * @slot-example verschiedene-stile
 */
export class DsButton {
  constructor() {
    /**
     * Modifies the button width behavior
     * @value false: Button width is determined by the button content
     * @value true: Button width is stretched to match the width of the host element.
     */
    this.stretch = false;
    /**
     * Disables user interaction on the button.
     * @value false: User can interaction with the button.
     * @value true: User can not interaction with the button.
     * @example disabled
     * @example verschiedene-stile
     */
    this.disabled = false;
    /**
     * Disables the ability to focus the button via tabulator navigation.
     * @value false: Button is reachable via tab navigation.
     * @value true: Button is ignored when using tab navigation.
     * @example unfocusable
     */
    this.unfocusable = false;
  }
  /**
   * Focuses this element.
   */
  async focusControl() {
    this.buttonElement.focus();
  }
  componentWillLoad() {
    verifyElementId(this.self);
  }
  render() {
    assertPropertyValue("ds-button", "type", this.type, [
      "primary",
      "secondary",
    ]);
    assertPropertyValue("ds-button", "size", this.size, ["medium", "small"]);
    return (h(Host, { class: {
        "type-primary": this.type === "primary",
        "type-secondary": this.type === "secondary",
        "size-medium": this.size === "medium",
        "size-small": this.size === "small",
        stretch: this.stretch,
        disabled: this.disabled,
        unfocusable: this.unfocusable,
      } }, h("button", { type: "button", ref: (el) => (this.buttonElement = el), id: ElementIdGenerator.createId(this.self, "control"), disabled: this.disabled, tabindex: this.unfocusable ? "-1" : "0" }, h("slot", null))));
  }
  static get is() { return "ds-button"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-button.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-button.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"primary\" | \"secondary\"",
          "resolved": "\"primary\" | \"secondary\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "primary: Flat colored button to indicate an important action."
            }, {
              "name": "value",
              "text": "secondary: Outlined transparent button for less important actions."
            }, {
              "name": "example",
              "text": "type"
            }, {
              "name": "example",
              "text": "verschiedene-stile"
            }],
          "text": "Defines the visual representation of the button."
        },
        "attribute": "type",
        "reflect": false
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"medium\" | \"small\"",
          "resolved": "\"medium\" | \"small\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "medium: Medium button size"
            }, {
              "name": "value",
              "text": "small: Small button size, optimized for usage in forms"
            }, {
              "name": "example",
              "text": "size"
            }, {
              "name": "example",
              "text": "verschiedene-stile"
            }],
          "text": "Defines the size of othe button."
        },
        "attribute": "size",
        "reflect": false
      },
      "stretch": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Button width is determined by the button content"
            }, {
              "name": "value",
              "text": "true: Button width is stretched to match the width of the host element."
            }],
          "text": "Modifies the button width behavior"
        },
        "attribute": "stretch",
        "reflect": false,
        "defaultValue": "false"
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: User can interaction with the button."
            }, {
              "name": "value",
              "text": "true: User can not interaction with the button."
            }, {
              "name": "example",
              "text": "disabled"
            }, {
              "name": "example",
              "text": "verschiedene-stile"
            }],
          "text": "Disables user interaction on the button."
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "unfocusable": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Button is reachable via tab navigation."
            }, {
              "name": "value",
              "text": "true: Button is ignored when using tab navigation."
            }, {
              "name": "example",
              "text": "unfocusable"
            }],
          "text": "Disables the ability to focus the button via tabulator navigation."
        },
        "attribute": "unfocusable",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get methods() {
    return {
      "focusControl": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Focuses this element.",
          "tags": []
        }
      }
    };
  }
  static get elementRef() { return "self"; }
}
