// @dvag/design-system/dist/collection/components/layout/ds-placeholder/ds-placeholder.js
import { Host, h } from "@stencil/core";
/**
 * Dient während der Entwicklung von UI-Oberflächen als Platzhalter, um Layout-Mechanismen zu testen.
 *
 * @path /Layout
 * @name Placeholder
 * @stable
 * @icon image
 * @slot - Label, dass innerhalb des Platzhalters anzeigt wird.
 *
 */
export class DsPlaceholder {
  render() {
    return (h(Host, { class: {
        box: this.box,
      } }, this.buildContent()));
  }
  buildContent() {
    if (this.box) {
      return h("slot", null);
    }
    if (this.card) {
      return (h("ds-card", { shadow: true }, h("ds-card-header", null, h("ds-text", { type: "u5" }, h("slot", null))), h("ds-card-content", null, h("ds-text", null, "Placeholder lorem ipsum dolor sit amet, consectetur adipisicing elit. Architecto cupiditate distinctio doloribus."))));
    }
  }
  static get is() { return "ds-placeholder"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-placeholder.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-placeholder.css"]
    };
  }
  static get properties() {
    return {
      "box": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "boolean: Box anzeigen"
            }, {
              "name": "example",
              "text": "box"
            }],
          "text": "Zeigt eine Box an."
        },
        "attribute": "box",
        "reflect": false
      },
      "card": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "boolean: Card anzeigen"
            }, {
              "name": "example",
              "text": "card"
            }],
          "text": "Zeigt eine Card an."
        },
        "attribute": "card",
        "reflect": false
      }
    };
  }
}
