// @dvag/design-system/dist/collection/new/scaffold/dx-footer/dx-footer.js
import { h, Host } from "@stencil/core";
/**
 * A footer component which should be placed at the end of the site.
 *
 * @path /Scaffold
 * @name Footer
 * @new
 * @icon shoe-prints
 * @stable
 *
 * @slot - A content shown in the footer.
 */
export class DxFooter {
  constructor() {
    /**
     * The color of the previous container to ensure a smooth transition for the separator.
     *
     * @value string: The name of a color from the color palette.
     *
     * @example from-color-white
     * @example from-color-gray
     */
    this.fromcolor = "white";
  }
  render() {
    return (h(Host, null, h("dx-fd-separator", { fromcolor: this.fromcolor, tocolor: "footer" }), h("dx-container", { class: "content-container", type: "page", color: "footer" }, h("slot", null))));
  }
  static get is() { return "dx-footer"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-footer.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-footer.css"]
    };
  }
  static get properties() {
    return {
      "fromcolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The name of a color from the color palette."
            }, {
              "name": "example",
              "text": "from-color-white"
            }, {
              "name": "example",
              "text": "from-color-gray"
            }],
          "text": "The color of the previous container to ensure a smooth transition for the separator."
        },
        "attribute": "fromcolor",
        "reflect": false,
        "defaultValue": "\"white\""
      }
    };
  }
}
