// @dvag/design-system/dist/collection/components/form/inputs/ds-text-area/ds-text-area.js
import { h, Host, } from "@stencil/core";
import { assert } from "../../../../classes/assert";
import { ElementIdGenerator, verifyElementId, } from "../../../../classes/element-id-handling";
import { ReadonlyTextElement } from "../../../functional/base-elements";
import { ErrorPopup } from "../../../functional/error-popup";
import { InputContainer } from "../../../functional/input-container";
/**
 * Formular-Element zur Erfassung von Fließtext
 *
 * @path /Form/Textfeld
 * @name Textarea
 * @shortname Textarea
 * @icon align-justify
 * @stable
 */
export class DsTextArea {
  constructor() {
    this.inputFocused = false;
    this.internalValue = "";
    /**
     * Colors the border of the textarea red.
     * @value true: red border.
     * @value false: grey border.
     * @example error
     */
    this.error = false;
    /**
     * Number of lines shown simultaneously.
     * @value number: Number of lines.
     * @example lines
     */
    this.lines = 4;
    /**
     * Text of the textarea.
     * @important
     * @value string: Content of the textarea.
     * @example value
     */
    this.value = "";
    /**
     * Textarea is disabled and cannot be used anymore.
     * Label and border are grey.
     * @value true: Textarea is disabled
     * @value false: Textarea is not disabled
     * @example disabled
     */
    this.disabled = false;
    /**
     * Textarea is in read-only mode.
     * Content of the textare is displayed in a simple way.
     * @value false: Formfield
     * @value true: Shows the content only as text.
     * @example readonly
     */
    this.readonly = false;
    /**
     * Marks the textarea as mandatory.
     * @value false: Not mandatory (Default).
     * @value true: mandatory field.
     * @example required
     * @example required-maxlength
     */
    this.required = false;
  }
  valueChanged() {
    this.obtainValueChangesFromOutside();
  }
  /**
   * Sets the focus to the textarea.
   */
  async focusControl() {
    this.inputElement.focus();
  }
  componentWillLoad() {
    verifyElementId(this.self);
    this.obtainValueChangesFromOutside(true);
  }
  obtainValueChangesFromOutside(force = false) {
    if (force || this.value !== this.internalValue) {
      this.internalValue = this.value;
      this.lastEmittedChangeValue = this.value;
      this.lastEmittedCompleteValue = this.value;
    }
  }
  onInputFocused() {
    this.inputFocused = true;
  }
  onInputBlurred() {
    this.inputFocused = false;
    this.emitValueChange();
    this.emitValueComplete();
  }
  onInputKeyUp(event) {
    this.internalValue = event.target.value;
    this.emitValueChange();
  }
  emitValueChange() {
    this.value = this.internalValue;
    if (this.lastEmittedChangeValue !== this.value) {
      this.valueChange.emit(this.value);
      this.lastEmittedChangeValue = this.value;
    }
  }
  emitValueComplete() {
    if (this.lastEmittedCompleteValue !== this.value) {
      this.valueComplete.emit(this.value);
      this.lastEmittedCompleteValue = this.value;
    }
  }
  render() {
    if (this.readonly) {
      return h(ReadonlyTextElement, { text: this.value, label: this.label });
    }
    return h(Host, null, this.renderInput());
  }
  renderInput() {
    var _a, _b;
    assert(this.maxlength == null || this.maxlength > 0, () => `<ds-text-area> 'maxlength' sollte nicht kleiner als 1 sein. `);
    assert(this.maxlength == null || this.maxlength > this.value.length, () => `<ds-text-area> 'value' darf nicht größer als 'maxlength' sein`);
    assert(this.lines > 0, () => `<ds-text-area> 'lines' muss größer als 0 sein`);
    const isErrorOutline = ((_a = this.errormessage) === null || _a === void 0 ? void 0 : _a.length) > 0 || this.error;
    const requirementFulfilled = ((_b = this.value) === null || _b === void 0 ? void 0 : _b.length) > 0;
    return (h("label", null, h("ds-input-label", { disabled: this.disabled, label: this.label }), h(InputContainer, { active: this.inputFocused, error: isErrorOutline, required: this.required, requirementFulfilled: requirementFulfilled, disabled: this.disabled, title: this.value }, h("ds-scroll-container", { style: { height: `${MARGIN_PX + this.lines * LINE_PX}px` }, scrollbarspace: 8 }, h("textarea", { id: ElementIdGenerator.createId(this.self, "control"), onKeyUp: (event) => this.onInputKeyUp(event), onFocus: () => this.onInputFocused(), onBlur: () => this.onInputBlurred(), ref: (el) => (this.inputElement = el), placeholder: "Bitte geben Sie hier Ihren Text ein ...", value: this.internalValue, maxLength: this.maxlength > 0 ? this.maxlength : undefined, disabled: this.disabled })), h(ErrorPopup, { text: this.errormessage, show: this.inputFocused })), this.renderRemainingLetters(), h("ds-required-hint", { required: this.required, fulfilled: requirementFulfilled })));
  }
  renderRemainingLetters() {
    if (this.maxlength > 0) {
      const remainingLetters = this.maxlength - this.internalValue.length;
      return (h("div", { class: "remaining-letters" }, h("ds-text", { type: "itk" }, remainingLetters, " Zeichen verbleibend")));
    }
  }
  static get is() { return "ds-text-area"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-text-area.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-text-area.css"]
    };
  }
  static get properties() {
    return {
      "errormessage": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Errormessage"
            }, {
              "name": "example",
              "text": "error"
            }],
          "text": "Text that is shown under the textarea in a red dialog."
        },
        "attribute": "errormessage",
        "reflect": false
      },
      "error": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: red border."
            }, {
              "name": "value",
              "text": "false: grey border."
            }, {
              "name": "example",
              "text": "error"
            }],
          "text": "Colors the border of the textarea red."
        },
        "attribute": "error",
        "reflect": false,
        "defaultValue": "false"
      },
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Label e.g `Notizen`"
            }, {
              "name": "important",
              "text": undefined
            }, {
              "name": "example",
              "text": "label"
            }],
          "text": "Label shown above the textarea."
        },
        "attribute": "label",
        "reflect": false
      },
      "maxlength": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: Number of characters"
            }, {
              "name": "example",
              "text": "maxlength"
            }],
          "text": "Maximum number of characters."
        },
        "attribute": "maxlength",
        "reflect": false
      },
      "lines": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: Number of lines."
            }, {
              "name": "example",
              "text": "lines"
            }],
          "text": "Number of lines shown simultaneously."
        },
        "attribute": "lines",
        "reflect": false,
        "defaultValue": "4"
      },
      "value": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "important",
              "text": undefined
            }, {
              "name": "value",
              "text": "string: Content of the textarea."
            }, {
              "name": "example",
              "text": "value"
            }],
          "text": "Text of the textarea."
        },
        "attribute": "value",
        "reflect": true,
        "defaultValue": "\"\""
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: Textarea is disabled"
            }, {
              "name": "value",
              "text": "false: Textarea is not disabled"
            }, {
              "name": "example",
              "text": "disabled"
            }],
          "text": "Textarea is disabled and cannot be used anymore.\nLabel and border are grey."
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "readonly": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Formfield"
            }, {
              "name": "value",
              "text": "true: Shows the content only as text."
            }, {
              "name": "example",
              "text": "readonly"
            }],
          "text": "Textarea is in read-only mode.\nContent of the textare is displayed in a simple way."
        },
        "attribute": "readonly",
        "reflect": false,
        "defaultValue": "false"
      },
      "required": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Not mandatory (Default)."
            }, {
              "name": "value",
              "text": "true: mandatory field."
            }, {
              "name": "example",
              "text": "required"
            }, {
              "name": "example",
              "text": "required-maxlength"
            }],
          "text": "Marks the textarea as mandatory."
        },
        "attribute": "required",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get states() {
    return {
      "inputFocused": {},
      "internalValue": {}
    };
  }
  static get events() {
    return [{
        "method": "valueChange",
        "name": "valueChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "example",
              "text": "change-events"
            }],
          "text": "Event is thrown when the user changes the content of the textarea."
        },
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        }
      }, {
        "method": "valueComplete",
        "name": "valueComplete",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "example",
              "text": "change-events"
            }],
          "text": "Event is thrown when the user checks the input. Event is triggered when the user presses Return or\nleaves the textarea."
        },
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        }
      }];
  }
  static get methods() {
    return {
      "focusControl": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Sets the focus to the textarea.",
          "tags": []
        }
      }
    };
  }
  static get elementRef() { return "self"; }
  static get watchers() {
    return [{
        "propName": "value",
        "methodName": "valueChanged"
      }];
  }
}
const MARGIN_PX = 8 + 8;
const LINE_PX = 22;
