// @dvag/design-system-angular/esm2015/generated/boolean-value-accessor.js
import { Directive, ElementRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ValueAccessor } from './value-accessor';
export class BooleanValueAccessor extends ValueAccessor {
    constructor(el) {
        super(el);
    }
    writeValue(value) {
        this.el.nativeElement.checked = this.lastValue = value == null ? false : value;
    }
}
BooleanValueAccessor.decorators = [
    { type: Directive, args: [{
                /* tslint:disable-next-line:directive-selector */
                selector: 'ds-checkbox, dx-checkbox, dx-toggle, ds-table-row, dx-table-row, dx-card-header',
                host: {
                    '(checkedChange)': 'handleChangeEvent($event.target.checked)',
                    '(rowSelected)': 'handleChangeEvent($event.target.selected)',
                    '(headerSelected)': 'handleChangeEvent($event.target.selected)'
                },
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: BooleanValueAccessor,
                        multi: true
                    }
                ]
            },] }
];
BooleanValueAccessor.ctorParameters = () => [
    { type: ElementRef }
];
