// @dvag/design-system-angular/esm2015/generated/proxies.js
import { __decorate } from "tslib";
/* tslint:disable */
/* auto-generated angular directive proxies */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, NgZone } from '@angular/core';
import { ProxyCmp, proxyOutputs } from './angular-component-lib/utils';
export class DsActionList {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsActionList.decorators = [
    { type: Component, args: [{
                selector: 'ds-action-list',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsActionList.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DsActionListItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsActionListItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-action-list-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsActionListItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsAlert = class DsAlert {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['buttonClick']);
    }
};
DsAlert.decorators = [
    { type: Component, args: [{
                selector: 'ds-alert',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['actionsdirection', 'visible'],
                outputs: ['buttonClick']
            },] }
];
DsAlert.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsAlert = __decorate([
    ProxyCmp({
        inputs: ['actionsdirection', 'visible'],
        methods: ['show', 'showAlert', 'showConfirm']
    })
], DsAlert);
export { DsAlert };
let DsAlertButton = class DsAlertButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsAlertButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-alert-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'confirm', 'disabled', 'highlight', 'label', 'value']
            },] }
];
DsAlertButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsAlertButton = __decorate([
    ProxyCmp({
        inputs: ['color', 'confirm', 'disabled', 'highlight', 'label', 'value']
    })
], DsAlertButton);
export { DsAlertButton };
export class DsAnnouncementBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsAnnouncementBar.decorators = [
    { type: Component, args: [{
                selector: 'ds-announcement-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsAnnouncementBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DsApp {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsApp.decorators = [
    { type: Component, args: [{
                selector: 'ds-app',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsApp.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsBorderedFormGroup = class DsBorderedFormGroup {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsBorderedFormGroup.decorators = [
    { type: Component, args: [{
                selector: 'ds-bordered-form-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['direction', 'size']
            },] }
];
DsBorderedFormGroup.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsBorderedFormGroup = __decorate([
    ProxyCmp({
        inputs: ['direction', 'size']
    })
], DsBorderedFormGroup);
export { DsBorderedFormGroup };
export class DsBorderedFormItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsBorderedFormItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-bordered-form-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsBorderedFormItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsButton = class DsButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'size', 'stretch', 'type', 'unfocusable']
            },] }
];
DsButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsButton = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'size', 'stretch', 'type', 'unfocusable'],
        methods: ['focusControl']
    })
], DsButton);
export { DsButton };
let DsButtonBar = class DsButtonBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsButtonBar.decorators = [
    { type: Component, args: [{
                selector: 'ds-button-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['direction']
            },] }
];
DsButtonBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsButtonBar = __decorate([
    ProxyCmp({
        inputs: ['direction']
    })
], DsButtonBar);
export { DsButtonBar };
let DsCard = class DsCard {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsCard.decorators = [
    { type: Component, args: [{
                selector: 'ds-card',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['shadow']
            },] }
];
DsCard.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCard = __decorate([
    ProxyCmp({
        inputs: ['shadow']
    })
], DsCard);
export { DsCard };
let DsCardCollapsable = class DsCardCollapsable {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsCardCollapsable.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-collapsable',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'open', 'type']
            },] }
];
DsCardCollapsable.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCardCollapsable = __decorate([
    ProxyCmp({
        inputs: ['label', 'open', 'type']
    })
], DsCardCollapsable);
export { DsCardCollapsable };
export class DsCardContent {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsCardContent.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-content',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsCardContent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DsCardHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsCardHeader.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsCardHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsCardImageHeader = class DsCardImageHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsCardImageHeader.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-image-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['url']
            },] }
];
DsCardImageHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCardImageHeader = __decorate([
    ProxyCmp({
        inputs: ['url']
    })
], DsCardImageHeader);
export { DsCardImageHeader };
let DsCardTabBar = class DsCardTabBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['showChange']);
    }
};
DsCardTabBar.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-tab-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['show'],
                outputs: ['showChange']
            },] }
];
DsCardTabBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCardTabBar = __decorate([
    ProxyCmp({
        inputs: ['show']
    })
], DsCardTabBar);
export { DsCardTabBar };
let DsCardTabPage = class DsCardTabPage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsCardTabPage.decorators = [
    { type: Component, args: [{
                selector: 'ds-card-tab-page',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'name']
            },] }
];
DsCardTabPage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCardTabPage = __decorate([
    ProxyCmp({
        inputs: ['label', 'name']
    })
], DsCardTabPage);
export { DsCardTabPage };
let DsCenteredBoard = class DsCenteredBoard {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsCenteredBoard.decorators = [
    { type: Component, args: [{
                selector: 'ds-centered-board',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['columns']
            },] }
];
DsCenteredBoard.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCenteredBoard = __decorate([
    ProxyCmp({
        inputs: ['columns']
    })
], DsCenteredBoard);
export { DsCenteredBoard };
export class DsCenteredBoardItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsCenteredBoardItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-centered-board-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsCenteredBoardItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsChartSample = class DsChartSample {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DsChartSample.decorators = [
    { type: Component, args: [{
                selector: 'ds-chart-sample',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'formattedValue', 'label', 'value'],
                outputs: ['valueChange']
            },] }
];
DsChartSample.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsChartSample = __decorate([
    ProxyCmp({
        inputs: ['color', 'formattedValue', 'label', 'value'],
        methods: ['toChartSample']
    })
], DsChartSample);
export { DsChartSample };
let DsCheckbox = class DsCheckbox {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['checkedChange']);
    }
};
DsCheckbox.decorators = [
    { type: Component, args: [{
                selector: 'ds-checkbox',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['checked', 'disabled', 'error', 'errormessage', 'label', 'readonly', 'required'],
                outputs: ['checkedChange']
            },] }
];
DsCheckbox.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsCheckbox = __decorate([
    ProxyCmp({
        inputs: ['checked', 'disabled', 'error', 'errormessage', 'label', 'readonly', 'required'],
        methods: ['focusControl']
    })
], DsCheckbox);
export { DsCheckbox };
let DsColumn = class DsColumn {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsColumn.decorators = [
    { type: Component, args: [{
                selector: 'ds-column',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align']
            },] }
];
DsColumn.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsColumn = __decorate([
    ProxyCmp({
        inputs: ['align']
    })
], DsColumn);
export { DsColumn };
let DsDateInput = class DsDateInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete', 'errorFound']);
    }
};
DsDateInput.decorators = [
    { type: Component, args: [{
                selector: 'ds-date-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'max', 'min', 'mode', 'readonly', 'required', 'value'],
                outputs: ['valueChange', 'valueComplete', 'errorFound']
            },] }
];
DsDateInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsDateInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'max', 'min', 'mode', 'readonly', 'required', 'value'],
        methods: ['focusControl']
    })
], DsDateInput);
export { DsDateInput };
let DsDateRange = class DsDateRange {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsDateRange.decorators = [
    { type: Component, args: [{
                selector: 'ds-date-range',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['max', 'min', 'mode', 'value']
            },] }
];
DsDateRange.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsDateRange = __decorate([
    ProxyCmp({
        inputs: ['max', 'min', 'mode', 'value'],
        methods: ['focusControl']
    })
], DsDateRange);
export { DsDateRange };
let DsDonutChart = class DsDonutChart {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['segmentSelect']);
    }
};
DsDonutChart.decorators = [
    { type: Component, args: [{
                selector: 'ds-donut-chart',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['offset', 'radius', 'scale', 'segmentwidth', 'startangle'],
                outputs: ['segmentSelect']
            },] }
];
DsDonutChart.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsDonutChart = __decorate([
    ProxyCmp({
        inputs: ['offset', 'radius', 'scale', 'segmentwidth', 'startangle']
    })
], DsDonutChart);
export { DsDonutChart };
let DsDropdown = class DsDropdown {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DsDropdown.decorators = [
    { type: Component, args: [{
                selector: 'ds-dropdown',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'elseoption', 'elseoptionbuttonlabel', 'elseoptioninputlabel', 'elseoptionselected', 'error', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'textalign', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DsDropdown.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsDropdown = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'elseoption', 'elseoptionbuttonlabel', 'elseoptioninputlabel', 'elseoptionselected', 'error', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'textalign', 'value'],
        methods: ['focusControl']
    })
], DsDropdown);
export { DsDropdown };
let DsDropdownOption = class DsDropdownOption {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsDropdownOption.decorators = [
    { type: Component, args: [{
                selector: 'ds-dropdown-option',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['icon', 'label', 'secondarylabel', 'tertiarylabel', 'value']
            },] }
];
DsDropdownOption.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsDropdownOption = __decorate([
    ProxyCmp({
        inputs: ['icon', 'label', 'secondarylabel', 'tertiarylabel', 'value']
    })
], DsDropdownOption);
export { DsDropdownOption };
let DsFmErgebnisuebersicht = class DsFmErgebnisuebersicht {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsFmErgebnisuebersicht.decorators = [
    { type: Component, args: [{
                selector: 'ds-fm-ergebnisuebersicht',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['donutcenterlabel', 'donutlabel', 'donutsublabel', 'donutvisible', 'firstcolumn', 'firstcolumninfo', 'firstcontributionlabel', 'firstcontributionsublabel', 'secondcolumn', 'secondcolumninfo', 'secondcontributionlabel', 'secondcontributionsublabel']
            },] }
];
DsFmErgebnisuebersicht.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFmErgebnisuebersicht = __decorate([
    ProxyCmp({
        inputs: ['donutcenterlabel', 'donutlabel', 'donutsublabel', 'donutvisible', 'firstcolumn', 'firstcolumninfo', 'firstcontributionlabel', 'firstcontributionsublabel', 'secondcolumn', 'secondcolumninfo', 'secondcontributionlabel', 'secondcontributionsublabel']
    })
], DsFmErgebnisuebersicht);
export { DsFmErgebnisuebersicht };
let DsFmErgebnisuebersichtItem = class DsFmErgebnisuebersichtItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsFmErgebnisuebersichtItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-fm-ergebnisuebersicht-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['firstcolumninfo', 'firstcolumnvalue', 'label', 'secondcolumninfo', 'secondcolumnvalue', 'sublabel']
            },] }
];
DsFmErgebnisuebersichtItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFmErgebnisuebersichtItem = __decorate([
    ProxyCmp({
        inputs: ['firstcolumninfo', 'firstcolumnvalue', 'label', 'secondcolumninfo', 'secondcolumnvalue', 'sublabel']
    })
], DsFmErgebnisuebersichtItem);
export { DsFmErgebnisuebersichtItem };
let DsFmErgebnisuebersichtTable = class DsFmErgebnisuebersichtTable {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsFmErgebnisuebersichtTable.decorators = [
    { type: Component, args: [{
                selector: 'ds-fm-ergebnisuebersicht-table',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label']
            },] }
];
DsFmErgebnisuebersichtTable.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFmErgebnisuebersichtTable = __decorate([
    ProxyCmp({
        inputs: ['label']
    })
], DsFmErgebnisuebersichtTable);
export { DsFmErgebnisuebersichtTable };
let DsFooter = class DsFooter {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['infoButtonClicked']);
    }
};
DsFooter.decorators = [
    { type: Component, args: [{
                selector: 'ds-footer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['companyinfo', 'showimprint', 'showtermsofuse'],
                outputs: ['infoButtonClicked']
            },] }
];
DsFooter.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFooter = __decorate([
    ProxyCmp({
        inputs: ['companyinfo', 'showimprint', 'showtermsofuse']
    })
], DsFooter);
export { DsFooter };
let DsFormLayout = class DsFormLayout {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsFormLayout.decorators = [
    { type: Component, args: [{
                selector: 'ds-form-layout',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['size']
            },] }
];
DsFormLayout.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFormLayout = __decorate([
    ProxyCmp({
        inputs: ['size']
    })
], DsFormLayout);
export { DsFormLayout };
export class DsFullscreenSpinner {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsFullscreenSpinner.decorators = [
    { type: Component, args: [{
                selector: 'ds-fullscreen-spinner',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsFullscreenSpinner.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsFunctionIconButton = class DsFunctionIconButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsFunctionIconButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-function-icon-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'unfocusable']
            },] }
];
DsFunctionIconButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsFunctionIconButton = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'unfocusable'],
        methods: ['focusControl']
    })
], DsFunctionIconButton);
export { DsFunctionIconButton };
let DsGrid = class DsGrid {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsGrid.decorators = [
    { type: Component, args: [{
                selector: 'ds-grid',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['base', 'l', 'm', 's', 'xl']
            },] }
];
DsGrid.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsGrid = __decorate([
    ProxyCmp({
        inputs: ['base', 'l', 'm', 's', 'xl']
    })
], DsGrid);
export { DsGrid };
let DsGroupedTextInput = class DsGroupedTextInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DsGroupedTextInput.decorators = [
    { type: Component, args: [{
                selector: 'ds-grouped-text-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'label', 'mask', 'readonly', 'required', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DsGroupedTextInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsGroupedTextInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'label', 'mask', 'readonly', 'required', 'value']
    })
], DsGroupedTextInput);
export { DsGroupedTextInput };
let DsGrowBox = class DsGrowBox {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['growStart', 'contentVisibilityChange']);
    }
};
DsGrowBox.decorators = [
    { type: Component, args: [{
                selector: 'ds-grow-box',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['animateoninit', 'open', 'slidecontent'],
                outputs: ['growStart', 'contentVisibilityChange']
            },] }
];
DsGrowBox.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsGrowBox = __decorate([
    ProxyCmp({
        inputs: ['animateoninit', 'open', 'slidecontent']
    })
], DsGrowBox);
export { DsGrowBox };
let DsHeader = class DsHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsHeader.decorators = [
    { type: Component, args: [{
                selector: 'ds-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['brand', 'fullscreen']
            },] }
];
DsHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeader = __decorate([
    ProxyCmp({
        inputs: ['brand', 'fullscreen']
    })
], DsHeader);
export { DsHeader };
let DsHeaderAction = class DsHeaderAction {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['interaction']);
    }
};
DsHeaderAction.decorators = [
    { type: Component, args: [{
                selector: 'ds-header-action',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['highlight', 'label'],
                outputs: ['interaction']
            },] }
];
DsHeaderAction.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeaderAction = __decorate([
    ProxyCmp({
        inputs: ['highlight', 'label']
    })
], DsHeaderAction);
export { DsHeaderAction };
let DsHeaderActionList = class DsHeaderActionList {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsHeaderActionList.decorators = [
    { type: Component, args: [{
                selector: 'ds-header-action-list',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'open']
            },] }
];
DsHeaderActionList.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeaderActionList = __decorate([
    ProxyCmp({
        inputs: ['label', 'open']
    })
], DsHeaderActionList);
export { DsHeaderActionList };
let DsHeaderCustomer = class DsHeaderCustomer {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsHeaderCustomer.decorators = [
    { type: Component, args: [{
                selector: 'ds-header-customer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['name', 'typelabel']
            },] }
];
DsHeaderCustomer.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeaderCustomer = __decorate([
    ProxyCmp({
        inputs: ['name', 'typelabel']
    })
], DsHeaderCustomer);
export { DsHeaderCustomer };
let DsHeaderUser = class DsHeaderUser {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['signOut', 'usernameInteraction']);
    }
};
DsHeaderUser.decorators = [
    { type: Component, args: [{
                selector: 'ds-header-user',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['imageurl', 'name', 'usernameinteractable'],
                outputs: ['signOut', 'usernameInteraction']
            },] }
];
DsHeaderUser.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeaderUser = __decorate([
    ProxyCmp({
        inputs: ['imageurl', 'name', 'usernameinteractable']
    })
], DsHeaderUser);
export { DsHeaderUser };
let DsHeroImage = class DsHeroImage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsHeroImage.decorators = [
    { type: Component, args: [{
                selector: 'ds-hero-image',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['url']
            },] }
];
DsHeroImage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsHeroImage = __decorate([
    ProxyCmp({
        inputs: ['url']
    })
], DsHeroImage);
export { DsHeroImage };
export class DsHr {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsHr.decorators = [
    { type: Component, args: [{
                selector: 'ds-hr',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsHr.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsIcon = class DsIcon {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsIcon.decorators = [
    { type: Component, args: [{
                selector: 'ds-icon',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'icon', 'size']
            },] }
];
DsIcon.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsIcon = __decorate([
    ProxyCmp({
        inputs: ['color', 'icon', 'size']
    })
], DsIcon);
export { DsIcon };
let DsIconButton = class DsIconButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsIconButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-icon-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'backgroundcolor', 'color', 'disabled', 'size', 'unfocusable']
            },] }
];
DsIconButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsIconButton = __decorate([
    ProxyCmp({
        inputs: ['align', 'backgroundcolor', 'color', 'disabled', 'size', 'unfocusable'],
        methods: ['focusControl']
    })
], DsIconButton);
export { DsIconButton };
let DsInfo = class DsInfo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsInfo.decorators = [
    { type: Component, args: [{
                selector: 'ds-info',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['hidebutton', 'popupvisible', 'size']
            },] }
];
DsInfo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsInfo = __decorate([
    ProxyCmp({
        inputs: ['hidebutton', 'popupvisible', 'size'],
        methods: ['setModal']
    })
], DsInfo);
export { DsInfo };
let DsLink = class DsLink {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsLink.decorators = [
    { type: Component, args: [{
                selector: 'ds-link',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['target', 'url']
            },] }
];
DsLink.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsLink = __decorate([
    ProxyCmp({
        inputs: ['target', 'url']
    })
], DsLink);
export { DsLink };
let DsLogo = class DsLogo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsLogo.decorators = [
    { type: Component, args: [{
                selector: 'ds-logo',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['type']
            },] }
];
DsLogo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsLogo = __decorate([
    ProxyCmp({
        inputs: ['type']
    })
], DsLogo);
export { DsLogo };
let DsModal = class DsModal {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['modalOpened', 'modalClosed']);
    }
};
DsModal.decorators = [
    { type: Component, args: [{
                selector: 'ds-modal',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['allowbackgroundclose', 'contentcolor', 'label', 'size', 'visible'],
                outputs: ['modalOpened', 'modalClosed']
            },] }
];
DsModal.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsModal = __decorate([
    ProxyCmp({
        inputs: ['allowbackgroundclose', 'contentcolor', 'label', 'size', 'visible']
    })
], DsModal);
export { DsModal };
export class DsNavigatedContent {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsNavigatedContent.decorators = [
    { type: Component, args: [{
                selector: 'ds-navigated-content',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsNavigatedContent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsNavigatedContentMenu = class DsNavigatedContentMenu {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['navigate']);
    }
};
DsNavigatedContentMenu.decorators = [
    { type: Component, args: [{
                selector: 'ds-navigated-content-menu',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['active'],
                outputs: ['navigate']
            },] }
];
DsNavigatedContentMenu.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsNavigatedContentMenu = __decorate([
    ProxyCmp({
        inputs: ['active']
    })
], DsNavigatedContentMenu);
export { DsNavigatedContentMenu };
let DsNavigatedContentMenuItem = class DsNavigatedContentMenuItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsNavigatedContentMenuItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-navigated-content-menu-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['error', 'label', 'success', 'value', 'warning']
            },] }
];
DsNavigatedContentMenuItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsNavigatedContentMenuItem = __decorate([
    ProxyCmp({
        inputs: ['error', 'label', 'success', 'value', 'warning']
    })
], DsNavigatedContentMenuItem);
export { DsNavigatedContentMenuItem };
let DsNotificationToast = class DsNotificationToast {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['closeNotification']);
    }
};
DsNotificationToast.decorators = [
    { type: Component, args: [{
                selector: 'ds-notification-toast',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['closeafter', 'icon', 'iconcolor', 'subcontent', 'titlestring', 'visible'],
                outputs: ['closeNotification']
            },] }
];
DsNotificationToast.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsNotificationToast = __decorate([
    ProxyCmp({
        inputs: ['closeafter', 'icon', 'iconcolor', 'subcontent', 'titlestring', 'visible']
    })
], DsNotificationToast);
export { DsNotificationToast };
let DsNumberInput = class DsNumberInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DsNumberInput.decorators = [
    { type: Component, args: [{
                selector: 'ds-number-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'limitlength', 'max', 'min', 'placeholder', 'precision', 'profile', 'readonly', 'required', 'thousandseparator', 'uniticon', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DsNumberInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsNumberInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'limitlength', 'max', 'min', 'placeholder', 'precision', 'profile', 'readonly', 'required', 'thousandseparator', 'uniticon', 'value'],
        methods: ['focusControl']
    })
], DsNumberInput);
export { DsNumberInput };
export class DsPage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsPage.decorators = [
    { type: Component, args: [{
                selector: 'ds-page',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsPage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsPaginator = class DsPaginator {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['pageChanged']);
    }
};
DsPaginator.decorators = [
    { type: Component, args: [{
                selector: 'ds-paginator',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['count', 'current', 'range'],
                outputs: ['pageChanged']
            },] }
];
DsPaginator.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsPaginator = __decorate([
    ProxyCmp({
        inputs: ['count', 'current', 'range']
    })
], DsPaginator);
export { DsPaginator };
let DsPlaceholder = class DsPlaceholder {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsPlaceholder.decorators = [
    { type: Component, args: [{
                selector: 'ds-placeholder',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['box', 'card']
            },] }
];
DsPlaceholder.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsPlaceholder = __decorate([
    ProxyCmp({
        inputs: ['box', 'card']
    })
], DsPlaceholder);
export { DsPlaceholder };
let DsRadioButton = class DsRadioButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsRadioButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-radio-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'label', 'value']
            },] }
];
DsRadioButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsRadioButton = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'label', 'value'],
        methods: ['focusControl']
    })
], DsRadioButton);
export { DsRadioButton };
let DsRadioButtonGroup = class DsRadioButtonGroup {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DsRadioButtonGroup.decorators = [
    { type: Component, args: [{
                selector: 'ds-radio-button-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'readonly', 'required', 'value'],
                outputs: ['valueChange']
            },] }
];
DsRadioButtonGroup.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsRadioButtonGroup = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'readonly', 'required', 'value'],
        methods: ['focusControl']
    })
], DsRadioButtonGroup);
export { DsRadioButtonGroup };
let DsSearch = class DsSearch {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['search', 'selectSearch', 'valueChange', 'valueComplete']);
    }
};
DsSearch.decorators = [
    { type: Component, args: [{
                selector: 'ds-search',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['autoscrollintoview', 'noresultstext', 'placeholder', 'value', 'visibleentries'],
                outputs: ['search', 'selectSearch', 'valueChange', 'valueComplete']
            },] }
];
DsSearch.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsSearch = __decorate([
    ProxyCmp({
        inputs: ['autoscrollintoview', 'noresultstext', 'placeholder', 'value', 'visibleentries'],
        methods: ['focusControl']
    })
], DsSearch);
export { DsSearch };
let DsSlider = class DsSlider {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DsSlider.decorators = [
    { type: Component, args: [{
                selector: 'ds-slider',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['formattedvalue', 'label', 'max', 'min', 'readonly', 'step', 'value'],
                outputs: ['valueChange']
            },] }
];
DsSlider.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsSlider = __decorate([
    ProxyCmp({
        inputs: ['formattedvalue', 'label', 'max', 'min', 'readonly', 'step', 'value']
    })
], DsSlider);
export { DsSlider };
let DsSpinner = class DsSpinner {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsSpinner.decorators = [
    { type: Component, args: [{
                selector: 'ds-spinner',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['size', 'theme']
            },] }
];
DsSpinner.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsSpinner = __decorate([
    ProxyCmp({
        inputs: ['size', 'theme']
    })
], DsSpinner);
export { DsSpinner };
let DsTable = class DsTable {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['tableStateChanged', 'rowsSelected']);
    }
};
DsTable.decorators = [
    { type: Component, args: [{
                selector: 'ds-table',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['initialsort', 'initialsortdirection', 'mode', 'pagesize', 'pagesizes', 'rowcount', 'scrollable', 'selectionenabled', 'selectionmode', 'view'],
                outputs: ['tableStateChanged', 'rowsSelected']
            },] }
];
DsTable.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTable = __decorate([
    ProxyCmp({
        inputs: ['initialsort', 'initialsortdirection', 'mode', 'pagesize', 'pagesizes', 'rowcount', 'scrollable', 'selectionenabled', 'selectionmode', 'view']
    })
], DsTable);
export { DsTable };
let DsTableCell = class DsTableCell {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsTableCell.decorators = [
    { type: Component, args: [{
                selector: 'ds-table-cell',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'type', 'value']
            },] }
];
DsTableCell.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTableCell = __decorate([
    ProxyCmp({
        inputs: ['color', 'type', 'value']
    })
], DsTableCell);
export { DsTableCell };
export class DsTableFooter {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsTableFooter.decorators = [
    { type: Component, args: [{
                selector: 'ds-table-footer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsTableFooter.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DsTableHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DsTableHeader.decorators = [
    { type: Component, args: [{
                selector: 'ds-table-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DsTableHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DsTableHeaderItem = class DsTableHeaderItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsTableHeaderItem.decorators = [
    { type: Component, args: [{
                selector: 'ds-table-header-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'columnid', 'customSort', 'label', 'linesize', 'maxwidth', 'minwidth', 'sortable', 'sortpriority', 'type']
            },] }
];
DsTableHeaderItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTableHeaderItem = __decorate([
    ProxyCmp({
        inputs: ['align', 'columnid', 'customSort', 'label', 'linesize', 'maxwidth', 'minwidth', 'sortable', 'sortpriority', 'type']
    })
], DsTableHeaderItem);
export { DsTableHeaderItem };
let DsTableRow = class DsTableRow {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['rowSelected']);
    }
};
DsTableRow.decorators = [
    { type: Component, args: [{
                selector: 'ds-table-row',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['rowid', 'selected'],
                outputs: ['rowSelected']
            },] }
];
DsTableRow.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTableRow = __decorate([
    ProxyCmp({
        inputs: ['rowid', 'selected']
    })
], DsTableRow);
export { DsTableRow };
let DsText = class DsText {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsText.decorators = [
    { type: Component, args: [{
                selector: 'ds-text',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'screensize', 'type']
            },] }
];
DsText.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsText = __decorate([
    ProxyCmp({
        inputs: ['color', 'screensize', 'type']
    })
], DsText);
export { DsText };
let DsTextArea = class DsTextArea {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DsTextArea.decorators = [
    { type: Component, args: [{
                selector: 'ds-text-area',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'label', 'lines', 'maxlength', 'readonly', 'required', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DsTextArea.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTextArea = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'label', 'lines', 'maxlength', 'readonly', 'required', 'value'],
        methods: ['focusControl']
    })
], DsTextArea);
export { DsTextArea };
let DsTextButton = class DsTextButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsTextButton.decorators = [
    { type: Component, args: [{
                selector: 'ds-text-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'disabled', 'screensize', 'unfocusable']
            },] }
];
DsTextButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTextButton = __decorate([
    ProxyCmp({
        inputs: ['color', 'disabled', 'screensize', 'unfocusable'],
        methods: ['focusControl']
    })
], DsTextButton);
export { DsTextButton };
let DsTextInput = class DsTextInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete', 'autocomplete', 'selectAutocomplete']);
    }
};
DsTextInput.decorators = [
    { type: Component, args: [{
                selector: 'ds-text-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'limitlength', 'mask', 'placeholder', 'readonly', 'required', 'value'],
                outputs: ['valueChange', 'valueComplete', 'autocomplete', 'selectAutocomplete']
            },] }
];
DsTextInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTextInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'kisynced', 'label', 'limitlength', 'mask', 'placeholder', 'readonly', 'required', 'value'],
        methods: ['focusControl']
    })
], DsTextInput);
export { DsTextInput };
let DsTextList = class DsTextList {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsTextList.decorators = [
    { type: Component, args: [{
                selector: 'ds-text-list',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['type']
            },] }
];
DsTextList.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsTextList = __decorate([
    ProxyCmp({
        inputs: ['type']
    })
], DsTextList);
export { DsTextList };
let DsUserAvatar = class DsUserAvatar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsUserAvatar.decorators = [
    { type: Component, args: [{
                selector: 'ds-user-avatar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['imageurl', 'initials', 'size']
            },] }
];
DsUserAvatar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsUserAvatar = __decorate([
    ProxyCmp({
        inputs: ['imageurl', 'initials', 'size']
    })
], DsUserAvatar);
export { DsUserAvatar };
let DsWorkflowProgress = class DsWorkflowProgress {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsWorkflowProgress.decorators = [
    { type: Component, args: [{
                selector: 'ds-workflow-progress',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['progress']
            },] }
];
DsWorkflowProgress.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsWorkflowProgress = __decorate([
    ProxyCmp({
        inputs: ['progress']
    })
], DsWorkflowProgress);
export { DsWorkflowProgress };
let DsWrappedLayout = class DsWrappedLayout {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DsWrappedLayout.decorators = [
    { type: Component, args: [{
                selector: 'ds-wrapped-layout',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'size']
            },] }
];
DsWrappedLayout.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsWrappedLayout = __decorate([
    ProxyCmp({
        inputs: ['align', 'size']
    })
], DsWrappedLayout);
export { DsWrappedLayout };
let DsYesNo = class DsYesNo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DsYesNo.decorators = [
    { type: Component, args: [{
                selector: 'ds-yes-no',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'error', 'errormessage', 'readonly', 'required', 'size', 'value'],
                outputs: ['valueChange']
            },] }
];
DsYesNo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DsYesNo = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'error', 'errormessage', 'readonly', 'required', 'size', 'value'],
        methods: ['focusControl']
    })
], DsYesNo);
export { DsYesNo };
let DxAccordion = class DxAccordion {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxAccordion.decorators = [
    { type: Component, args: [{
                selector: 'dx-accordion',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'open']
            },] }
];
DxAccordion.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxAccordion = __decorate([
    ProxyCmp({
        inputs: ['label', 'open']
    })
], DxAccordion);
export { DxAccordion };
export class DxAccordionGroup {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxAccordionGroup.decorators = [
    { type: Component, args: [{
                selector: 'dx-accordion-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxAccordionGroup.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxAlert = class DxAlert {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxAlert.decorators = [
    { type: Component, args: [{
                selector: 'dx-alert',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['actionsdirection', 'content', 'icon', 'sublabel', 'titlestring', 'type', 'visible']
            },] }
];
DxAlert.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxAlert = __decorate([
    ProxyCmp({
        inputs: ['actionsdirection', 'content', 'icon', 'sublabel', 'titlestring', 'type', 'visible']
    })
], DxAlert);
export { DxAlert };
let DxAnnouncementBar = class DxAnnouncementBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxAnnouncementBar.decorators = [
    { type: Component, args: [{
                selector: 'dx-announcement-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'icon', 'iconcolor']
            },] }
];
DxAnnouncementBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxAnnouncementBar = __decorate([
    ProxyCmp({
        inputs: ['color', 'icon', 'iconcolor']
    })
], DxAnnouncementBar);
export { DxAnnouncementBar };
export class DxApp {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxApp.decorators = [
    { type: Component, args: [{
                selector: 'dx-app',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxApp.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DxBreadcrumb {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['navigateHome', 'navigate']);
    }
}
DxBreadcrumb.decorators = [
    { type: Component, args: [{
                selector: 'dx-breadcrumb',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                outputs: ['navigateHome', 'navigate']
            },] }
];
DxBreadcrumb.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxBreadcrumbItem = class DxBreadcrumbItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxBreadcrumbItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-breadcrumb-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'value']
            },] }
];
DxBreadcrumbItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxBreadcrumbItem = __decorate([
    ProxyCmp({
        inputs: ['label', 'value']
    })
], DxBreadcrumbItem);
export { DxBreadcrumbItem };
let DxButton = class DxButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxButton.decorators = [
    { type: Component, args: [{
                selector: 'dx-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'icon', 'iconposition', 'label', 'loading', 'stretch', 'theme', 'type', 'unfocusable']
            },] }
];
DxButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxButton = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'icon', 'iconposition', 'label', 'loading', 'stretch', 'theme', 'type', 'unfocusable'],
        methods: ['focusControl']
    })
], DxButton);
export { DxButton };
export class DxCard {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxCard.decorators = [
    { type: Component, args: [{
                selector: 'dx-card',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxCard.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxCardCollapsable = class DxCardCollapsable {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxCardCollapsable.decorators = [
    { type: Component, args: [{
                selector: 'dx-card-collapsable',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['open']
            },] }
];
DxCardCollapsable.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxCardCollapsable = __decorate([
    ProxyCmp({
        inputs: ['open']
    })
], DxCardCollapsable);
export { DxCardCollapsable };
export class DxCardContent {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxCardContent.decorators = [
    { type: Component, args: [{
                selector: 'dx-card-content',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxCardContent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxCardHeader = class DxCardHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxCardHeader.decorators = [
    { type: Component, args: [{
                selector: 'dx-card-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['icon', 'iconcolor', 'label']
            },] }
];
DxCardHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxCardHeader = __decorate([
    ProxyCmp({
        inputs: ['icon', 'iconcolor', 'label']
    })
], DxCardHeader);
export { DxCardHeader };
let DxCardImage = class DxCardImage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxCardImage.decorators = [
    { type: Component, args: [{
                selector: 'dx-card-image',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['coverheightmq1', 'coverheightmq2', 'coverheightmq3', 'coverheightmq4', 'coverheightmq5', 'coverpositionmq1', 'coverpositionmq2', 'coverpositionmq3', 'coverpositionmq4', 'coverpositionmq5', 'fit', 'url']
            },] }
];
DxCardImage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxCardImage = __decorate([
    ProxyCmp({
        inputs: ['coverheightmq1', 'coverheightmq2', 'coverheightmq3', 'coverheightmq4', 'coverheightmq5', 'coverpositionmq1', 'coverpositionmq2', 'coverpositionmq3', 'coverpositionmq4', 'coverpositionmq5', 'fit', 'url']
    })
], DxCardImage);
export { DxCardImage };
let DxChartSample = class DxChartSample {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DxChartSample.decorators = [
    { type: Component, args: [{
                selector: 'dx-chart-sample',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'icon', 'label', 'selectedcolor', 'value'],
                outputs: ['valueChange']
            },] }
];
DxChartSample.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxChartSample = __decorate([
    ProxyCmp({
        inputs: ['color', 'icon', 'label', 'selectedcolor', 'value'],
        methods: ['toChartSample']
    })
], DxChartSample);
export { DxChartSample };
let DxCheckbox = class DxCheckbox {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['checkedChange']);
    }
};
DxCheckbox.decorators = [
    { type: Component, args: [{
                selector: 'dx-checkbox',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['checked', 'disabled', 'errormessage', 'kisynced', 'label', 'readonly', 'required'],
                outputs: ['checkedChange']
            },] }
];
DxCheckbox.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxCheckbox = __decorate([
    ProxyCmp({
        inputs: ['checked', 'disabled', 'errormessage', 'kisynced', 'label', 'readonly', 'required'],
        methods: ['focusControl']
    })
], DxCheckbox);
export { DxCheckbox };
let DxCombobox = class DxCombobox {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete', 'search', 'selectSearchOption', 'enterPress']);
    }
};
DxCombobox.decorators = [
    { type: Component, args: [{
                selector: 'dx-combobox',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'emptysearchtext', 'errormessage', 'kisynced', 'label', 'limitlength', 'loadingtext', 'maxlistheight', 'placeholder', 'presearchconditiontext', 'presearchlettercount', 'readonly', 'required', 'size', 'value'],
                outputs: ['valueChange', 'valueComplete', 'search', 'selectSearchOption', 'enterPress']
            },] }
];
DxCombobox.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxCombobox = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'emptysearchtext', 'errormessage', 'kisynced', 'label', 'limitlength', 'loadingtext', 'maxlistheight', 'placeholder', 'presearchconditiontext', 'presearchlettercount', 'readonly', 'required', 'size', 'value'],
        methods: ['focusControl']
    })
], DxCombobox);
export { DxCombobox };
let DxContainer = class DxContainer {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxContainer.decorators = [
    { type: Component, args: [{
                selector: 'dx-container',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'type']
            },] }
];
DxContainer.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxContainer = __decorate([
    ProxyCmp({
        inputs: ['color', 'type']
    })
], DxContainer);
export { DxContainer };
let DxDateInput = class DxDateInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete', 'errorFound']);
    }
};
DxDateInput.decorators = [
    { type: Component, args: [{
                selector: 'dx-date-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'max', 'min', 'mode', 'readonly', 'required', 'size', 'value'],
                outputs: ['valueChange', 'valueComplete', 'errorFound']
            },] }
];
DxDateInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxDateInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'max', 'min', 'mode', 'readonly', 'required', 'size', 'value'],
        methods: ['focusControl']
    })
], DxDateInput);
export { DxDateInput };
let DxDateRange = class DxDateRange {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxDateRange.decorators = [
    { type: Component, args: [{
                selector: 'dx-date-range',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['max', 'min', 'mode', 'value']
            },] }
];
DxDateRange.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxDateRange = __decorate([
    ProxyCmp({
        inputs: ['max', 'min', 'mode', 'value'],
        methods: ['focusControl']
    })
], DxDateRange);
export { DxDateRange };
let DxDonutChart = class DxDonutChart {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['segmentSelect']);
    }
};
DxDonutChart.decorators = [
    { type: Component, args: [{
                selector: 'dx-donut-chart',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['colorscheme', 'label', 'selected', 'showsegmentlabels', 'sublabel'],
                outputs: ['segmentSelect']
            },] }
];
DxDonutChart.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxDonutChart = __decorate([
    ProxyCmp({
        inputs: ['colorscheme', 'label', 'selected', 'showsegmentlabels', 'sublabel'],
        methods: ['update']
    })
], DxDonutChart);
export { DxDonutChart };
let DxDropdown = class DxDropdown {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DxDropdown.decorators = [
    { type: Component, args: [{
                selector: 'dx-dropdown',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'elseoption', 'elseoptionbuttonlabel', 'elseoptioninputlabel', 'elseoptionselected', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'size', 'textalign', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DxDropdown.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxDropdown = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'elseoption', 'elseoptionbuttonlabel', 'elseoptioninputlabel', 'elseoptionselected', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'size', 'textalign', 'value'],
        methods: ['focusControl']
    })
], DxDropdown);
export { DxDropdown };
let DxDropdownOption = class DxDropdownOption {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxDropdownOption.decorators = [
    { type: Component, args: [{
                selector: 'dx-dropdown-option',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['icon', 'label', 'selected', 'value']
            },] }
];
DxDropdownOption.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxDropdownOption = __decorate([
    ProxyCmp({
        inputs: ['icon', 'label', 'selected', 'value']
    })
], DxDropdownOption);
export { DxDropdownOption };
let DxErrorMessage = class DxErrorMessage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['forceUpdateOnChange']);
    }
};
DxErrorMessage.decorators = [
    { type: Component, args: [{
                selector: 'dx-error-message',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['lines'],
                outputs: ['forceUpdateOnChange']
            },] }
];
DxErrorMessage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxErrorMessage = __decorate([
    ProxyCmp({
        inputs: ['lines']
    })
], DxErrorMessage);
export { DxErrorMessage };
let DxErrorPage = class DxErrorPage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['action']);
    }
};
DxErrorPage.decorators = [
    { type: Component, args: [{
                selector: 'dx-error-page',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['actiondescription', 'actionlabel', 'description', 'reason'],
                outputs: ['action']
            },] }
];
DxErrorPage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxErrorPage = __decorate([
    ProxyCmp({
        inputs: ['actiondescription', 'actionlabel', 'description', 'reason']
    })
], DxErrorPage);
export { DxErrorPage };
export class DxExpanded {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxExpanded.decorators = [
    { type: Component, args: [{
                selector: 'dx-expanded',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxExpanded.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DxFdOutlineBox {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxFdOutlineBox.decorators = [
    { type: Component, args: [{
                selector: 'dx-fd-outline-box',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxFdOutlineBox.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxFdSeparator = class DxFdSeparator {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxFdSeparator.decorators = [
    { type: Component, args: [{
                selector: 'dx-fd-separator',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['fromcolor', 'tocolor']
            },] }
];
DxFdSeparator.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxFdSeparator = __decorate([
    ProxyCmp({
        inputs: ['fromcolor', 'tocolor']
    })
], DxFdSeparator);
export { DxFdSeparator };
let DxFooter = class DxFooter {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxFooter.decorators = [
    { type: Component, args: [{
                selector: 'dx-footer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['fromcolor']
            },] }
];
DxFooter.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxFooter = __decorate([
    ProxyCmp({
        inputs: ['fromcolor']
    })
], DxFooter);
export { DxFooter };
let DxFormWrapper = class DxFormWrapper {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxFormWrapper.decorators = [
    { type: Component, args: [{
                selector: 'dx-form-wrapper',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['errormessage', 'kisynced', 'label', 'required']
            },] }
];
DxFormWrapper.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxFormWrapper = __decorate([
    ProxyCmp({
        inputs: ['errormessage', 'kisynced', 'label', 'required']
    })
], DxFormWrapper);
export { DxFormWrapper };
let DxGoldenStage = class DxGoldenStage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxGoldenStage.decorators = [
    { type: Component, args: [{
                selector: 'dx-golden-stage',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['alignment', 'backgroundcolor', 'contentoffsety', 'size']
            },] }
];
DxGoldenStage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxGoldenStage = __decorate([
    ProxyCmp({
        inputs: ['alignment', 'backgroundcolor', 'contentoffsety', 'size']
    })
], DxGoldenStage);
export { DxGoldenStage };
let DxGrid = class DxGrid {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxGrid.decorators = [
    { type: Component, args: [{
                selector: 'dx-grid',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['base', 'mq1', 'mq2', 'mq3', 'mq4', 'mq5', 'rowgap']
            },] }
];
DxGrid.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxGrid = __decorate([
    ProxyCmp({
        inputs: ['base', 'mq1', 'mq2', 'mq3', 'mq4', 'mq5', 'rowgap']
    })
], DxGrid);
export { DxGrid };
let DxHeader = class DxHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['navigateHome']);
    }
};
DxHeader.decorators = [
    { type: Component, args: [{
                selector: 'dx-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['appname', 'brand', 'logo', 'showannouncement', 'theme'],
                outputs: ['navigateHome']
            },] }
];
DxHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeader = __decorate([
    ProxyCmp({
        inputs: ['appname', 'brand', 'logo', 'showannouncement', 'theme']
    })
], DxHeader);
export { DxHeader };
let DxHeaderAppFavorites = class DxHeaderAppFavorites {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['menuInteraction', 'deleteApp', 'sortApps', 'addApp', 'editApps', 'launchApp']);
    }
};
DxHeaderAppFavorites.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-app-favorites',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['notifications', 'notificationstheme', 'priority'],
                outputs: ['menuInteraction', 'deleteApp', 'sortApps', 'addApp', 'editApps', 'launchApp']
            },] }
];
DxHeaderAppFavorites.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderAppFavorites = __decorate([
    ProxyCmp({
        inputs: ['notifications', 'notificationstheme', 'priority']
    })
], DxHeaderAppFavorites);
export { DxHeaderAppFavorites };
let DxHeaderAppFavoritesApp = class DxHeaderAppFavoritesApp {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxHeaderAppFavoritesApp.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-app-favorites-app',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['appid', 'appname', 'icon']
            },] }
];
DxHeaderAppFavoritesApp.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderAppFavoritesApp = __decorate([
    ProxyCmp({
        inputs: ['appid', 'appname', 'icon'],
        methods: ['toAppFavoritesApp']
    })
], DxHeaderAppFavoritesApp);
export { DxHeaderAppFavoritesApp };
let DxHeaderMenuCustom = class DxHeaderMenuCustom {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxHeaderMenuCustom.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-menu-custom',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['priority']
            },] }
];
DxHeaderMenuCustom.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderMenuCustom = __decorate([
    ProxyCmp({
        inputs: ['priority']
    })
], DxHeaderMenuCustom);
export { DxHeaderMenuCustom };
let DxHeaderMenuGroup = class DxHeaderMenuGroup {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxHeaderMenuGroup.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-menu-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['priority']
            },] }
];
DxHeaderMenuGroup.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderMenuGroup = __decorate([
    ProxyCmp({
        inputs: ['priority']
    })
], DxHeaderMenuGroup);
export { DxHeaderMenuGroup };
let DxHeaderMenuItem = class DxHeaderMenuItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['menuInteraction']);
    }
};
DxHeaderMenuItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-menu-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['connectedslot', 'icon', 'notifications', 'notificationstheme', 'priority'],
                outputs: ['menuInteraction']
            },] }
];
DxHeaderMenuItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderMenuItem = __decorate([
    ProxyCmp({
        inputs: ['connectedslot', 'icon', 'notifications', 'notificationstheme', 'priority']
    })
], DxHeaderMenuItem);
export { DxHeaderMenuItem };
let DxHeaderMenuLink = class DxHeaderMenuLink {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['menuInteraction']);
    }
};
DxHeaderMenuLink.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-menu-link',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['connectednavigationvalue', 'label', 'priority', 'selected'],
                outputs: ['menuInteraction']
            },] }
];
DxHeaderMenuLink.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderMenuLink = __decorate([
    ProxyCmp({
        inputs: ['connectednavigationvalue', 'label', 'priority', 'selected']
    })
], DxHeaderMenuLink);
export { DxHeaderMenuLink };
let DxHeaderNavigation = class DxHeaderNavigation {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['menuInteraction']);
    }
};
DxHeaderNavigation.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-navigation',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['autoclose', 'notification', 'notificationstheme'],
                outputs: ['menuInteraction']
            },] }
];
DxHeaderNavigation.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderNavigation = __decorate([
    ProxyCmp({
        inputs: ['autoclose', 'notification', 'notificationstheme']
    })
], DxHeaderNavigation);
export { DxHeaderNavigation };
let DxHeaderNavigationItem = class DxHeaderNavigationItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['interaction']);
    }
};
DxHeaderNavigationItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-navigation-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'size', 'value'],
                outputs: ['interaction']
            },] }
];
DxHeaderNavigationItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderNavigationItem = __decorate([
    ProxyCmp({
        inputs: ['label', 'size', 'value']
    })
], DxHeaderNavigationItem);
export { DxHeaderNavigationItem };
let DxHeaderUser = class DxHeaderUser {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['signOut', 'accountInteraction', 'menuInteraction']);
    }
};
DxHeaderUser.decorators = [
    { type: Component, args: [{
                selector: 'dx-header-user',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['accountavailable', 'imageurl', 'name', 'notifications', 'notificationstheme', 'priority'],
                outputs: ['signOut', 'accountInteraction', 'menuInteraction']
            },] }
];
DxHeaderUser.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeaderUser = __decorate([
    ProxyCmp({
        inputs: ['accountavailable', 'imageurl', 'name', 'notifications', 'notificationstheme', 'priority']
    })
], DxHeaderUser);
export { DxHeaderUser };
let DxHeroImage = class DxHeroImage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxHeroImage.decorators = [
    { type: Component, args: [{
                selector: 'dx-hero-image',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5']
            },] }
];
DxHeroImage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxHeroImage = __decorate([
    ProxyCmp({
        inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5']
    })
], DxHeroImage);
export { DxHeroImage };
let DxIcon = class DxIcon {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxIcon.decorators = [
    { type: Component, args: [{
                selector: 'dx-icon',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'icon', 'size']
            },] }
];
DxIcon.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxIcon = __decorate([
    ProxyCmp({
        inputs: ['color', 'icon', 'size']
    })
], DxIcon);
export { DxIcon };
let DxInfo = class DxInfo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxInfo.decorators = [
    { type: Component, args: [{
                selector: 'dx-info',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['hidebutton', 'size']
            },] }
];
DxInfo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxInfo = __decorate([
    ProxyCmp({
        inputs: ['hidebutton', 'size'],
        methods: ['setModal']
    })
], DxInfo);
export { DxInfo };
let DxLegend = class DxLegend {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxLegend.decorators = [
    { type: Component, args: [{
                selector: 'dx-legend',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['kisynchint', 'requiredhint']
            },] }
];
DxLegend.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxLegend = __decorate([
    ProxyCmp({
        inputs: ['kisynchint', 'requiredhint']
    })
], DxLegend);
export { DxLegend };
let DxLink = class DxLink {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxLink.decorators = [
    { type: Component, args: [{
                selector: 'dx-link',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['target', 'url']
            },] }
];
DxLink.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxLink = __decorate([
    ProxyCmp({
        inputs: ['target', 'url']
    })
], DxLink);
export { DxLink };
let DxList = class DxList {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxList.decorators = [
    { type: Component, args: [{
                selector: 'dx-list',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['contentbelow', 'size', 'width']
            },] }
];
DxList.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxList = __decorate([
    ProxyCmp({
        inputs: ['contentbelow', 'size', 'width']
    })
], DxList);
export { DxList };
let DxListItem = class DxListItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxListItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-list-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['actionindicator', 'disabled', 'icon', 'label', 'selected', 'sublabel', 'textalign', 'type', 'value']
            },] }
];
DxListItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxListItem = __decorate([
    ProxyCmp({
        inputs: ['actionindicator', 'disabled', 'icon', 'label', 'selected', 'sublabel', 'textalign', 'type', 'value'],
        methods: ['getRadioElement', 'focusControl']
    })
], DxListItem);
export { DxListItem };
let DxLogo = class DxLogo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxLogo.decorators = [
    { type: Component, args: [{
                selector: 'dx-logo',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['type']
            },] }
];
DxLogo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxLogo = __decorate([
    ProxyCmp({
        inputs: ['type']
    })
], DxLogo);
export { DxLogo };
let DxMediaQueryValue = class DxMediaQueryValue {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxMediaQueryValue.decorators = [
    { type: Component, args: [{
                selector: 'dx-media-query-value',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5', 'property']
            },] }
];
DxMediaQueryValue.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxMediaQueryValue = __decorate([
    ProxyCmp({
        inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5', 'property']
    })
], DxMediaQueryValue);
export { DxMediaQueryValue };
let DxModal = class DxModal {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['modalOpened', 'modalClosed']);
    }
};
DxModal.decorators = [
    { type: Component, args: [{
                selector: 'dx-modal',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['height', 'label', 'visible', 'width'],
                outputs: ['modalOpened', 'modalClosed']
            },] }
];
DxModal.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxModal = __decorate([
    ProxyCmp({
        inputs: ['height', 'label', 'visible', 'width']
    })
], DxModal);
export { DxModal };
let DxMultiDropdown = class DxMultiDropdown {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DxMultiDropdown.decorators = [
    { type: Component, args: [{
                selector: 'dx-multi-dropdown',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'selectionlabel', 'size', 'textalign'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DxMultiDropdown.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxMultiDropdown = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'icon', 'kisynced', 'label', 'placeholder', 'popupposition', 'readonly', 'required', 'selectionlabel', 'size', 'textalign'],
        methods: ['focusControl']
    })
], DxMultiDropdown);
export { DxMultiDropdown };
let DxNotificationBar = class DxNotificationBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['closeNotification']);
    }
};
DxNotificationBar.decorators = [
    { type: Component, args: [{
                selector: 'dx-notification-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['visible'],
                outputs: ['closeNotification']
            },] }
];
DxNotificationBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxNotificationBar = __decorate([
    ProxyCmp({
        inputs: ['visible']
    })
], DxNotificationBar);
export { DxNotificationBar };
let DxNotificationToast = class DxNotificationToast {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['closeNotification']);
    }
};
DxNotificationToast.decorators = [
    { type: Component, args: [{
                selector: 'dx-notification-toast',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['closeafter', 'icon', 'iconcolor', 'subcontent', 'titlestring', 'visible'],
                outputs: ['closeNotification']
            },] }
];
DxNotificationToast.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxNotificationToast = __decorate([
    ProxyCmp({
        inputs: ['closeafter', 'icon', 'iconcolor', 'subcontent', 'titlestring', 'visible']
    })
], DxNotificationToast);
export { DxNotificationToast };
let DxNumberInput = class DxNumberInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DxNumberInput.decorators = [
    { type: Component, args: [{
                selector: 'dx-number-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'limitlength', 'max', 'min', 'placeholder', 'precision', 'profile', 'readonly', 'required', 'size', 'step', 'thousandseparator', 'unit', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DxNumberInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxNumberInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'limitlength', 'max', 'min', 'placeholder', 'precision', 'profile', 'readonly', 'required', 'size', 'step', 'thousandseparator', 'unit', 'value'],
        methods: ['focusControl']
    })
], DxNumberInput);
export { DxNumberInput };
let DxNumberPaginator = class DxNumberPaginator {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['pageChanged']);
    }
};
DxNumberPaginator.decorators = [
    { type: Component, args: [{
                selector: 'dx-number-paginator',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['count', 'current'],
                outputs: ['pageChanged']
            },] }
];
DxNumberPaginator.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxNumberPaginator = __decorate([
    ProxyCmp({
        inputs: ['count', 'current']
    })
], DxNumberPaginator);
export { DxNumberPaginator };
let DxNumberPicker = class DxNumberPicker {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DxNumberPicker.decorators = [
    { type: Component, args: [{
                selector: 'dx-number-picker',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'value', 'values'],
                outputs: ['valueChange']
            },] }
];
DxNumberPicker.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxNumberPicker = __decorate([
    ProxyCmp({
        inputs: ['align', 'value', 'values']
    })
], DxNumberPicker);
export { DxNumberPicker };
let DxOverflowMenu = class DxOverflowMenu {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['selection', 'listOpenChange']);
    }
};
DxOverflowMenu.decorators = [
    { type: Component, args: [{
                selector: 'dx-overflow-menu',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'label', 'maxlabels', 'maxselections', 'selectionmode'],
                outputs: ['selection', 'listOpenChange']
            },] }
];
DxOverflowMenu.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxOverflowMenu = __decorate([
    ProxyCmp({
        inputs: ['align', 'label', 'maxlabels', 'maxselections', 'selectionmode']
    })
], DxOverflowMenu);
export { DxOverflowMenu };
let DxOverflowMenuItem = class DxOverflowMenuItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxOverflowMenuItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-overflow-menu-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['icon', 'label', 'selected', 'value']
            },] }
];
DxOverflowMenuItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxOverflowMenuItem = __decorate([
    ProxyCmp({
        inputs: ['icon', 'label', 'selected', 'value']
    })
], DxOverflowMenuItem);
export { DxOverflowMenuItem };
let DxPaginator = class DxPaginator {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['pageChanged']);
    }
};
DxPaginator.decorators = [
    { type: Component, args: [{
                selector: 'dx-paginator',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['count', 'current'],
                outputs: ['pageChanged']
            },] }
];
DxPaginator.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxPaginator = __decorate([
    ProxyCmp({
        inputs: ['count', 'current']
    })
], DxPaginator);
export { DxPaginator };
let DxPlaceholder = class DxPlaceholder {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxPlaceholder.decorators = [
    { type: Component, args: [{
                selector: 'dx-placeholder',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['box', 'card']
            },] }
];
DxPlaceholder.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxPlaceholder = __decorate([
    ProxyCmp({
        inputs: ['box', 'card']
    })
], DxPlaceholder);
export { DxPlaceholder };
let DxProcessHeader = class DxProcessHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxProcessHeader.decorators = [
    { type: Component, args: [{
                selector: 'dx-process-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['headline']
            },] }
];
DxProcessHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxProcessHeader = __decorate([
    ProxyCmp({
        inputs: ['headline']
    })
], DxProcessHeader);
export { DxProcessHeader };
let DxRadioButton = class DxRadioButton {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxRadioButton.decorators = [
    { type: Component, args: [{
                selector: 'dx-radio-button',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'label', 'value']
            },] }
];
DxRadioButton.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxRadioButton = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'label', 'value'],
        methods: ['focusControl']
    })
], DxRadioButton);
export { DxRadioButton };
let DxRadioButtonGroup = class DxRadioButtonGroup {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DxRadioButtonGroup.decorators = [
    { type: Component, args: [{
                selector: 'dx-radio-button-group',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'readonly', 'required', 'value'],
                outputs: ['valueChange']
            },] }
];
DxRadioButtonGroup.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxRadioButtonGroup = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'readonly', 'required', 'value'],
        methods: ['focusControl', 'connectRadioButton', 'disconnectRadioButton']
    })
], DxRadioButtonGroup);
export { DxRadioButtonGroup };
let DxRepresentativeInfo = class DxRepresentativeInfo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxRepresentativeInfo.decorators = [
    { type: Component, args: [{
                selector: 'dx-representative-info',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['name', 'type']
            },] }
];
DxRepresentativeInfo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxRepresentativeInfo = __decorate([
    ProxyCmp({
        inputs: ['name', 'type']
    })
], DxRepresentativeInfo);
export { DxRepresentativeInfo };
export class DxSidebarContent {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxSidebarContent.decorators = [
    { type: Component, args: [{
                selector: 'dx-sidebar-content',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxSidebarContent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxSiteNavigation = class DxSiteNavigation {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSiteNavigation.decorators = [
    { type: Component, args: [{
                selector: 'dx-site-navigation',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['expandedlayoutat']
            },] }
];
DxSiteNavigation.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSiteNavigation = __decorate([
    ProxyCmp({
        inputs: ['expandedlayoutat']
    })
], DxSiteNavigation);
export { DxSiteNavigation };
let DxSiteNavigationMenu = class DxSiteNavigationMenu {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['navigate']);
    }
};
DxSiteNavigationMenu.decorators = [
    { type: Component, args: [{
                selector: 'dx-site-navigation-menu',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['autonavigation', 'selectedvalue', 'titlestring'],
                outputs: ['navigate']
            },] }
];
DxSiteNavigationMenu.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSiteNavigationMenu = __decorate([
    ProxyCmp({
        inputs: ['autonavigation', 'selectedvalue', 'titlestring']
    })
], DxSiteNavigationMenu);
export { DxSiteNavigationMenu };
let DxSiteNavigationMenuItem = class DxSiteNavigationMenuItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSiteNavigationMenuItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-site-navigation-menu-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'state', 'value']
            },] }
];
DxSiteNavigationMenuItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSiteNavigationMenuItem = __decorate([
    ProxyCmp({
        inputs: ['label', 'state', 'value']
    })
], DxSiteNavigationMenuItem);
export { DxSiteNavigationMenuItem };
let DxSiteNavigationScope = class DxSiteNavigationScope {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSiteNavigationScope.decorators = [
    { type: Component, args: [{
                selector: 'dx-site-navigation-scope',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['value']
            },] }
];
DxSiteNavigationScope.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSiteNavigationScope = __decorate([
    ProxyCmp({
        inputs: ['value']
    })
], DxSiteNavigationScope);
export { DxSiteNavigationScope };
let DxSlideShow = class DxSlideShow {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['pageChange']);
    }
};
DxSlideShow.decorators = [
    { type: Component, args: [{
                selector: 'dx-slide-show',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['initialpage', 'mq1amount', 'mq2amount', 'mq3amount', 'mq4amount', 'mq5amount', 'spacebetween', 'viewmode'],
                outputs: ['pageChange']
            },] }
];
DxSlideShow.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSlideShow = __decorate([
    ProxyCmp({
        inputs: ['initialpage', 'mq1amount', 'mq2amount', 'mq3amount', 'mq4amount', 'mq5amount', 'spacebetween', 'viewmode'],
        methods: ['setPage']
    })
], DxSlideShow);
export { DxSlideShow };
let DxSlider = class DxSlider {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DxSlider.decorators = [
    { type: Component, args: [{
                selector: 'dx-slider',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'formattingpattern', 'label', 'max', 'min', 'readonly', 'showminmaxlabel', 'showvaluelabel', 'step', 'value'],
                outputs: ['valueChange']
            },] }
];
DxSlider.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSlider = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'formattingpattern', 'label', 'max', 'min', 'readonly', 'showminmaxlabel', 'showvaluelabel', 'step', 'value']
    })
], DxSlider);
export { DxSlider };
let DxSort = class DxSort {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['sort']);
    }
};
DxSort.decorators = [
    { type: Component, args: [{
                selector: 'dx-sort',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['sortable'],
                outputs: ['sort']
            },] }
];
DxSort.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSort = __decorate([
    ProxyCmp({
        inputs: ['sortable']
    })
], DxSort);
export { DxSort };
let DxSortItem = class DxSortItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSortItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-sort-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['value']
            },] }
];
DxSortItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSortItem = __decorate([
    ProxyCmp({
        inputs: ['value']
    })
], DxSortItem);
export { DxSortItem };
let DxSpacer = class DxSpacer {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSpacer.decorators = [
    { type: Component, args: [{
                selector: 'dx-spacer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5']
            },] }
];
DxSpacer.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSpacer = __decorate([
    ProxyCmp({
        inputs: ['mq1', 'mq2', 'mq3', 'mq4', 'mq5']
    })
], DxSpacer);
export { DxSpacer };
let DxSpinner = class DxSpinner {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxSpinner.decorators = [
    { type: Component, args: [{
                selector: 'dx-spinner',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['size', 'theme']
            },] }
];
DxSpinner.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxSpinner = __decorate([
    ProxyCmp({
        inputs: ['size', 'theme']
    })
], DxSpinner);
export { DxSpinner };
let DxTabBar = class DxTabBar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['tabChange']);
    }
};
DxTabBar.decorators = [
    { type: Component, args: [{
                selector: 'dx-tab-bar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['selected'],
                outputs: ['tabChange']
            },] }
];
DxTabBar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTabBar = __decorate([
    ProxyCmp({
        inputs: ['selected']
    })
], DxTabBar);
export { DxTabBar };
let DxTabPage = class DxTabPage {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxTabPage.decorators = [
    { type: Component, args: [{
                selector: 'dx-tab-page',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['label', 'value']
            },] }
];
DxTabPage.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTabPage = __decorate([
    ProxyCmp({
        inputs: ['label', 'value'],
        methods: ['toTabPage']
    })
], DxTabPage);
export { DxTabPage };
let DxTable = class DxTable {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['tableStateChanged', 'selectionStateChanged', 'expansionChanged', 'rowsSelected']);
    }
};
DxTable.decorators = [
    { type: Component, args: [{
                selector: 'dx-table',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['initialsort', 'initialsortdirection', 'mode', 'pagesize', 'pagesizes', 'rowcount', 'scrollable', 'selectionenabled', 'selectionmode'],
                outputs: ['tableStateChanged', 'selectionStateChanged', 'expansionChanged', 'rowsSelected']
            },] }
];
DxTable.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTable = __decorate([
    ProxyCmp({
        inputs: ['initialsort', 'initialsortdirection', 'mode', 'pagesize', 'pagesizes', 'rowcount', 'scrollable', 'selectionenabled', 'selectionmode']
    })
], DxTable);
export { DxTable };
let DxTableCell = class DxTableCell {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxTableCell.decorators = [
    { type: Component, args: [{
                selector: 'dx-table-cell',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'color', 'type', 'value']
            },] }
];
DxTableCell.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTableCell = __decorate([
    ProxyCmp({
        inputs: ['align', 'color', 'type', 'value']
    })
], DxTableCell);
export { DxTableCell };
export class DxTableFooter {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxTableFooter.decorators = [
    { type: Component, args: [{
                selector: 'dx-table-footer',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxTableFooter.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
export class DxTableHeader {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
}
DxTableHeader.decorators = [
    { type: Component, args: [{
                selector: 'dx-table-header',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>'
            },] }
];
DxTableHeader.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
let DxTableHeaderItem = class DxTableHeaderItem {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxTableHeaderItem.decorators = [
    { type: Component, args: [{
                selector: 'dx-table-header-item',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['align', 'columnid', 'customSort', 'label', 'linesize', 'maxwidth', 'minwidth', 'sortable', 'sortpriority', 'sticky', 'type']
            },] }
];
DxTableHeaderItem.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTableHeaderItem = __decorate([
    ProxyCmp({
        inputs: ['align', 'columnid', 'customSort', 'label', 'linesize', 'maxwidth', 'minwidth', 'sortable', 'sortpriority', 'sticky', 'type'],
        methods: ['getMinWidth', 'getMaxWidth']
    })
], DxTableHeaderItem);
export { DxTableHeaderItem };
let DxTableRow = class DxTableRow {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['rowSelected']);
    }
};
DxTableRow.decorators = [
    { type: Component, args: [{
                selector: 'dx-table-row',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['rowid', 'selected'],
                outputs: ['rowSelected']
            },] }
];
DxTableRow.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTableRow = __decorate([
    ProxyCmp({
        inputs: ['rowid', 'selected']
    })
], DxTableRow);
export { DxTableRow };
let DxTag = class DxTag {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['remove', 'selectChange']);
    }
};
DxTag.decorators = [
    { type: Component, args: [{
                selector: 'dx-tag',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'label', 'removeable', 'selectable', 'selected', 'textcolor', 'theme', 'type'],
                outputs: ['remove', 'selectChange']
            },] }
];
DxTag.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTag = __decorate([
    ProxyCmp({
        inputs: ['color', 'label', 'removeable', 'selectable', 'selected', 'textcolor', 'theme', 'type']
    })
], DxTag);
export { DxTag };
let DxText = class DxText {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxText.decorators = [
    { type: Component, args: [{
                selector: 'dx-text',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['color', 'type']
            },] }
];
DxText.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxText = __decorate([
    ProxyCmp({
        inputs: ['color', 'type']
    })
], DxText);
export { DxText };
let DxTextArea = class DxTextArea {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete']);
    }
};
DxTextArea.decorators = [
    { type: Component, args: [{
                selector: 'dx-text-area',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'lines', 'maxlength', 'placeholder', 'readonly', 'required', 'value'],
                outputs: ['valueChange', 'valueComplete']
            },] }
];
DxTextArea.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTextArea = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'kisynced', 'label', 'lines', 'maxlength', 'placeholder', 'readonly', 'required', 'value'],
        methods: ['focusControl']
    })
], DxTextArea);
export { DxTextArea };
let DxTextInput = class DxTextInput {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange', 'valueComplete', 'autocomplete', 'selectAutocomplete', 'enterPress']);
    }
};
DxTextInput.decorators = [
    { type: Component, args: [{
                selector: 'dx-text-input',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'icon', 'kisynced', 'label', 'limitlength', 'placeholder', 'readonly', 'required', 'size', 'value'],
                outputs: ['valueChange', 'valueComplete', 'autocomplete', 'selectAutocomplete', 'enterPress']
            },] }
];
DxTextInput.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTextInput = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'icon', 'kisynced', 'label', 'limitlength', 'placeholder', 'readonly', 'required', 'size', 'value'],
        methods: ['focusControl']
    })
], DxTextInput);
export { DxTextInput };
let DxTextList = class DxTextList {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxTextList.decorators = [
    { type: Component, args: [{
                selector: 'dx-text-list',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['type']
            },] }
];
DxTextList.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxTextList = __decorate([
    ProxyCmp({
        inputs: ['type']
    })
], DxTextList);
export { DxTextList };
let DxToggle = class DxToggle {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['checkedChange']);
    }
};
DxToggle.decorators = [
    { type: Component, args: [{
                selector: 'dx-toggle',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['checked', 'disabled'],
                outputs: ['checkedChange']
            },] }
];
DxToggle.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxToggle = __decorate([
    ProxyCmp({
        inputs: ['checked', 'disabled']
    })
], DxToggle);
export { DxToggle };
let DxUserAvatar = class DxUserAvatar {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
    }
};
DxUserAvatar.decorators = [
    { type: Component, args: [{
                selector: 'dx-user-avatar',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['imageurl', 'initials', 'size']
            },] }
];
DxUserAvatar.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxUserAvatar = __decorate([
    ProxyCmp({
        inputs: ['imageurl', 'initials', 'size']
    })
], DxUserAvatar);
export { DxUserAvatar };
let DxWizard = class DxWizard {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['currentStepChange']);
    }
};
DxWizard.decorators = [
    { type: Component, args: [{
                selector: 'dx-wizard',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['currentstep', 'expandedlayoutat', 'headline', 'orientation'],
                outputs: ['currentStepChange']
            },] }
];
DxWizard.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxWizard = __decorate([
    ProxyCmp({
        inputs: ['currentstep', 'expandedlayoutat', 'headline', 'orientation']
    })
], DxWizard);
export { DxWizard };
let DxWizardStep = class DxWizardStep {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['interaction']);
    }
};
DxWizardStep.decorators = [
    { type: Component, args: [{
                selector: 'dx-wizard-step',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['finished', 'label', 'value'],
                outputs: ['interaction']
            },] }
];
DxWizardStep.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxWizardStep = __decorate([
    ProxyCmp({
        inputs: ['finished', 'label', 'value']
    })
], DxWizardStep);
export { DxWizardStep };
let DxYesNo = class DxYesNo {
    constructor(c, r, z) {
        this.z = z;
        c.detach();
        this.el = r.nativeElement;
        proxyOutputs(this, this.el, ['valueChange']);
    }
};
DxYesNo.decorators = [
    { type: Component, args: [{
                selector: 'dx-yes-no',
                changeDetection: ChangeDetectionStrategy.OnPush,
                template: '<ng-content></ng-content>',
                inputs: ['disabled', 'errormessage', 'size', 'value'],
                outputs: ['valueChange']
            },] }
];
DxYesNo.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: NgZone }
];
DxYesNo = __decorate([
    ProxyCmp({
        inputs: ['disabled', 'errormessage', 'size', 'value']
    })
], DxYesNo);
export { DxYesNo };
