// @dvag/design-system/dist/collection/new/site-navigation/dx-site-navigation/dx-site-navigation.js
import { h, Host } from "@stencil/core";
/**
 * Adds a navigation menu to a content area.
 * It uses a responsive layout which transitions between the compact layout from MQ1 to MQ3 or MQ4 (configurable) and an
 * expanded layout which starts at MQ3 or MQ4 respectively.
 *
 * @name Site Navigation
 * @path /Scaffold
 * @icon compass
 * @stable
 */
export class DxSiteNavigation {
  constructor() {
    /**
     * Defines from which breakpoint the expanded large screen layout of the component is displayed.
     *
     * @value mq3: The expanded layout starts at mq3 or bigger.
     * @value mq4: The expanded layout starts at mq4 or bigger.
     *
     * @example expanded-layout-at
     */
    this.expandedlayoutat = "mq4";
  }
  render() {
    return (h(Host, { ref: (ref) => {
        const menu = ref.querySelector("dx-site-navigation-menu");
        menu.expandedlayoutat = this.expandedlayoutat;
      } }, h("div", { class: {
        layout: true,
        "expanded-at-mq4": this.expandedlayoutat === "mq4",
        "expanded-at-mq3": this.expandedlayoutat === "mq3",
      } }, h("div", { class: "menu" }, h("slot", { name: "menu" })), h("div", { class: "content" }, h("slot", null)))));
  }
  static get is() { return "dx-site-navigation"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-site-navigation.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-site-navigation.css"]
    };
  }
  static get properties() {
    return {
      "expandedlayoutat": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "SiteNavigationExpansionBreakpoint",
          "resolved": "\"mq3\" | \"mq4\"",
          "references": {
            "SiteNavigationExpansionBreakpoint": {
              "location": "local"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "mq3: The expanded layout starts at mq3 or bigger."
            }, {
              "name": "value",
              "text": "mq4: The expanded layout starts at mq4 or bigger."
            }, {
              "name": "example",
              "text": "expanded-layout-at"
            }],
          "text": "Defines from which breakpoint the expanded large screen layout of the component is displayed."
        },
        "attribute": "expandedlayoutat",
        "reflect": false,
        "defaultValue": "\"mq4\""
      }
    };
  }
}
