// @dvag/design-system/dist/collection/classes/render-util.js
export function onNextFrame(fn) {
  requestAnimationFrame(() => requestAnimationFrame(() => fn()));
}
export function getSlottedElements(element) {
  return Array.from(element.querySelectorAll("[slot]"));
}
export function getDefinedSlots(element) {
  return getSlottedElements(element).map((element) => element.slot);
}
export function isSlotDefined(element, slotName) {
  return !!element.querySelector(`[slot="${slotName}"]`);
}
export function isSlotUndefined(element, slotName) {
  return !isSlotDefined(element, slotName);
}
// Color calculations
export function toHex(dec) {
  return Math.round(dec).toString(16);
}
export function lighten(hex, factor) {
  const lightenColor = lightenRgb(toRgb(hex), factor);
  return rgbToHex(lightenColor);
}
function lightenRgb(rgb, factor) {
  return rgb.map((value) => 255 - factor * (255 - value));
}
function rgbToHex(rgb) {
  return "#" + rgb.map((val) => toHex(val)).join("");
}
function toRgb(hex) {
  const hexValues = hex.replace("#", "").match(/.{1,2}/g);
  return hexValues.map((hexValue) => parseInt(hexValue, 16));
}
