// @dvag/design-system/dist/collection/new/form/dx-radio-button-group/dx-radio-button/dx-radio-button.js
import { forceUpdate, h, Host, } from "@stencil/core";
import { ElementIdGenerator, verifyElementId, } from "../../../../classes/element-id-handling";
import { assert } from "../../../../classes/assert";
import { findUpperElementByPredicate, tagOf, } from "../../../../classes/misc-util";
/**
 * A radio button representing an option inside a `<dx-radio-button-group>` element.
 *
 * @path /Form/Radio
 * @name Radio-Button
 * @shortname Button
 * @icon check-circle
 *
 * @stable
 */
export class DxRadioButton {
  constructor() {
    /**
     * Disables the ability to interact with this component.
     *
     * @value false: User can interact with this component.
     * @value true: User cannot interact with this component.
     *
     * @example modes
     */
    this.disabled = false;
    /**
     * The group id of the parent radio button group. If not defined the first radio-button-group which is found will be elected as the owner of this radio-button
     * @internal
     */
    this.groupid = null;
  }
  /**
   * @internal
   */
  async internalSetChecked(checked) {
    this.inputElement.checked = checked;
    this.internalCheckedChange.emit(checked);
  }
  /**
   * @internal
   */
  async internalSelect() {
    this.select();
  }
  /**
   * Sets the focus on this component.
   * @example focus
   */
  async focusControl() {
    this.inputElement.focus();
  }
  /**
   * @internal
   */
  async forceUpdate() {
    forceUpdate(this);
  }
  connectedCallback() {
    var _a;
    (_a = this.group()) === null || _a === void 0 ? void 0 : _a.connectRadioButton(this.self);
  }
  disconnectedCallback() {
    var _a;
    (_a = this.group()) === null || _a === void 0 ? void 0 : _a.disconnectRadioButton(this.self);
  }
  componentWillLoad() {
    verifyElementId(this.self);
  }
  containsLabel() {
    var _a;
    return ((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0;
  }
  select() {
    if (!this.disabled) {
      const radioButtonGroup = this.group();
      radioButtonGroup.select(this.value).then();
    }
  }
  group() {
    const radioButtonGroup = findUpperElementByPredicate(this.self, (elem) => {
      const isRadioGroup = tagOf(elem) === "dx-radio-button-group";
      if (isRadioGroup) {
        return (elem.radiogroupid ===
          this.groupid ||
          (elem.radiogroupid == null &&
            this.groupid == null));
      }
      return false;
    });
    assert(radioButtonGroup != null, () => "<dx-radio-button> has to be within a <dx-radio-button-group>.", this.self);
    return radioButtonGroup;
  }
  render() {
    const radioButtonGroup = this.group();
    const groupId = radioButtonGroup.getAttribute("radiogroupid");
    const disabled = this.disabled || radioButtonGroup.disabled;
    return (h(Host, { class: {
        disabled: disabled,
      } }, h("div", null, h("input", { id: ElementIdGenerator.createId(this.self, "control"), ref: (el) => (this.inputElement = el), name: groupId ? `form-dx-radio-button-group-${groupId}` : undefined, type: "radio", onChange: () => this.select(), disabled: disabled, class: {
        "native-input": true,
        "event-delegate-receiver": true,
      }, tabIndex: disabled ? -1 : 1 }), h("div", { class: {
        "radio-label-container": true,
        "contains-label": this.containsLabel(),
      } }, h("dx-text", null, this.label), h("div", { class: "baseline-container" }, h("dx-text", { class: "baseline-text" }, "\u00A0"), h("div", { class: {
        "dot-container": true,
      } }))))));
  }
  static get is() { return "dx-radio-button"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-radio-button.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-radio-button.css"]
    };
  }
  static get properties() {
    return {
      "value": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The identifier of this option."
            }, {
              "name": "example",
              "text": "value"
            }],
          "text": "Identifier of this option. When selected this value overwrites the `<dx-radio-button-group>`'s value."
        },
        "attribute": "value",
        "reflect": true
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: User can interact with this component."
            }, {
              "name": "value",
              "text": "true: User cannot interact with this component."
            }, {
              "name": "example",
              "text": "modes"
            }],
          "text": "Disables the ability to interact with this component."
        },
        "attribute": "disabled",
        "reflect": true,
        "defaultValue": "false"
      },
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "important",
              "text": undefined
            }, {
              "name": "value",
              "text": "string: Label text."
            }, {
              "name": "example",
              "text": "label"
            }],
          "text": "Text, which is displayed next to the radio button."
        },
        "attribute": "label",
        "reflect": false
      },
      "groupid": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": undefined
            }],
          "text": "The group id of the parent radio button group. If not defined the first radio-button-group which is found will be elected as the owner of this radio-button"
        },
        "attribute": "groupid",
        "reflect": false,
        "defaultValue": "null"
      }
    };
  }
  static get events() {
    return [{
        "method": "internalCheckedChange",
        "name": "internalCheckedChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": undefined
            }],
          "text": ""
        },
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        }
      }];
  }
  static get methods() {
    return {
      "internalSetChecked": {
        "complexType": {
          "signature": "(checked: boolean) => Promise<void>",
          "parameters": [{
              "tags": [],
              "text": ""
            }],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      },
      "internalSelect": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      },
      "focusControl": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Sets the focus on this component.",
          "tags": [{
              "name": "example",
              "text": "focus"
            }]
        }
      },
      "forceUpdate": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      }
    };
  }
  static get elementRef() { return "self"; }
}
