// @dvag/design-system/dist/collection/components/card/ds-card-header/ds-card-header.js
import { h, Host } from "@stencil/core";
import { SlotObserver } from "../../../classes/slot-observer";
/**
 * Header element for cards.
 *
 * @path /Card
 * @name Card Header
 * @shortname Header
 * @icon heading
 * @stable
 *
 * @slot - Elements within the header. If not expicilty defined all text elements are rendered bold within this header.
 * The slot automatically aligns the content by defining a space between all elements.
 *
 * @slot-example simple
 * @slot-example header-with-text-button
 */
export class DsCardHeader {
  connectedCallback() {
    this.slotObserver = new SlotObserver(this.self, "ds-function-icon-button", async (nodes) => {
      nodes.forEach((node) => {
        node.style.marginRight = "-6px";
      });
    });
  }
  disconnectedCallback() {
    var _a;
    (_a = this.slotObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
  }
  render() {
    return (h(Host, null, h("div", { class: "header-content" }, h("slot", null)), h("ds-hr", null)));
  }
  static get is() { return "ds-card-header"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-card-header.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-card-header.css"]
    };
  }
  static get elementRef() { return "self"; }
}
