// @dvag/design-system/dist/collection/components/fachmodule/ergebnisuebersicht/ds-fm-ergebnisuebersicht-item/ds-fm-ergebnisuebersicht-item.js
/**
 * @path /Specialized/Erg.-Übersicht
 * @name Ergebnisübersicht-Eintrag
 * @shortname Eintrag
 * @icon database
 *
 */
export class DsFmErgebnisuebersichtItem {
  /**
   * @internal
   * Erzeugt aus diesem Item ein JS Objekt
   */
  async toErgebnisUebersichtItem() {
    return {
      label: this.label,
      sublabel: this.sublabel,
      firstcolumnvalue: this.firstcolumnvalue,
      secondcolumnvalue: this.secondcolumnvalue,
      firstcolumninfo: this.firstcolumninfo,
      secondcolumninfo: this.secondcolumninfo,
    };
  }
  static get is() { return "ds-fm-ergebnisuebersicht-item"; }
  static get encapsulation() { return "shadow"; }
  static get properties() {
    return {
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: e.g: Hausrat"
            }],
          "text": "Displays the label within the chart and the table."
        },
        "attribute": "label",
        "reflect": true
      },
      "sublabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: e.g Anwartschaft."
            }],
          "text": "Additional information below the label. Not visible within the chart."
        },
        "attribute": "sublabel",
        "reflect": true
      },
      "firstcolumnvalue": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: an amount."
            }],
          "text": "The value visible within the first column."
        },
        "attribute": "firstcolumnvalue",
        "reflect": true
      },
      "secondcolumnvalue": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: an amount."
            }],
          "text": "The value shown in the second column.\nThis value is optional. If no value is entered \"unbekannt\" will be displayed."
        },
        "attribute": "secondcolumnvalue",
        "reflect": true
      },
      "firstcolumninfo": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: A short text describing this column."
            }],
          "text": "Information text describing the first column.\nIf set an info-button will be displayed next to the column label."
        },
        "attribute": "firstcolumninfo",
        "reflect": true
      },
      "secondcolumninfo": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string:  A short text describing this column."
            }],
          "text": "Information text describing the second column.\nIf set an info-button will be displayed next to the column label."
        },
        "attribute": "secondcolumninfo",
        "reflect": true
      }
    };
  }
  static get methods() {
    return {
      "toErgebnisUebersichtItem": {
        "complexType": {
          "signature": "() => Promise<ErgebnisUebersichtItem>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "ErgebnisUebersichtItem": {
              "location": "local"
            }
          },
          "return": "Promise<ErgebnisUebersichtItem>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": "Erzeugt aus diesem Item ein JS Objekt"
            }]
        }
      }
    };
  }
}
