// @dvag/design-system/example-webapps/angular-app/src/app/app.component.ts
import { Component, OnInit } from '@angular/core';
import {
  FormControl,
  UntypedFormControl,
  UntypedFormGroup,
} from '@angular/forms';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent implements OnInit {
  formGroupDs = new UntypedFormGroup({
    text: new UntypedFormControl('Text'),
    number: new UntypedFormControl(1),
    date: new UntypedFormControl('2020-01-01'),
    area: new UntypedFormControl(
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Cupiditate, distinctio doloremque enim est fugit laudantium minima molestiae natus quia, quo recusandae reprehenderit unde voluptate? Consectetur id repudiandae sunt tempore ut.'
    ),
    dropdown: new UntypedFormControl('4'),
    radio: new UntypedFormControl('2'),
    search: new UntypedFormControl('Search'),
    grouped: new UntypedFormControl('Hiho'),
    yesno: new UntypedFormControl('no'),
    slider: new UntypedFormControl(25),
  });

  formGroupDx = new UntypedFormGroup({
    text: new UntypedFormControl('Text'),
    number: new UntypedFormControl(1),
    date: new UntypedFormControl('2020-01-01'),
    area: new UntypedFormControl(
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Cupiditate, distinctio doloremque enim est fugit laudantium minima molestiae natus quia, quo recusandae reprehenderit unde voluptate? Consectetur id repudiandae sunt tempore ut.'
    ),
    dropdown: new UntypedFormControl('4'),
    radio: new UntypedFormControl('2'),
    yesno: new UntypedFormControl('no'),
    toggle: new UntypedFormControl(false),
    slider: new UntypedFormControl(25),
  });

  money = 50;
  text = 'Ein Text';
  text2 = 'yes';
  active = false;
  textWerte = ['1', '2', '3', '4'];
  checked = 'unset';
  rows: Row[] = [
    createRow('2020-01-05', 'Fernseher', 899.99),
    createRow('2020-01-05', 'Auto', 27000),
    createRow('2020-01-05', 'Kaugummi', 0.4),
    createRow('2020-07-02', 'Kopfhörer', 39.95),
    createRow('2020-12-12', 'Weihnachtsbaum', 70),
    createRow('2020-12-13', 'Weihnachtsbaum', 70),
    createRow('2020-12-14', 'Weihnachtsbaum', 70),
  ];

  products = [
    'Fernseher',
    'Auto',
    'Weihnachtsbaum',
    'Kopfhörer',
    'Computer',
    'Spielzeug',
  ];

  ngOnInit(): void {}

  leadingZero(num: number): string {
    return ('0' + num.toString()).slice(-2);
  }

  addRow(): void {
    const startDate = new Date(2020, 1, 1);
    const randomDaysInMs = 24 * 60 * 60 * 1000 * Math.random() * 365;
    const randomDate = new Date(startDate.getTime() + randomDaysInMs);
    const randomProductsIndex = Math.floor(
      Math.random() * (this.products.length - 1)
    );
    this.rows.push(
      createRow(
        randomDate.getFullYear() +
          '-' +
          this.leadingZero(randomDate.getMonth() + 1) +
          '-' +
          this.leadingZero(randomDate.getDate()),
        this.products[randomProductsIndex],
        Math.random() * 100000
      )
    );
  }

  change(delta: number): void {
    this.money = Math.max(0, Math.min(this.money + delta, 100));
  }

  uncheck(): void {
    this.checked = 'unset';
  }

  onSubmitDs(): void {
    console.log(this.formGroupDs.value);
  }

  onSubmitDx(): void {
    console.log(this.formGroupDx.value);
  }
}

function createRow(date: string, item: string, price: number): Row {
  return {
    date,
    item,
    price,
  };
}

interface Row {
  date: string;
  item: string;
  price: number;
}
