// @dvag/design-system/dist/collection/new/internal/dx-chevron/dx-chevron.js
import { h } from "@stencil/core";
/**
 * @internal
 */
export class DxChevron {
  constructor() {
    /**
     * Defines the direction the arrow is pointing to.
     * @value true: the arrow is pointing up.
     * @value false: the arrow is pointing down.
     */
    this.directiontop = false;
    this.size = 24;
    /**
     * If `false` the chevron will be displayed with color `default`
     * If `true` the chevron will be displayed with color `active` or the color set via `activecolor`
     */
    this.active = false;
    /**
     * Color of the chevron when active property is `false`.
     * Use a color from the color palette.
     * @value string: The name of a color from the color palette.
     */
    this.inactivecolor = "default";
    /**
     * Color of the chevron when active property is `true`.
     * Use a color from the color palette.
     * @value string: The name of a color from the color palette.
     */
    this.activecolor = "active";
  }
  render() {
    return (h("dx-icon", { class: {
        rotated: this.directiontop,
        "size-16": this.size === 16,
        "size-24": this.size === 24,
      }, icon: "chevron-runter", color: this.active ? this.activecolor : this.inactivecolor, size: this.size }));
  }
  static get is() { return "dx-chevron"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-chevron.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-chevron.css"]
    };
  }
  static get properties() {
    return {
      "directiontop": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: the arrow is pointing up."
            }, {
              "name": "value",
              "text": "false: the arrow is pointing down."
            }],
          "text": "Defines the direction the arrow is pointing to."
        },
        "attribute": "directiontop",
        "reflect": false,
        "defaultValue": "false"
      },
      "size": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "16 | 24",
          "resolved": "16 | 24",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "size",
        "reflect": false,
        "defaultValue": "24"
      },
      "active": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "If `false` the chevron will be displayed with color `default`\nIf `true` the chevron will be displayed with color `active` or the color set via `activecolor`"
        },
        "attribute": "active",
        "reflect": false,
        "defaultValue": "false"
      },
      "inactivecolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The name of a color from the color palette."
            }],
          "text": "Color of the chevron when active property is `false`.\nUse a color from the color palette."
        },
        "attribute": "inactivecolor",
        "reflect": false,
        "defaultValue": "\"default\""
      },
      "activecolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The name of a color from the color palette."
            }],
          "text": "Color of the chevron when active property is `true`.\nUse a color from the color palette."
        },
        "attribute": "activecolor",
        "reflect": false,
        "defaultValue": "\"active\""
      }
    };
  }
}
