// @dvag/design-system/dist/collection/new/table/dx-table-cell/dx-table-cell.js
import { h, Host, } from "@stencil/core";
import { ColorPalette } from "../../../classes/color-palette";
import { displayNumberAsCurrency } from "../../../classes/currency-util";
/**
 * @path /Table
 * @name Table Cell
 * @shortname Cell
 * @icon border-all
 * @stable
 *
 */
export class DxTableCell {
  constructor() {
    this.touchDetected = false;
    this.tooltipText = "";
    this.showTooltip = false;
  }
  onShowTooltipChanged() {
    if (this.showTooltip) {
      this.updateTooltipText();
    }
  }
  /**
   * @internal
   */
  async obtainSettings(settings) {
    this.settings = settings;
  }
  updateTooltipText() {
    requestAnimationFrame(() => {
      var _a;
      const cell = this.self.shadowRoot.querySelector(".cell");
      this.tooltipText =
        ((_a = ((cell === null || cell === void 0 ? void 0 : cell.textContent) || this.self.textContent)) === null || _a === void 0 ? void 0 : _a.trim()) || this.value;
    });
  }
  getInnerElement() {
    var _a, _b, _c, _d;
    const currentType = (_a = this.type) !== null && _a !== void 0 ? _a : (_b = this.settings) === null || _b === void 0 ? void 0 : _b.type;
    if (currentType == null) {
      return h("div", null);
    }
    if (currentType === "html") {
      return (h("div", { class: "html-alignment-container" }, h("div", { class: "overflow-container" }, h("slot", null))));
    }
    if (currentType === "text") {
      return this.renderCellText(this.value);
    }
    if (currentType === "numeric") {
      const numericText = (_c = parseFloat(this.value)) === null || _c === void 0 ? void 0 : _c.toLocaleString("de-DE");
      return this.renderCellText(numericText);
    }
    if (currentType === "currency") {
      const currencyText = displayNumberAsCurrency(parseFloat(this.value), 2);
      return this.renderCellText(`${currencyText} €`);
    }
    if (currentType === "date") {
      const split = (_d = this.value) === null || _d === void 0 ? void 0 : _d.split("-");
      const text = split != null ? `${split[2]}.${split[1]}.${split[0]}` : "";
      return this.renderCellText(text);
    }
  }
  renderCellText(text) {
    return (h("div", { class: "overflow-container" }, h("dx-text", null, text)));
  }
  handleContextMenu(e) {
    e.preventDefault();
    this.showTooltip = true;
  }
  getStyle() {
    var _a, _b;
    const style = Object.assign({}, (_a = this.settings) === null || _a === void 0 ? void 0 : _a.sizeStyle);
    if (this.color != null) {
      style["background-color"] = new ColorPalette(true).getColorCode(this.color);
    }
    const lines = (_b = this.settings) === null || _b === void 0 ? void 0 : _b.lines;
    style["--dx-table-cell-lines"] = lines !== null && lines !== void 0 ? lines : 1;
    if (lines > 1) {
      style["padding-top"] = "8px";
      style["padding-bottom"] = "8px";
    }
    return style;
  }
  getAlign() {
    var _a;
    return this.align || ((_a = this.settings) === null || _a === void 0 ? void 0 : _a.align);
  }
  render() {
    var _a, _b;
    const alignClass = {};
    if (this.align || this.settings != null) {
      alignClass["align-" + this.getAlign()] = true;
    }
    return (h(Host, { role: "cell", style: this.getStyle(), slot: ((_a = this.settings) === null || _a === void 0 ? void 0 : _a.sticky) ? "sticky" : undefined }, h("div", { class: Object.assign({ cell: true, "multiple-lines": ((_b = this.settings) === null || _b === void 0 ? void 0 : _b.lines) > 1 }, alignClass), title: this.tooltipText, onTouchStart: () => (this.touchDetected = true), onContextMenu: (e) => {
        if (this.touchDetected) {
          this.handleContextMenu(e);
          this.touchDetected = false;
        }
      } }, this.getInnerElement()), this.renderPopup(alignClass)));
  }
  renderPopup(alignClass) {
    var _a;
    if (!this.showTooltip || !(((_a = this.tooltipText) === null || _a === void 0 ? void 0 : _a.length) > 0)) {
      return;
    }
    const alignment = this.getAlign();
    return (h("div", { class: Object.assign({ "popup-container": true }, alignClass) }, h("div", { class: "overlay", onTouchEnd: (e) => {
        e.stopPropagation();
        e.preventDefault();
        this.showTooltip = false;
      } }), h("dx-popup", { arrowposition: alignment !== "center" ? alignment !== null && alignment !== void 0 ? alignment : "left" : "left", show: true }, h("dx-text", null, this.tooltipText))));
  }
  static get is() { return "dx-table-cell"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-table-cell.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-table-cell.css"]
    };
  }
  static get properties() {
    return {
      "value": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The value which is formatted within the cell."
            }],
          "text": "The value of this cell. It will be formatted depending on the cell's `type`."
        },
        "attribute": "value",
        "reflect": true
      },
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "text: The `value` is a string which will be rendered as text.\nIt is left-aligned by default."
            }, {
              "name": "value",
              "text": "numeric: The `value` is a number which will be rendered as text.\nIt is right-aligned by default."
            }, {
              "name": "value",
              "text": "date:  The `value` is a string ('YYYY-mm-dd') which will be formatted as a date string.\nIt is left-aligned by default."
            }, {
              "name": "value",
              "text": "currency: The `value` is a number which will be formatted as a currency string (e.g: 1.120,00 \u20AC).\nThe currency itself is rendered as an icon.\nIt is right-aligned by default."
            }, {
              "name": "value",
              "text": "html: The default slot will be rendered in place of the `value`."
            }],
          "text": "The type of this cell.\nIf not explicity set the cell inherits the type by the corresponding `<dx-table-header-item>`."
        },
        "attribute": "type",
        "reflect": false
      },
      "align": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "ColumnAlignment",
          "resolved": "\"center\" | \"left\" | \"right\"",
          "references": {
            "ColumnAlignment": {
              "location": "import",
              "path": "../table-common-types"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The alignment of this cell.\nIf not explicity set the cell derives the alignment"
        },
        "attribute": "align",
        "reflect": false
      },
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: A color which has to be defined within the color palette of the design system. (e.g.: blue-100, error, etc.)"
            }],
          "text": "The background color."
        },
        "attribute": "color",
        "reflect": false
      }
    };
  }
  static get states() {
    return {
      "settings": {},
      "tooltipText": {},
      "showTooltip": {}
    };
  }
  static get methods() {
    return {
      "obtainSettings": {
        "complexType": {
          "signature": "(settings: ColumnSettings) => Promise<void>",
          "parameters": [{
              "tags": [],
              "text": ""
            }],
          "references": {
            "Promise": {
              "location": "global"
            },
            "ColumnSettings": {
              "location": "import",
              "path": "../table-common-types"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      }
    };
  }
  static get elementRef() { return "self"; }
  static get watchers() {
    return [{
        "propName": "showTooltip",
        "methodName": "onShowTooltipChanged"
      }];
  }
}
