// @dvag/design-system/dist/collection/components/card/ds-card/ds-card.js
import { h, Host } from "@stencil/core";
/**
 * Card base element.
 *
 * @path /
 * @name Card
 * @icon id-card
 * @stable
 *
 * @slot - Elements which are displayed within the card.
 */
export class DsCard {
  constructor() {
    /**
     * The component throws a shadow. It is used as a contrast to light backgrounds.
     *
     * @value false: Card without shadow
     * @value true: Card with shadow
     *
     * @example shadow
     */
    this.shadow = false;
  }
  render() {
    return (h(Host, { class: { shadow: this.shadow } }, h("slot", null)));
  }
  static get is() { return "ds-card"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-card.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-card.css"]
    };
  }
  static get properties() {
    return {
      "shadow": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Card without shadow"
            }, {
              "name": "value",
              "text": "true: Card with shadow"
            }, {
              "name": "example",
              "text": "shadow"
            }],
          "text": "The component throws a shadow. It is used as a contrast to light backgrounds."
        },
        "attribute": "shadow",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
