// @dvag/design-system/dist/collection/components/card/ds-card-image-header/ds-card-image-header.js
import { h, Host } from "@stencil/core";
import { assert } from "../../../classes/assert";
/**
 * An image which is displayed within a `<ds-card>` element.
 *
 * @path /Card
 * @name Card Image Header
 * @shortname Image Header
 * @icon image
 */
export class DsCardImageHeader {
  render() {
    assert(this.url != null, () => "<ds-card-image-header> Eigenschaft 'url' muss gesetzt sein.");
    return (h(Host, { style: {
        backgroundImage: `url(${this.url})`,
      } }));
  }
  static get is() { return "ds-card-image-header"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-card-image-header.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-card-image-header.css"]
    };
  }
  static get properties() {
    return {
      "url": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The path to the image file."
            }],
          "text": "Specifies a path to the image file."
        },
        "attribute": "url",
        "reflect": false
      }
    };
  }
}
