// @dvag/design-system/dist/collection/new/scaffold/dx-logo/dx-logo.js
import { h, Host, } from "@stencil/core";
/**
 * Displays the DVAG logo
 * @path /Scaffold
 * @name Logo
 * @icon sign
 * @stable
 */
export class DxLogo {
  constructor() {
    /**
     * Provides the logo in different styles.
     *
     * @value bildmarke: DVAG-Logo banner image.
     * @value wortmarke: DVAG-Logo with word mark.
     * @example logo
     */
    this.type = "bildmarke";
  }
  render() {
    // https://stackoverflow.com/questions/33471880/dynamic-tag-name-in-react-jsx
    const Logo = `dx-logo-${this.type}`;
    return (h(Host, null, h(Logo, null)));
  }
  static get is() { return "dx-logo"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-logo.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-logo.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"bildmarke\" | \"wortmarke\"",
          "resolved": "\"bildmarke\" | \"wortmarke\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "bildmarke: DVAG-Logo banner image."
            }, {
              "name": "value",
              "text": "wortmarke: DVAG-Logo with word mark."
            }, {
              "name": "example",
              "text": "logo"
            }],
          "text": "Provides the logo in different styles."
        },
        "attribute": "type",
        "reflect": false,
        "defaultValue": "\"bildmarke\""
      }
    };
  }
  static get elementRef() { return "self"; }
}
