// @dvag/design-system/dist/collection/new/scaffold/dx-overflow-menu/dx-overflow-menu-item/dx-overflow-menu-item.js
import { assert } from "../../../../classes/assert";
/**
 * A menu item within a `<dx-overflow-menu>` element which describes an action to perform within a specific application context.
 *
 * @path /Miscellaneous/Menu
 * @name Overflow Menu Item
 * @shortname Item
 * @icon list-ul
 * @stable
 */
export class DxOverflowMenuItem {
  /**
   * @internal
   * Extracts this element's properties into a complex javascript object.
   */
  async toOverflowMenuItem() {
    var _a;
    assert(typeof this.value !== "number", () => `<dx-overflow-menu-item> Property 'value' must not be of type 'number'.`);
    return {
      id: this.self.id,
      icon: this.icon,
      value: (_a = this.value) !== null && _a !== void 0 ? _a : this.label,
      label: this.label,
      selected: this.selected,
      setSelected: (selected) => this.handleSelection(selected),
    };
  }
  handleSelection(selected) {
    this.selected = selected;
  }
  static get is() { return "dx-overflow-menu-item"; }
  static get properties() {
    return {
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The icon name."
            }, {
              "name": "example",
              "text": "icon"
            }],
          "text": "An icon which is displayed next to the label text."
        },
        "attribute": "icon",
        "reflect": true
      },
      "value": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "StringOrObject",
          "resolved": "string | { [key: string]: any; }",
          "references": {
            "StringOrObject": {
              "location": "import",
              "path": "../../../layout/dx-list/list.util"
            }
          }
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Internal value as string. (e.g. `\"apple\"`)"
            }, {
              "name": "value",
              "text": "object: Complex type e.g. `{name: \"apple\", amount: 3}`"
            }, {
              "name": "example",
              "text": "label"
            }],
          "text": "Internal value of this option. Only `string` and `object` types are supported. The `number` type is explicitly not supported."
        },
        "attribute": "value",
        "reflect": true
      },
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Displayed text e.g. `Apple`"
            }, {
              "name": "example",
              "text": "label"
            }],
          "text": "The label which is displayed in the option list.\nIn case `value` is undefined, this `label` represents the internal value."
        },
        "attribute": "label",
        "reflect": true
      },
      "selected": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: This option is selected (Depending on the selection rule of the parent dropdown)"
            }, {
              "name": "value",
              "text": "false: This option is not selected."
            }],
          "text": "Whether this option is selected or not. In case the parent dropdown only can select one option at a time, the last\noption which selected attribute is true, will be selected."
        },
        "attribute": "selected",
        "reflect": true
      }
    };
  }
  static get methods() {
    return {
      "toOverflowMenuItem": {
        "complexType": {
          "signature": "() => Promise<OverflowMenuItem>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "OverflowMenuItem": {
              "location": "local"
            }
          },
          "return": "Promise<OverflowMenuItem>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": "Extracts this element's properties into a complex javascript object."
            }]
        }
      }
    };
  }
  static get elementRef() { return "self"; }
}
