// @dvag/design-system/dist/collection/components/scaffold/ds-header/ds-header-customer/ds-header-customer.js
/**
 * Konfiguriert die Kunden-Integration im Header.
 * @name Header Kundenkonfiguration
 * @shortname Kunde
 * @path /Scaffold/Header
 */
export class DsHeaderCustomer {
  /**
   * @internal
   */
  async toHeaderCustomerConfiguration() {
    return {
      name: this.name,
      typeLabel: this.typelabel
    };
  }
  static get is() { return "ds-header-customer"; }
  static get encapsulation() { return "shadow"; }
  static get properties() {
    return {
      "name": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: first name and last name"
            }],
          "text": "Customer name"
        },
        "attribute": "name",
        "reflect": false
      },
      "typelabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The type of the customer"
            }],
          "text": "The type of the customer e.g: Kunde"
        },
        "attribute": "typelabel",
        "reflect": false
      }
    };
  }
  static get methods() {
    return {
      "toHeaderCustomerConfiguration": {
        "complexType": {
          "signature": "() => Promise<HeaderCustomerConfiguration>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "HeaderCustomerConfiguration": {
              "location": "local"
            }
          },
          "return": "Promise<HeaderCustomerConfiguration>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      }
    };
  }
}
