// @dvag/design-system/dist/collection/new/scaffold/dx-header/dx-header-menu-link/dx-header-menu-link.js
import { h, } from "@stencil/core";
import { ElementIdGenerator, verifyElementId, } from "../../../../classes/element-id-handling";
/**
 * A clickable item in the header menu.
 * @name Header Menu link
 * @shortname Link
 * @path /Scaffold/Header/Menu
 * @icon link
 * @stable
 */
export class DxHeaderMenuLink {
  constructor() {
    /**
     * Highlights this entry.
     *
     * @value true: Entry highlighted as selected.
     * @value false: Entry not highlighted as selected.
     * @example selected
     */
    this.selected = false;
    /**
     * @internal
     */
    this.internalopen = true;
  }
  componentWillLoad() {
    verifyElementId(this.self);
  }
  /**
   * @internal
   */
  async getDefaultWidth() {
    return this.self.shadowRoot.querySelector("dx-header-menu-button")
      .offsetWidth;
  }
  render() {
    return (h("dx-horizontal-grow-box", { open: this.internalopen }, h("dx-header-menu-button", { id: ElementIdGenerator.createId(this.self, "button"), selected: this.selected, onInteraction: () => this.menuInteraction.emit(this.connectednavigationvalue) }, h("dx-text", null, this.label))));
  }
  static get is() { return "dx-header-menu-link"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-header-menu-link.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-header-menu-link.css"]
    };
  }
  static get properties() {
    return {
      "selected": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: Entry highlighted as selected."
            }, {
              "name": "value",
              "text": "false: Entry not highlighted as selected."
            }, {
              "name": "example",
              "text": "selected"
            }],
          "text": "Highlights this entry."
        },
        "attribute": "selected",
        "reflect": false,
        "defaultValue": "false"
      },
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The label which is displayed in the menu item."
            }, {
              "name": "example",
              "text": "label"
            }],
          "text": "The label which is displayed in the menu item."
        },
        "attribute": "label",
        "reflect": false
      },
      "connectednavigationvalue": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The name of the connected value which is displayed within the `<dx-header>` sidebar navigation when interacted with."
            }, {
              "name": "example",
              "text": "connected-value"
            }],
          "text": "The `<dx-header-navigation-item>` value which is connected with this menu item. It also acts as an identifier which is emitted with the `menuInteraction` event payload."
        },
        "attribute": "connectednavigationvalue",
        "reflect": false
      },
      "priority": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: The priority of the user menu item."
            }, {
              "name": "example",
              "text": "priority"
            }],
          "text": "The priority indicates the order in which the entries can be shown and hidden.\nThis property does not have any effect when the component is not within a `<dx-header>` element."
        },
        "attribute": "priority",
        "reflect": false
      },
      "internalopen": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": undefined
            }],
          "text": ""
        },
        "attribute": "internalopen",
        "reflect": false,
        "defaultValue": "true"
      }
    };
  }
  static get events() {
    return [{
        "method": "menuInteraction",
        "name": "menuInteraction",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "example",
              "text": "menu-interaction"
            }],
          "text": "Fired when the header menu item is interacted with."
        },
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        }
      }];
  }
  static get methods() {
    return {
      "getDefaultWidth": {
        "complexType": {
          "signature": "() => Promise<number>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<number>"
        },
        "docs": {
          "text": "",
          "tags": [{
              "name": "internal",
              "text": undefined
            }]
        }
      }
    };
  }
  static get elementRef() { return "self"; }
}
