// @dvag/design-system/dist/cjs/breakpoint-aware-cc9d0ee6.js
'use strict';

const BREAKPOINTS = [0, 768, 992, 1200];
const NEW_BREAKPOINTS = [0, 476, 760, 1024, 1280];
function getBreakpointIndex(width, breakpoints) {
  return breakpoints.findIndex((_, index) => {
    return isBreakpointIndex(width, index, breakpoints);
  });
}
function isBreakpointIndex(width, index, breakpoints) {
  if (index === breakpoints.length - 1) {
    return width >= breakpoints[index];
  }
  return width >= breakpoints[index] && width < breakpoints[index + 1];
}
class BreakpointAware {
  constructor(onSizeChange, useNewBreakpoints = false) {
    this.onSizeChange = onSizeChange;
    this.internalCallback = () => {
      const newBreakpoint = getBreakpointIndex(window.innerWidth, this.breakpoints);
      if (this.currentBreakpoint !== newBreakpoint) {
        this.currentBreakpoint = newBreakpoint;
        this.onSizeChange();
      }
    };
    window.addEventListener("resize", this.internalCallback);
    this.breakpoints = useNewBreakpoints ? NEW_BREAKPOINTS : BREAKPOINTS;
    this.currentBreakpoint = getBreakpointIndex(window.innerWidth, this.breakpoints);
  }
  getEffectiveValue(mediaQueryValues) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    if (this.mq1) {
      return mediaQueryValues.mq1;
    }
    if (this.mq2) {
      return (_a = mediaQueryValues.mq2) !== null && _a !== void 0 ? _a : mediaQueryValues.mq1;
    }
    if (this.mq3) {
      return ((_c = (_b = mediaQueryValues.mq3) !== null && _b !== void 0 ? _b : mediaQueryValues.mq2) !== null && _c !== void 0 ? _c : mediaQueryValues.mq1);
    }
    if (this.mq4) {
      return ((_f = (_e = (_d = mediaQueryValues.mq4) !== null && _d !== void 0 ? _d : mediaQueryValues.mq3) !== null && _e !== void 0 ? _e : mediaQueryValues.mq2) !== null && _f !== void 0 ? _f : mediaQueryValues.mq1);
    }
    if (this.mq5) {
      return ((_k = (_j = (_h = (_g = mediaQueryValues.mq5) !== null && _g !== void 0 ? _g : mediaQueryValues.mq4) !== null && _h !== void 0 ? _h : mediaQueryValues.mq3) !== null && _j !== void 0 ? _j : mediaQueryValues.mq2) !== null && _k !== void 0 ? _k : mediaQueryValues.mq1);
    }
    throw new Error("Illegal state! Cannot find breakpoint");
  }
  get xs() {
    return this.currentBreakpoint === 0;
  }
  get md() {
    return this.currentBreakpoint === 1;
  }
  get lg() {
    return this.currentBreakpoint === 2;
  }
  get xl() {
    return this.currentBreakpoint === 3;
  }
  // New Breakpoints
  get mq1() {
    return this.currentBreakpoint === 0;
  }
  get mq2() {
    return this.currentBreakpoint === 1;
  }
  get mq3() {
    return this.currentBreakpoint === 2;
  }
  get mq4() {
    return this.currentBreakpoint === 3;
  }
  get mq5() {
    return this.currentBreakpoint === 4;
  }
  disconnect() {
    window.removeEventListener("resize", this.internalCallback);
  }
}

exports.BreakpointAware = BreakpointAware;
