// @dvag/design-system/dist/cjs/color-palette-ed0f6557.js
'use strict';

const assert = require('./assert-d8ea75c6.js');

const ColorMapDX = {
  "#c8aa22": ["gold", "logo", "golden-stage"],
  "#00090c": ["blue-190"],
  "#001219": ["blue-180"],
  "#001a25": ["blue-170"],
  "#002332": ["blue-160"],
  "#002c3e": ["blue-150"],
  "#00354a": ["blue-140"],
  "#003e57": ["blue-130"],
  "#004763": ["blue-120", "active"],
  "#004f70": ["blue-110"],
  "#00587c": ["blue-100", "hover"],
  "#196989": ["blue-90"],
  "#337a96": ["blue-80", "default"],
  "#4c8aa3": ["blue-70"],
  "#669bb0": ["blue-60"],
  "#80acbe": ["blue-50"],
  "#99bdcb": ["blue-40"],
  "#b2cdd8": ["blue-30"],
  "#ccdee5": ["blue-20"],
  "#e5eef2": ["blue-10"],
  "#181f00": ["green-180"],
  "#313e00": ["green-160"],
  "#495c01": ["green-140"],
  "#617b01": ["green-120"],
  "#7a9a01": ["green-100", "success"],
  "#94ae34": ["green-80"],
  "#afc267": ["green-60"],
  "#cad799": ["green-40"],
  "#e4ebcc": ["green-20"],
  "#1f070b": ["red-180"],
  "#3f0e15": ["red-160"],
  "#5e141f": ["red-140"],
  "#7e1b2a": ["red-120"],
  "#9d2235": ["red-100", "error"],
  "#b14e5d": ["red-80"],
  "#c47a85": ["red-60"],
  "#d8a7ae": ["red-40"],
  "#ebd3d7": ["red-20"],
  "#2e1807": ["orange-180"],
  "#5d300e": ["orange-160"],
  "#8b4714": ["orange-140"],
  "#ba5f1b": ["orange-120"],
  "#e87722": ["orange-100", "warning"],
  "#ed924e": ["orange-80"],
  "#f1ad7a": ["orange-60"],
  "#f6c9a7": ["orange-40"],
  "#fae4d3": ["orange-20"],
  "#001715": ["turquoise-180"],
  "#002e29": ["turquoise-160"],
  "#00453e": ["turquoise-140"],
  "#005c53": ["turquoise-120"],
  "#007367": ["turquoise-100"],
  "#338f86": ["turquoise-80"],
  "#66aba4": ["turquoise-60"],
  "#99c7c2": ["turquoise-40"],
  "#cce3e1": ["turquoise-20"],
  "#13011c": ["purple-180"],
  "#250238": ["purple-160"],
  "#380454": ["purple-140"],
  "#4a0570": ["purple-120"],
  "#5c068c": ["purple-100"],
  "#7d38a3": ["purple-80"],
  "#9e6aba": ["purple-60"],
  "#be9bd1": ["purple-40"],
  "#dfcde8": ["purple-20"],
  "#000000": ["black"],
  "#2b2b2b": ["gray-83", "headline"],
  "#575757": ["gray-66", "paragraph"],
  "#808080": ["gray-50", "tags", "icons"],
  "#ababab": ["gray-33"],
  "#d6d6d6": ["gray-16", "border"],
  "#ededed": ["gray-7", "footer"],
  "#f7f7f7": ["gray-3", "background"],
  "#ffffff": ["white"],
  "transparent": ["transparent"],
};

class ColorPalette {
  constructor(useNewColors = false) {
    this.useNewColors = useNewColors;
  }
  getColorCode(colorName) {
    const lowerCaseColor = colorName.toLowerCase();
    return this.useNewColors
      ? this.getDXColorCode(lowerCaseColor)
      : this.getDSColorCode(lowerCaseColor);
  }
  getDXColorCode(colorName) {
    const hexCode = getDXColorCode(colorName);
    assert.assert(hexCode != null, () => this.getErrorMessage(colorName, getDXColorAliases()));
    return hexCode;
  }
  getDSColorCode(colorName) {
    const hexCode = getDSColorCode(colorName);
    assert.assert(hexCode != null, () => this.getErrorMessage(colorName, getDSColorAliases()));
    return hexCode;
  }
  getErrorMessage(colorName, aliases) {
    return `There is no color with the name '${colorName}'. Possible names: \n${aliases.join("\n")}`;
  }
  getWheelColor(index) {
    return this.useNewColors
      ? getDXWheelColorAlias(index)
      : getDSWheelColorAlias(index);
  }
  getWheelColorCode(index) {
    return this.getColorCode(this.getWheelColor(index));
  }
}
const ColorMapDS = {
  "#a7893d": ["gold"],
  "#957b36": ["gold-mittel"],
  "#856D30": ["gold-dunkel"],
  "#E3DAC4": ["gold-hell"],
  "#ffffff": ["weiss"],
  "#424242": ["anthrazit"],
  "#4c4c4c": ["dunkelgrau-1"],
  "#666666": ["dunkelgrau-2"],
  "#7f7f7f": ["dunkelgrau-3"],
  "#EBEBEB": ["hellgrau-1"],
  "#DEDEDE": ["hellgrau-2"],
  "#C1C1C1": ["hellgrau-3"],
  "#F5F5F5": ["hellgrau-4"],
  "#004165": ["dunkelblau-1"],
  "#336783": ["dunkelblau-2"],
  "#668DA2": ["dunkelblau-3"],
  "#99B3C1": ["dunkelblau-4"],
  "#CCD9E0": ["dunkelblau-5"],
  "#BE1413": ["rot-1"],
  "#CB4342": ["rot-2"],
  "#D87271": ["rot-3"],
  "#E5A1A0": ["rot-4"],
  "#F2D0CF": ["rot-5"],
  "#3B94A3": ["petrol-1"],
  "#62A9B5": ["petrol-2"],
  "#89BEC7": ["petrol-3"],
  "#B0D4DA": ["petrol-4"],
  "#D7E9EC": ["petrol-5"],
  "#E8B50E": ["gelb-1"],
  "#ECC33E": ["gelb-2"],
  "#F1D26E": ["gelb-3"],
  "#F5E19E": ["gelb-4"],
  "#FAF0CE": ["gelb-5"],
  "#6EA318": ["gruen-1"],
  "#8BB546": ["gruen-2"],
  "#A8C774": ["gruen-3"],
  "#C5DAA2": ["gruen-4"],
  "#E2ECD0": ["gruen-5"],
};
const WHEEL_COLORS_DS = [
  "gold",
  "dunkelblau-1",
  "petrol-1",
  "gruen-1",
  "gelb-1",
  "rot-1",
  "gold-hell",
  "dunkelblau-3",
  "petrol-3",
  "gruen-3",
  "gelb-3",
  "rot-3",
  "dunkelblau-4",
  "petrol-4",
  "gruen-4",
  "gelb-4",
  "rot-4",
];
const WHEEL_COLORS_DX = ["gold", "default", "error", "warning"];
function getDXColorAliases() {
  return Object.values(ColorMapDX)
    .flat()
    .map((alias) => alias);
}
function getDXColorCode(color) {
  return Object.keys(ColorMapDX).find((hexCode) => {
    return ColorMapDX[hexCode].includes(color);
  });
}
function getDXWheelColorAlias(index) {
  return WHEEL_COLORS_DX[index];
}
function getDSColorAliases() {
  return Object.values(ColorMapDS)
    .flat()
    .map((alias) => alias);
}
function getDSColorCode(color) {
  return Object.keys(ColorMapDS).find((hexCode) => {
    return ColorMapDS[hexCode].includes(color);
  });
}
function getDSWheelColorAlias(index) {
  return WHEEL_COLORS_DS[index];
}

exports.ColorPalette = ColorPalette;
