// @dvag/design-system/dist/cjs/ds-column.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const assert = require('./assert-d8ea75c6.js');

const dsColumnCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .column{display:flex;flex-direction:column}";

const DsColumn = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Gibt an, welche Ausrichtung die Elemente haben sollen.
     * @value string: left
     * @value string: center
     * @value string: right
     * @value string: stretch
     * @example example
     */
    this.align = "left";
  }
  render() {
    assert.assertPropertyValue("ds-column", "align", this.align, [
      "left",
      "center",
      "right",
      "stretch",
    ]);
    return (index.h(index.Host, null, index.h("div", { class: "column", style: { "align-items": this.getAlignment() } }, index.h("slot", null))));
  }
  getAlignment() {
    return alignments[this.align];
  }
};
const alignments = {
  center: "center",
  left: "flex-start",
  right: "flex-end",
  stretch: "stretch",
};
DsColumn.style = dsColumnCss;

exports.ds_column = DsColumn;
