// @dvag/design-system/dist/cjs/ds-card-collapsable.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const renderUtil = require('./render-util-6b278c0d.js');
const assert = require('./assert-d8ea75c6.js');

const dsCardCollapsableCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .header{cursor:pointer}:host .header .header-content{display:flex;align-items:center}:host .header .header-content ds-checkbox,:host .header .header-content ds-icon{margin-right:12px}";

const DsCardCollapsable = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Sets the theme of the `<ds-card-header>` element within this component.
     *
     * @value arrow: Renders an arrow in the header which visualizes the current collapse state.
     * @value checkbox: Render a checkbox in the header which is checked whether the card is collapsed or not.
     * @value custom: Enables the feature to define a custom card header in the `header` slot.
     *
     * @example arrow
     * @example checkbox
     * @example custom
     */
    this.type = "arrow";
    /**
     * The current collapse state.
     *
     * @value false: The card is collapsed.
     * @value true: The card is unfolded.
     *
     * @example open
     * @example close
     */
    this.open = false;
  }
  toggleOpen() {
    this.open = !this.open;
  }
  componentWillLoad() {
    this.showBorder = this.open;
  }
  handleHeaderClick() {
    if (this.type !== "custom") {
      this.toggleOpen();
    }
  }
  render() {
    return (index.h(index.Host, { style: {
        "--ds-card-header-separator-opacity": `${this.showBorder ? 1 : 0}`,
      } }, [this.renderHeader(), this.renderContent()]));
  }
  renderToggleButton() {
    if (this.type === "checkbox") {
      return index.h("ds-checkbox", { checked: this.open, class: "no-id-warning" });
    }
    else if (this.type === "arrow") {
      return (index.h("ds-icon", { size: 24, icon: this.open ? "collapse-close" : "collapse-open" }));
    }
  }
  renderHeaderContent() {
    return (index.h("div", { class: {
        "header-content": true,
      } }, this.renderToggleButton(), index.h("ds-text", { type: "u5", screensize: "small" }, this.label)));
  }
  renderHeader() {
    assert.assert(renderUtil.isSlotDefined(this.self, "header")
      ? this.type === "custom"
      : this.type !== "custom", () => "Header-Slot should be defined when type === 'custom'");
    return this.type === "custom" ? (index.h("slot", { name: "header" })) : (index.h("ds-card-header", { class: {
        header: true,
      }, onClick: () => this.handleHeaderClick() }, this.renderHeaderContent()));
  }
  renderContent() {
    return (index.h("dx-simple-grow-box", { slidecontent: true, class: {
        content: true,
      }, animationmode: "top-to-bottom", open: this.open, onGrowBoxStateChange: (event) => {
        this.showBorder = event.detail !== "closed";
      } }, index.h("slot", null)));
  }
  get self() { return index.getElement(this); }
};
DsCardCollapsable.style = dsCardCollapsableCss;

exports.ds_card_collapsable = DsCardCollapsable;
