// @dvag/design-system/dist/cjs/ds-fullscreen-spinner.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const modalScrollHelper = require('./modal-scroll-helper-97bdd5aa.js');
require('./misc-util-63b423bd.js');
require('./value-store-f683b26f.js');

const dsFullscreenSpinnerCss = ":host{display:flex;justify-content:center;align-items:center;position:fixed;top:0;left:0;width:100%;height:100%;background-color:rgba(102, 102, 102, 0.6);z-index:1000}:host:host([hidden]){display:none !important}";

const DsFullscreenSpinner = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
  }
  connectedCallback() {
    var _a;
    (_a = this.mutationObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    this.mutationObserver = new MutationObserver(() => this.handleBlocking());
    this.mutationObserver.observe(this.self, { attributes: true });
  }
  disconnectedCallback() {
    var _a;
    modalScrollHelper.ModalScrollHelper.instance().unblock(this.self);
    (_a = this.mutationObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
  }
  preventInputIfVisible(e) {
    if (this.isVisible) {
      e.preventDefault();
      e.stopPropagation();
      e.stopImmediatePropagation();
    }
  }
  get isVisible() {
    return getComputedStyle(this.self).display !== "none";
  }
  onKeyDown(e) {
    this.preventInputIfVisible(e);
  }
  onKeyUp(e) {
    this.preventInputIfVisible(e);
  }
  handleBlocking() {
    const scrollHelper = modalScrollHelper.ModalScrollHelper.instance();
    if (this.isVisible) {
      scrollHelper.block(this.self, null);
    }
    else {
      scrollHelper.unblock(this.self);
    }
  }
  render() {
    return (index.h(index.Host, null, index.h("ds-spinner", { size: "large", theme: "dark", animationduration: 1500 })));
  }
  get self() { return index.getElement(this); }
};
DsFullscreenSpinner.style = dsFullscreenSpinnerCss;

exports.ds_fullscreen_spinner = DsFullscreenSpinner;
