// @dvag/design-system/dist/cjs/ds-header-action-list.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const slotObserver = require('./slot-observer-2e83212e.js');
const miscUtil = require('./misc-util-63b423bd.js');
const headerActionHelper = require('./header-action-helper-195b9f78.js');
require('./query-util-8642f149.js');

const DsHeaderActionList = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.children = [];
  }
  toggleOpen() {
    this.open = !this.open;
  }
  /**
   * @internal
   */
  async toHeaderActionListConfiguration() {
    return {
      id: this.self.id,
      label: this.label,
      open: this.open,
      toggleOpen: () => this.toggleOpen(),
      children: await headerActionHelper.extractHeaderActionConfigurations(this.children),
    };
  }
  connectedCallback() {
    this.childObserver = slotObserver.createMetaComponentsSlotObserver(this.self, ":scope > *", (nodes) => {
      this.children = miscUtil.filterElementsByTagPrefix(nodes, "ds-header-action");
    });
  }
  disconnectedCallback() {
    this.childObserver.disconnect();
  }
  get self() { return index.getElement(this); }
};

exports.ds_header_action_list = DsHeaderActionList;
