// @dvag/design-system/dist/cjs/ds-table-header-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const assert = require('./assert-d8ea75c6.js');
const elementIdHandling = require('./element-id-handling-89898a1c.js');
require('./value-store-f683b26f.js');

const dsTableHeaderItemCss = ":host{margin-right:16px;overflow:hidden;display:flex;flex:1;padding-top:8px;padding-bottom:8px}:host:host([hidden]){display:none !important}:host:first-child{padding-left:16px}:host:host(:first-child){padding-left:16px}:host:last-child{margin-right:0;padding-right:16px}:host:host(:host:last-child){margin-right:0;padding-right:16px}:host:host(.align-left){justify-content:flex-start}:host:host(.align-right){justify-content:flex-end}:host:host(.align-center){justify-content:center}:host .sort-icon{cursor:pointer;margin-left:4px;flex:0 0 16px}:host .header{width:100%;overflow:hidden;width:auto;text-align:left;height:100%;border:none;border-radius:4px;text-decoration:none;background:none;padding:4px 0;cursor:pointer;transition:background 250ms ease-in-out;-webkit-appearance:none;-moz-appearance:none;outline:none}:host .header .cell{font-family:\"Helvetica Neue\", \"Helvetica Neue LT W01\";color:#666666;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;width:100%;box-sizing:border-box;padding-left:8px;padding-right:8px}:host .header .cell.multiple-lines{display:-webkit-box;-webkit-line-clamp:var(--ds-table-cell-lines, 1);-webkit-box-orient:vertical;white-space:break-spaces}:host .header .cell.align-left{text-align:start}:host .header .cell.align-right{text-align:end}:host .header .cell.align-center{text-align:center}:host .header .cell ds-text{overflow:hidden;text-overflow:ellipsis}:host .header .cell ::slotted(*){overflow:hidden;text-overflow:ellipsis}:host .header:host(.cell){font-family:\"Helvetica Neue\", \"Helvetica Neue LT W01\";color:#666666;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;width:100%;box-sizing:border-box;padding-left:8px;padding-right:8px}:host .header:host(.cell).multiple-lines{display:-webkit-box;-webkit-line-clamp:var(--ds-table-cell-lines, 1);-webkit-box-orient:vertical;white-space:break-spaces}:host .header:host(.cell).align-left{text-align:start}:host .header:host(.cell).align-right{text-align:end}:host .header:host(.cell).align-center{text-align:center}:host .header:host(.cell) ds-text{overflow:hidden;text-overflow:ellipsis}:host .header:host(.cell) ::slotted(*){overflow:hidden;text-overflow:ellipsis}:host .header[disabled]{cursor:default}:host .header:focus{background-color:#ebebeb !important;--ds-text-color:colors.$anthrazit}:host .header .cell .label-container{display:flex;align-items:center}:host .header .cell .label-container .label{flex:1;display:flex;overflow:hidden;text-overflow:ellipsis}:host .header .cell .label-container .label.multiple-lines{display:-webkit-box;-webkit-box-orient:vertical;white-space:break-spaces}";

const DsTableHeaderItem = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Allows the table to sort this column.
     *
     * @value true: The column is sortable.
     * @value false: The column is not sortable.
     */
    this.sortable = false;
    /**
     * The alignment of the title text and the cell contents within this column.
     * **Warning:** The default alignment which is defined by the `type` property will be overriden.
     *
     * @value left: Child elements are left-aligned.
     * @value right: Child elements right-aligned.
     * @value center: Child elements are centered.
     */
    this.align = null;
    /**
     * Specifies the priority of this column which is considered in the table's secondary sorting.
     *
     * @value number: Priority. The lower this number the higher the chance this column will be considered first
     * in the secondary sorting of the table.
     *
     */
    this.sortpriority = null;
    /**
     * Specifies the number of rows within an column.
     *
     * @value s: Table cells have only 1 text row
     * @value m: Table cells have up to 2 text rows.
     * @value l: Table cell have up to 3 text rows.
     */
    this.linesize = "s";
    /**
     * @internal
     */
    this.usemaxwidth = false;
    this.sortState = "neutral";
    this.internAlign = null;
  }
  onSortableChange() {
    if (this.sortable) {
      elementIdHandling.verifyElementId(this.self);
    }
  }
  onCustomSortChanged() {
    if (this.sortState !== "neutral") {
      this.invokeSort(this.sortState);
    }
  }
  onChangeAlign() {
    this.internAlign = this.getAlignment();
  }
  componentWillLoad() {
    this.internAlign = this.getAlignment();
    assert.assert(!!(this.label || this.columnid) || !this.sortable, () => "Sortierbare Spalten sollten die Eigenschaft 'columnid' oder 'label', zur Unterscheidung, definieren.");
    if (this.sortable) {
      elementIdHandling.verifyElementId(this.self);
    }
  }
  /**
   * @internal
   */
  async setSortDirection(state) {
    this.invokeSort(state);
  }
  /**
   * @internal
   */
  async toColumnSettings() {
    return {
      type: this.type,
      align: this.internAlign,
      lines: this.getCellLines(),
      sizeStyle: this.buildStyle(),
    };
  }
  getCellLines() {
    var _a;
    return {
      s: 1,
      m: 2,
      l: 3,
    }[(_a = this.linesize) !== null && _a !== void 0 ? _a : "s"];
  }
  buildStyle() {
    const style = {
      "max-width": this.getMaxWidth(),
      "min-width": this.getMinWidth(),
    };
    if (this.usemaxwidth) {
      const width = this.getMaxWidth();
      style["width"] = width;
      style["flex"] = `0 0 ${width}`;
    }
    return style;
  }
  getMinWidth() {
    return (this.minwidth || this.maxwidth) + "px";
  }
  getMaxWidth() {
    return (this.maxwidth || this.minwidth) + "px";
  }
  getAlignment() {
    var _a;
    return ((_a = this.align) !== null && _a !== void 0 ? _a : (this.type === "currency" || this.type === "numeric" ? "right" : "left"));
  }
  getSortIcon() {
    if (this.sortState === "neutral") {
      return "sortieren";
    }
    else if (this.sortState === "asc") {
      return "sortieren-aufsteigend";
    }
    else {
      return "sortieren-absteigend";
    }
  }
  getHeaderCellLines() {
    return this.useMultipleLines() ? 2 : 1;
  }
  useMultipleLines() {
    return /\s|\xAD/.test(this.label); // &shy;
  }
  render() {
    const alignClass = {};
    alignClass["align-" + this.internAlign] = true;
    return (index.h(index.Host, { style: this.buildStyle(), class: alignClass, role: "columnheader" }, index.h("button", { id: elementIdHandling.ElementIdGenerator.createId(this.self, "sort-button"), type: "button", disabled: !this.sortable, onClick: () => this.invokeSort(), class: "header" }, index.h("div", { class: Object.assign({ cell: true }, alignClass) }, index.h("div", { class: "label-container" }, index.h("div", { class: {
        label: true,
        "multiple-lines": this.useMultipleLines(),
      }, style: {
        "-webkit-line-clamp": `${this.getHeaderCellLines()}`,
      } }, this.renderLabel(), index.h("slot", null)), this.renderSortButton())))));
  }
  renderLabel() {
    if (this.label != null) {
      return (index.h("ds-text", { type: "it", color: this.sortState !== "neutral" ? "gold" : undefined }, this.label));
    }
  }
  renderSortButton() {
    if (this.sortable) {
      return index.h("ds-icon", { class: "sort-icon", icon: this.getSortIcon(), size: 16 });
    }
  }
  getNextSortState() {
    if (this.sortState == "neutral") {
      return "asc";
    }
    return this.sortState === "asc" ? "desc" : "asc";
  }
  invokeSort(state = undefined) {
    if (this.sortable) {
      this.sortState = state !== null && state !== void 0 ? state : this.getNextSortState();
      if (this.sortState !== "neutral") {
        const table = this.getParentTable();
        table.invokeSort(this.self, this.sortState === "desc").then();
      }
    }
  }
  getParentTable() {
    return this.self.closest("ds-table");
  }
  get self() { return index.getElement(this); }
  static get watchers() { return {
    "sortable": ["onSortableChange"],
    "customSort": ["onCustomSortChanged"],
    "align": ["onChangeAlign"]
  }; }
};
DsTableHeaderItem.style = dsTableHeaderItemCss;

exports.ds_table_header_item = DsTableHeaderItem;
