// @dvag/design-system/dist/cjs/dx-announcement-bar.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const renderUtil = require('./render-util-6b278c0d.js');

const dxAnnouncementBarCss = ":host:host([hidden]){display:none !important}:host .announcement-content{display:flex;--dx-text-color:#2b2b2b;--dx-icon-color:#2b2b2b;padding:16px 0}:host .announcement-content .icon-container{flex-shrink:0;margin-right:8px}:host .announcement-content .text-container{flex:1}:host .announcement-content .sub-label{margin-top:8px;--dx-text-font-size-small:12px;--dx-text-line-height-small:16px;--dx-text-font-size-large:14px;--dx-text-line-height-large:20px}";

const DxAnnouncementBar = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Specifies the color of the icon.
     *
     * @value string: The background color of this component.
     *
     * @example color
     */
    this.color = "white";
  }
  renderIconContainer() {
    if (!this.icon) {
      return;
    }
    return (index.h("div", { class: "icon-container" }, index.h("dx-icon", { icon: this.icon, color: this.iconcolor, size: 24 })));
  }
  renderSubLabel() {
    if (renderUtil.isSlotUndefined(this.self, "sublabel")) {
      return;
    }
    return (index.h("div", { class: "sub-label" }, index.h("slot", { name: "sublabel" })));
  }
  renderContent() {
    return (index.h("div", { class: "content-container" }, index.h("slot", null), this.renderSubLabel()));
  }
  render() {
    return (index.h("dx-container", { color: this.color, type: "page" }, index.h("div", { class: "announcement-content" }, this.renderIconContainer(), this.renderContent())));
  }
  get self() { return index.getElement(this); }
};
DxAnnouncementBar.style = dxAnnouncementBarCss;

exports.dx_announcement_bar = DxAnnouncementBar;
