// @dvag/design-system/dist/cjs/dx-container.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const colorPalette = require('./color-palette-ed0f6557.js');
require('./assert-d8ea75c6.js');

const dxContainerCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .container.type-page{padding:0 16px;max-width:1460px;box-sizing:border-box}@media (min-width: 476px){:host .container.type-page{padding:0 24px}}@media (min-width: 760px){:host .container.type-page{padding:0 32px}}@media (min-width: 1280px){:host .container.type-page{padding:0 40px}}@media (min-width: 1460px){:host .container.type-page{margin:auto}}";

const DxContainer = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * The type of the container.
     *
     * @value page: The container contains paddings and margins necessary to position the content in the middle of the screen.
     * This type should only be used in a top level context.
     * @value div: The container behaves like a div.
     */
    this.type = "div";
  }
  renderDivType() {
    return index.h("slot", null);
  }
  renderPageContainerType() {
    return (index.h("div", { class: {
        container: true,
        "type-page": this.type === "page",
      } }, index.h("slot", null)));
  }
  render() {
    var _a;
    const hexColor = (_a = new colorPalette.ColorPalette(true).getColorCode(this.color)) !== null && _a !== void 0 ? _a : "white";
    return (index.h(index.Host, { style: {
        "background-color": hexColor,
      } }, this.type === "div"
      ? this.renderDivType()
      : this.renderPageContainerType()));
  }
};
DxContainer.style = dxContainerCss;

exports.dx_container = DxContainer;
