// @dvag/design-system/dist/cjs/dx-card-image.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const assert = require('./assert-d8ea75c6.js');
const breakpointAware = require('./breakpoint-aware-cc9d0ee6.js');

const dxCardImageCss = ":host{display:block;line-height:0}:host:host([hidden]){display:none !important}:host .proportional-image{width:100%}:host .cover-image{background-size:cover}";

const DxCardImage = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Defines the fit of the image.
     *
     * @value  proportional: The image is scaled proportional to its width.
     * @value cover: The image is displayed with a set height. The height is set via the `coverheightmq` properties. The visible part of the image can be controlled using the `coverpositionmq` properties.
     */
    this.fit = "proportional";
  }
  connectedCallback() {
    this.breakpointAware = new breakpointAware.BreakpointAware(() => index.forceUpdate(this), true);
  }
  disconnectedCallback() {
    this.breakpointAware.disconnect();
  }
  render() {
    var _a, _b;
    assert.assert(this.url != null, () => "<dx-card-image-header> Eigenschaft 'url' muss gesetzt sein.");
    if (this.fit === "proportional") {
      return (index.h(index.Host, null, index.h("img", { src: this.url, alt: "", class: "proportional-image" })));
    }
    else {
      const coverHeight = (_a = this.breakpointAware.getEffectiveValue({
        mq1: this.coverheightmq1,
        mq2: this.coverheightmq2,
        mq3: this.coverheightmq3,
        mq4: this.coverheightmq4,
        mq5: this.coverheightmq5,
      })) !== null && _a !== void 0 ? _a : "300px";
      const coverPosition = (_b = this.breakpointAware.getEffectiveValue({
        mq1: this.coverpositionmq1,
        mq2: this.coverpositionmq2,
        mq3: this.coverpositionmq3,
        mq4: this.coverpositionmq4,
        mq5: this.coverpositionmq5,
      })) !== null && _b !== void 0 ? _b : "center";
      return (index.h(index.Host, null, index.h("div", { class: "cover-image", style: {
          height: coverHeight,
          "background-position": coverPosition,
          "background-image": `url(${this.url})`,
        } })));
    }
  }
};
DxCardImage.style = dxCardImageCss;

exports.dx_card_image = DxCardImage;
