// @dvag/design-system/dist/cjs/dx-element-swapper_2.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const renderUtil = require('./render-util-6b278c0d.js');

const dxElementSwapperCss = ":host{display:block;position:relative;width:max-content;height:max-content}:host:host([hidden]){display:none !important}:host .swappable{line-height:0;will-change:transform, opacity;transition:transform var(--dx-element-swapper-transition-duration, 150ms) ease-in-out, opacity var(--dx-element-swapper-transition-duration, 150ms) ease-in-out;z-index:1}:host .swappable.hidden{opacity:0;transform:scale(0.25);z-index:-1}:host .swappable.overlap-parent{position:absolute;top:0;left:0;width:100%;height:100%;display:flex;align-items:center;justify-content:center}";

const DxElementSwapper = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Defines whether the first or the second slot is currently visible.
     *
     * @value first: First slot is visible.
     * @value second: Second slot is visible.
     * @value none: No slot is visible.
     */
    this.active = "first";
  }
  render() {
    return (index.h(index.Host, null, index.h("div", { class: {
        swappable: true,
        hidden: this.active !== "first",
      } }, index.h("slot", { name: "first" })), index.h("div", { class: {
        swappable: true,
        "overlap-parent": true,
        hidden: this.active !== "second",
      } }, index.h("slot", { name: "second" }))));
  }
};
DxElementSwapper.style = dxElementSwapperCss;

const dxHorizontalGrowBoxCss = ".sc-dx-horizontal-grow-box-h{display:flex;flex-direction:row;visibility:hidden}.sc-dx-horizontal-grow-box-h([hidden]).sc-dx-horizontal-grow-box-h{display:none !important}.animate-from-left.sc-dx-horizontal-grow-box-h{flex-direction:row-reverse}.sc-dx-horizontal-grow-box-h .slot-container.sc-dx-horizontal-grow-box{display:flex;width:max-content}.overflow-hidden.sc-dx-horizontal-grow-box-h{overflow:hidden}.transition-enabled.sc-dx-horizontal-grow-box-h{transition:width 300ms ease-in-out}.visible.sc-dx-horizontal-grow-box-h{visibility:visible}";

const DxHorizontalGrowBox = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.overflowHidden = true;
    this.visible = true;
    this.transitionEnabled = false;
    /**
     * Whether the container is open or not. When changed the container transitions smoothly to the correct width.
     *
     * @value true: The container displays it's content.
     * @value false: The container hides it's content.
     */
    this.open = true;
    /**
     * Disconnects the resizeObserver after
     */
    this.limitresizes = -1;
    this.animationmode = "open-from-right";
  }
  onForceWidthChanged() {
    var _a;
    if (this.forcewidth == null) {
      this.initializeResizeObserver();
    }
    else {
      (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
      this.resizeObserver = null;
      this.updateWidth();
    }
  }
  onOpenChanged() {
    this.updateWidth();
  }
  connectedCallback() {
    if (this.forcewidth == null) {
      this.initializeResizeObserver();
    }
    else {
      this.updateWidth();
    }
  }
  componentDidLoad() {
    renderUtil.onNextFrame(() => {
      this.transitionEnabled = true;
      this.overflowHidden = !this.open;
      this.visible = this.open;
    });
  }
  onTransitionStart(e) {
    if (e.target !== this.self) {
      return;
    }
    this.overflowHidden = true;
    this.visible = true;
  }
  onTransitionEnd(e) {
    if (e.target !== this.self) {
      return;
    }
    this.overflowHidden = !this.open;
    this.visible = this.open;
  }
  getWidthOfContent() {
    var _a, _b, _c;
    if (!this.open) {
      return 0;
    }
    return (_c = (_a = this.forcewidth) !== null && _a !== void 0 ? _a : (_b = this.slotContainer) === null || _b === void 0 ? void 0 : _b.scrollWidth) !== null && _c !== void 0 ? _c : 0;
  }
  initializeResizeObserver() {
    this.resizeObserver = new ResizeObserver(() => {
      renderUtil.onNextFrame(() => this.updateWidth());
    });
  }
  updateWidth() {
    // Immediately update width on self
    this.self.style.width = this.getWidthOfContent() + "px";
  }
  setSlotContainer(slotContainer) {
    if (this.forcewidth != null) {
      return;
    }
    if (this.slotContainer !== slotContainer) {
      this.resizeObserver.disconnect();
    }
    this.slotContainer = slotContainer;
    this.resizeObserver.observe(this.slotContainer);
  }
  disconnectedCallback() {
    var _a;
    (_a = this.resizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
  }
  render() {
    return (index.h(index.Host, { class: {
        "animate-from-left": this.animationmode === "open-from-left",
        "overflow-hidden": this.overflowHidden,
        visible: this.visible,
        "transition-enabled": this.transitionEnabled,
      } }, index.h("div", { class: "slot-container", ref: (ref) => this.setSlotContainer(ref) }, index.h("slot", null))));
  }
  get self() { return index.getElement(this); }
  static get watchers() { return {
    "forcewidth": ["onForceWidthChanged"],
    "open": ["onOpenChanged"]
  }; }
};
DxHorizontalGrowBox.style = dxHorizontalGrowBoxCss;

exports.dx_element_swapper = DxElementSwapper;
exports.dx_horizontal_grow_box = DxHorizontalGrowBox;
