// @dvag/design-system/dist/cjs/dx-fd-outline-box.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const colorPalette = require('./color-palette-ed0f6557.js');
require('./assert-d8ea75c6.js');

const dxFdOutlineBoxCss = ":host:host([hidden]){display:none !important}:host .content{--dx-content-padding-left:16px;--dx-content-padding-right:16px;--dx-content-padding-top:32px;--dx-content-padding-bottom:16px;padding:var(--dx-content-padding-top) var(--dx-content-padding-left) var(--dx-content-padding-bottom) var(--dx-content-padding-right);position:relative;box-sizing:border-box}@media (min-width: 476px){:host .content{--dx-content-padding-left:24px;--dx-content-padding-right:24px}}@media (min-width: 1024px){:host .content{--dx-content-padding-left:32px;--dx-content-padding-right:32px}}@media (min-width: 1280px){:host .content{--dx-content-padding-left:40px;--dx-content-padding-right:40px}}:host .content svg{position:absolute;overflow:visible;top:0;left:0;width:100%;height:100%;pointer-events:none}:host .triangle{display:block;position:relative;width:100%;padding-top:8.74887%}:host .triangle .svg-container{position:absolute;top:0;left:0;width:100%;height:100%}:host .triangle .svg-container svg{display:block;overflow:visible;width:100%;height:100%}:host svg{padding:0 1.5px;box-sizing:border-box}";

const DxFdOutlineBox = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * The color of the outline.
     *
     * @value string: Name of the color.
     * @internal
     */
    this.color = "gold";
    /**
     * @internal
     */
    this.strokewidth = 3;
  }
  render() {
    const hexColor = new colorPalette.ColorPalette(true).getColorCode(this.color);
    const svgLineStyles = {
      "vector-effect": "non-scaling-stroke",
      "stroke-linejoin": "miter",
      "stroke-width": `${this.strokewidth}px`,
      stroke: hexColor,
      fill: "none",
      "shape-rendering": "geometricPrecision",
    };
    return (index.h(index.Host, null, index.h("div", { class: "content" }, index.h("slot", null), index.h("svg", { preserveAspectRatio: "none", viewBox: "0 0 100 100" }, index.h("line", { x1: 0, y1: 0, x2: 0, y2: 100, style: svgLineStyles }), index.h("line", { x1: 100, y1: 0, x2: 100, y2: 100, style: svgLineStyles }))), index.h("div", { class: "triangle" }, index.h("div", { class: "svg-container" }, index.h("svg", { preserveAspectRatio: "none", viewBox: "0 0 100 100" }, index.h("polyline", { points: "0,-5 0,100 100,0 100,-5", style: svgLineStyles }))))));
  }
};
DxFdOutlineBox.style = dxFdOutlineBoxCss;

exports.dx_fd_outline_box = DxFdOutlineBox;
