// @dvag/design-system/dist/cjs/dx-header-navigation-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const elementIdHandling = require('./element-id-handling-89898a1c.js');
const slotObserver = require('./slot-observer-2e83212e.js');
const navigation_util = require('./navigation.util-ddaead16.js');
require('./value-store-f683b26f.js');
require('./query-util-8642f149.js');

const DxHeaderNavigationItem = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.interaction = index.createEvent(this, "interaction", 7);
    this.internalChangeSidebarVisibility = index.createEvent(this, "internalChangeSidebarVisibility", 7);
    /**
     * The size of the dropdown.
     *
     * @value m: A medium sized navigation item is displayed.
     * @value l: A large sized navigation item is displayed.
     */
    this.size = "l";
  }
  connectedCallback() {
    this.slotObserver =
      slotObserver.createMetaComponentsSlotObserver(this.self, ":scope > dx-header-navigation-item", (nodes) => {
        this.children = nodes.sort(navigation_util.sortByNavigationItemSize);
      });
  }
  componentWillLoad() {
    elementIdHandling.verifyElementId(this.self);
  }
  disconnectedCallback() {
    this.slotObserver.disconnect();
  }
  /**
   * @internal
   *
   * Inserts a navigationItem recursively into the tree.
   * @param belongsToPage: The page the navigation item belongs to
   * @param pageIndexCount: An iterating pageIndex reference which
   */
  async insertNavigationTree(belongsToPage, pageIndexCount) {
    if (belongsToPage == null) {
      return;
    }
    const navigationItem = this.createNavigationItem();
    // If we have any children we create a new page for them and recursively add new navigation pages and items.
    if (this.children.length > 0) {
      const refersToPage = {
        pageIndex: pageIndexCount.value,
        previousPage: belongsToPage,
        navigationItems: [],
      };
      navigationItem.refersToPage = refersToPage;
      // The pageIndexCount is a reference which tracks the recursive count. If we add a new navigation page we have to increment the count.
      pageIndexCount.value += 1;
      for (const child of this.children) {
        await child.insertNavigationTree(refersToPage, pageIndexCount);
      }
    }
    belongsToPage.navigationItems.push(navigationItem);
  }
  createNavigationItem() {
    var _a;
    const value = (_a = this.value) !== null && _a !== void 0 ? _a : this.label;
    return {
      id: this.self.id,
      size: this.size,
      label: this.label,
      value: value,
      refersToPage: null,
      interact: () => this.interaction.emit({
        value: value,
        closeSidebar: () => {
          this.internalChangeSidebarVisibility.emit(false);
        },
      }),
    };
  }
  get self() { return index.getElement(this); }
};

exports.dx_header_navigation_item = DxHeaderNavigationItem;
