// @dvag/design-system/dist/cjs/dx-header-user-sidebar-content.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const miscUtil = require('./misc-util-63b423bd.js');
const elementIdHandling = require('./element-id-handling-89898a1c.js');
require('./value-store-f683b26f.js');

const dxHeaderUserSidebarContentCss = ":host:host([hidden]){display:none !important}:host .container dx-button{display:block}";

const DxHeaderUserSidebarContent = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.signOut = index.createEvent(this, "signOut", 7);
    this.accountInteraction = index.createEvent(this, "accountInteraction", 7);
    /**
     * Defines whether the account button "Benutzerkonto" is displayed or not.
     *
     * @value true: The account button is visible.
     * @value false: The account button is not visible.
     */
    this.accountavailable = true;
  }
  renderAccountButton() {
    if (!this.accountavailable) {
      return;
    }
    return (index.h("dx-button", { id: elementIdHandling.ElementIdGenerator.createId(this.self, "account"), type: "text", label: "Benutzerkonto", onClick: () => this.accountInteraction.emit() }));
  }
  renderLogoutButton() {
    return (index.h("dx-button", { id: elementIdHandling.ElementIdGenerator.createId(this.self, "sign-out"), type: "text", label: "Abmelden", onClick: () => this.signOut.emit() }));
  }
  render() {
    return (index.h("dx-sidebar-content", { class: "container" }, index.h("dx-user-avatar", { imageurl: this.imageurl, initials: miscUtil.generateInitialsForName(this.name), size: "xl" }), index.h("dx-spacer", { mq1: "24v" }), index.h("dx-text", { type: "its", color: "tags" }, "Angemeldete Person"), index.h("dx-spacer", { mq1: "8v" }), index.h("dx-text", { type: "ps", color: "headline" }, this.name), index.h("dx-spacer", { mq1: "32v" }), this.renderAccountButton(), this.renderLogoutButton()));
  }
  get self() { return index.getElement(this); }
};
DxHeaderUserSidebarContent.style = dxHeaderUserSidebarContentCss;

exports.dx_header_user_sidebar_content = DxHeaderUserSidebarContent;
