// @dvag/design-system/dist/cjs/dx-legend.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const newKiSnycIcon = require('./new-ki-snyc-icon-665f6e39.js');

const dxLegendCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .text{display:flex;align-items:center}:host .text:not(:last-child){margin-bottom:8px}:host .required-icon{margin:0 4px}";

const DxLegend = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Indicates if the required hint is shown.
     * @value false: Required hint is not shown.
     * @value true: Required hint is shown.
     */
    this.requiredhint = true;
    /**
     * Indicates if the KI-Sync hint is shown.
     * @value true: KI-Sync hint is shown.
     * @value false: KI-Sync hint is not shown.
     */
    this.kisynchint = false;
  }
  renderRequiredhint() {
    if (!this.requiredhint) {
      return;
    }
    return (index.h("dx-text", { class: "text", type: "its", color: "headline" }, index.h("dx-icon", { class: "required-icon", size: 16, icon: "state-required" }), "markiert Pflichtfeld."));
  }
  renderKiSyncedHint() {
    if (!this.kisynchint) {
      return;
    }
    return (index.h("dx-text", { class: "text", type: "its", color: "headline" }, index.h(newKiSnycIcon.NewKiSyncIcon, null), "synchronisiert system\u00FCbergreifend."));
  }
  render() {
    return (index.h(index.Host, null, this.renderRequiredhint(), this.renderKiSyncedHint()));
  }
};
DxLegend.style = dxLegendCss;

exports.dx_legend = DxLegend;
