// @dvag/design-system/dist/cjs/dx-scrollable-shadow_2.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-2f41386b.js');
const renderUtil = require('./render-util-6b278c0d.js');

const dxScrollableShadowCss = ".sc-dx-scrollable-shadow-h{display:block;position:relative}.sc-dx-scrollable-shadow-h([hidden]).sc-dx-scrollable-shadow-h{display:none !important}.sc-dx-scrollable-shadow-h .shadow.sc-dx-scrollable-shadow{position:absolute;height:3px;background-color:rgba(0, 0, 0, 0.03);left:0;width:100%;transition:opacity 200ms ease-in-out;opacity:0;z-index:1}.sc-dx-scrollable-shadow-h .shadow.top.sc-dx-scrollable-shadow{top:0}.sc-dx-scrollable-shadow-h .shadow.top.show-border.sc-dx-scrollable-shadow{border-top:1px solid #d6d6d6}.sc-dx-scrollable-shadow-h .shadow.bottom.sc-dx-scrollable-shadow{bottom:0}.sc-dx-scrollable-shadow-h .shadow.bottom.show-border.sc-dx-scrollable-shadow{border-bottom:1px solid #d6d6d6}.sc-dx-scrollable-shadow-h .shadow.visible.sc-dx-scrollable-shadow{opacity:1}";

const DxScrollableShadow = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.showborders = false;
    this.showTopShadow = false;
    this.showBottomShadow = false;
  }
  connectedCallback() {
    this.updateScrollShadowVisibilities();
    this.resizeObserver = new ResizeObserver(() => {
      this.updateScrollShadowVisibilities();
    });
    this.resizeObserver.observe(this.self);
  }
  disconnectedCallback() {
    this.resizeObserver.disconnect();
  }
  onScrolling() {
    this.updateScrollShadowVisibilities();
  }
  updateScrollShadowVisibilities() {
    const slottedScrollElement = this.self.querySelector("*:not(.shadow)");
    if (slottedScrollElement) {
      this.showTopShadow = slottedScrollElement.scrollTop > 0;
      this.showBottomShadow =
        slottedScrollElement.offsetHeight + slottedScrollElement.scrollTop <
          slottedScrollElement.scrollHeight;
    }
  }
  render() {
    return (index.h(index.Host, null, index.h("div", { class: {
        shadow: true,
        top: true,
        visible: this.showTopShadow,
        "show-border": this.showborders,
      } }), index.h("div", { class: {
        shadow: true,
        bottom: true,
        visible: this.showBottomShadow,
        "show-border": this.showborders,
      } }), index.h("slot", null)));
  }
  get self() { return index.getElement(this); }
};
DxScrollableShadow.style = dxScrollableShadowCss;

const dxSidebarContentCss = ":host{--dx-page-padding:16px;display:flex;flex-direction:column;box-sizing:border-box;height:100%;padding:16px 0}:host:host([hidden]){display:none !important}@media (min-width: 476px){:host{--dx-page-padding:24px}}@media (min-width: 760px){:host{--dx-page-padding:32px}}@media (min-width: 1280px){:host{--dx-page-padding:40px}}:host .header{margin-bottom:16px;padding:0 var(--dx-sidebar-content-padding, var(--dx-page-padding, 16px))}:host .scrollable-shadow-container{overflow:hidden}:host .scrollable-shadow-container.content-grows{flex:1}:host .scrollable-shadow-container .content{overflow-y:auto;overflow-x:hidden;padding:0 var(--dx-sidebar-content-padding, var(--dx-page-padding, 16px));height:100%}:host .scrollable-shadow-container .content.content-grow{flex:1}:host .footer{margin-top:16px;padding:0 var(--dx-sidebar-content-padding, var(--dx-page-padding, 16px))}";

const DxSidebarContent = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Defines if the content area should claim the available space.
     * @internal
     */
    this.contentgrows = true;
  }
  renderSlot(slot) {
    if (renderUtil.isSlotUndefined(this.self, slot)) {
      return;
    }
    return (index.h("div", { class: slot }, index.h("slot", { name: slot })));
  }
  render() {
    return (index.h(index.Host, null, this.renderSlot("header"), index.h("dx-scrollable-shadow", { showborders: true, class: {
        "scrollable-shadow-container": true,
        "content-grows": this.contentgrows,
      } }, index.h("div", { class: { content: true } }, index.h("slot", null))), this.renderSlot("footer")));
  }
  get self() { return index.getElement(this); }
};
DxSidebarContent.style = dxSidebarContentCss;

exports.dx_scrollable_shadow = DxScrollableShadow;
exports.dx_sidebar_content = DxSidebarContent;
