// @dvag/design-system/dist/cjs/render-util-6b278c0d.js
'use strict';

function onNextFrame(fn) {
  requestAnimationFrame(() => requestAnimationFrame(() => fn()));
}
function getSlottedElements(element) {
  return Array.from(element.querySelectorAll("[slot]"));
}
function getDefinedSlots(element) {
  return getSlottedElements(element).map((element) => element.slot);
}
function isSlotDefined(element, slotName) {
  return !!element.querySelector(`[slot="${slotName}"]`);
}
function isSlotUndefined(element, slotName) {
  return !isSlotDefined(element, slotName);
}
// Color calculations
function toHex(dec) {
  return Math.round(dec).toString(16);
}
function lighten(hex, factor) {
  const lightenColor = lightenRgb(toRgb(hex), factor);
  return rgbToHex(lightenColor);
}
function lightenRgb(rgb, factor) {
  return rgb.map((value) => 255 - factor * (255 - value));
}
function rgbToHex(rgb) {
  return "#" + rgb.map((val) => toHex(val)).join("");
}
function toRgb(hex) {
  const hexValues = hex.replace("#", "").match(/.{1,2}/g);
  return hexValues.map((hexValue) => parseInt(hexValue, 16));
}

exports.getDefinedSlots = getDefinedSlots;
exports.isSlotDefined = isSlotDefined;
exports.isSlotUndefined = isSlotUndefined;
exports.lighten = lighten;
exports.onNextFrame = onNextFrame;
exports.toHex = toHex;
