// @dvag/design-system/dist/collection/components/scaffold/ds-logo/ds-logo.js
import { h, Host, } from "@stencil/core";
/**
 * Displays the DVAG logo
 * @path /Scaffold
 * @name Logo
 * @icon sign
 * @stable
 *
 */
export class DsLogo {
  constructor() {
    /**
     * Provides the logo in different styles.
     *
     * @value bildmarke: DVAG-Logo without word mark.
     * @value wortmarke: DVAG-Logo with word-mark.
     * @value wortmarke-allfinanz: DVAG-Logo with `Allfinanz` word mark.
     * @example logo
     */
    this.type = "bildmarke";
  }
  componentDidRender() {
    this.self.shadowRoot.querySelector("span").innerHTML = `<ds-logo-${this.type}></ds-logo-${this.type}>`;
  }
  render() {
    return (h(Host, null, h("span", null)));
  }
  static get is() { return "ds-logo"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-logo.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-logo.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "bildmarke: DVAG-Logo without word mark."
            }, {
              "name": "value",
              "text": "wortmarke: DVAG-Logo with word-mark."
            }, {
              "name": "value",
              "text": "wortmarke-allfinanz: DVAG-Logo with `Allfinanz` word mark."
            }, {
              "name": "example",
              "text": "logo"
            }],
          "text": "Provides the logo in different styles."
        },
        "attribute": "type",
        "reflect": false,
        "defaultValue": "\"bildmarke\""
      }
    };
  }
  static get elementRef() { return "self"; }
}
