// @dvag/design-system/dist/esm/ds-column.entry.js
import { r as registerInstance, h, H as Host } from './index-04c470ea.js';
import { c as assertPropertyValue } from './assert-7fc7463a.js';

const dsColumnCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .column{display:flex;flex-direction:column}";

const DsColumn = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Gibt an, welche Ausrichtung die Elemente haben sollen.
     * @value string: left
     * @value string: center
     * @value string: right
     * @value string: stretch
     * @example example
     */
    this.align = "left";
  }
  render() {
    assertPropertyValue("ds-column", "align", this.align, [
      "left",
      "center",
      "right",
      "stretch",
    ]);
    return (h(Host, null, h("div", { class: "column", style: { "align-items": this.getAlignment() } }, h("slot", null))));
  }
  getAlignment() {
    return alignments[this.align];
  }
};
const alignments = {
  center: "center",
  left: "flex-start",
  right: "flex-end",
  stretch: "stretch",
};
DsColumn.style = dsColumnCss;

export { DsColumn as ds_column };
