// @dvag/design-system/dist/esm/ds-footer.entry.js
import { r as registerInstance, c as createEvent, h, H as Host } from './index-04c470ea.js';

const dsFooterCss = ":host{display:block;position:-webkit-sticky;position:sticky;z-index:900;bottom:0;left:0;width:100%;background-color:#424242}:host:host([hidden]){display:none !important}:host .footer-bar{padding:0 16px;display:block;box-sizing:border-box;box-shadow:0 4px 12px 0 rgba(0, 0, 0, 0.16)}@media (min-width: 768px){:host .footer-bar{padding:0 24px}}@media (min-width: 1200px){:host .footer-bar{padding:0 32px}}:host .footer-bar .page-wrapped-content{max-width:1200px;margin:0 auto}:host .footer-bar .content{height:100%;display:flex;align-items:center;justify-content:space-between;padding-top:24px;padding-bottom:24px}@media (max-width: 768px){:host .footer-bar .content{flex-direction:column;align-items:flex-start}}:host .footer-bar .content .link-container{display:flex;justify-content:flex-end}:host .footer-bar .content .link-container ds-text-button{margin-left:64px}:host .footer-bar .content .link-container ds-text-button ds-text{--ds-text-color:#c1c1c1}:host .footer-bar .content .link-container ds-text-button:hover ds-text{--ds-text-color:#a7893d}@media (max-width: 768px){:host .footer-bar .content .link-container{flex-direction:column;align-items:flex-start}:host .footer-bar .content .link-container ds-text-button{margin-left:0}}";

const DsFooter = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.infoButtonClicked = createEvent(this, "infoButtonClicked", 7);
    /**
     * Defines the visibility of "Nutzungsbedingungen".
     *
     * @value true: shows "Nutzungsbedinungen".
     * @value false: hides "Nutzungsbedinungen".
     *
     * @example show-terms-of-use
     */
    this.showtermsofuse = true;
    /**
     * Defines the visibility of "Impressum".
     *
     * @value true: shows "Impressum".
     * @value false: hides "Impressum".
     *
     * @example show-imprint
     */
    this.showimprint = true;
  }
  renderInfoLabels() {
    const infoLabels = [PRIVACY_BUTTON_LABEL];
    if (this.showtermsofuse) {
      infoLabels.unshift(TERMS_OF_USE_BUTTON_LABEL);
    }
    if (this.showimprint) {
      infoLabels.push(IMPRINT_BUTTON_LABEL);
    }
    return infoLabels.map((infoButtonLabel) => (h("ds-text-button", { id: `ds-footer-${infoButtonLabel}`, onClick: () => this.infoButtonClicked.emit(infoButtonLabel) }, h("ds-text", { type: "itk" }, infoButtonLabel))));
  }
  render() {
    return (h(Host, null, h("div", { class: "footer-bar" }, h("div", { class: "content page-wrapped-content" }, h("ds-text", { type: "itk", color: "hellgrau-3" }, this.companyinfo), h("div", { class: "link-container" }, this.renderInfoLabels())))));
  }
};
const IMPRINT_BUTTON_LABEL = "Impressum";
const TERMS_OF_USE_BUTTON_LABEL = "Nutzungsbedingungen";
const PRIVACY_BUTTON_LABEL = "Datenschutz";
DsFooter.style = dsFooterCss;

export { DsFooter as ds_footer };
