// @dvag/design-system/dist/esm/ds-navigated-content-menu.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-04c470ea.js';
import { c as createMetaComponentsSlotObserver } from './slot-observer-9196738e.js';
import { v as verifyElementId, E as ElementIdGenerator } from './element-id-handling-bb37154e.js';
import { o as onNextFrame } from './render-util-1dee6655.js';
import { s as subscribeToHeaderHeight, a as subscribeToContentHeightBelowHeader, H as HeaderService } from './header-service-3e7097ef.js';
import { n as notHidden } from './query-util-02753b43.js';
import './value-store-f4663db1.js';
import './multicast-observable-f2e53ec0.js';

const dsNavigatedContentMenuCss = ":host{display:block;position:sticky;position:-webkit-sticky;top:0}:host:host([hidden]){display:none !important}:host .menu-item{margin-top:16px}:host .menu-item:first-child{margin-top:0}:host .menu-item .menu-item-label{display:flex}:host .menu-item .menu-item-label .main{flex-grow:1;overflow:hidden}:host .menu-item .menu-item-label .main .navigate-button{-moz-appearance:none;-webkit-appearance:none;appearance:none;border:0;padding:0;margin:0;background:none;cursor:pointer;text-align:left;display:flex;align-items:center;--ds-text-color:#424242}:host .menu-item .menu-item-label .main .navigate-button:focus{outline:none}:host .menu-item .menu-item-label .main .navigate-button ds-icon{margin-left:8px;display:none}:host .menu-item .menu-item-label .main .navigate-button ds-icon.visible{display:inline}:host .menu-item .menu-item-label .main .navigate-button ds-icon.flip{transform:rotate(180deg)}:host .menu-item .menu-item-label .status{flex:0 0 16px;margin-left:16px;position:relative;top:5px}:host .menu-item.level-1.active>.menu-item-label .navigate-button{--ds-text-color:#a7893d;--ds-text-font-weight:500}:host .menu-item .children{display:flex;box-sizing:border-box;transition:max-height 300ms ease-in-out;overflow:hidden}:host .menu-item .children .slider{position:relative;top:0;left:0;margin:16px 24px 0 12px;width:4px;flex-shrink:0}:host .menu-item .children .slider .line{height:100%;background-color:#dedede;margin:0 1px}:host .menu-item .children .slider .knob{position:absolute;left:0;top:0;width:100%;height:24px;background-color:#a7893d;border-radius:2px;transition:transform 300ms ease-in-out, height 300ms ease-in-out}:host .menu-item .children .items{width:calc(100% - 40px);margin-top:16px}:host .menu-item .children:not(.visible){max-height:0 !important;pointer-events:none}:host .menu-item .measurement-box{position:absolute;width:194px;left:100px;pointer-events:none;visibility:hidden}:host .menu-item .measurement-box .items{width:100%;margin-top:16px}";

const DsNavigatedContentMenu = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.navigate = createEvent(this, "navigate", 7);
    this.knobs = [];
    this.menuItems = [];
    this.headerHeight = 0;
    this.contentBelowHeaderHeight = 0;
    this.childrenBoxHeights = [];
  }
  connectedCallback() {
    this.registerMenuItemObserver();
    this.headerHeightSubscription = subscribeToHeaderHeight((height) => (this.headerHeight = height));
    this.contentHeightBelowHeaderSubscription =
      subscribeToContentHeightBelowHeader((height) => (this.contentBelowHeaderHeight = height));
    const headerService = HeaderService.instance();
    this.headerHeight = headerService.headerHeight;
    this.contentBelowHeaderHeight = headerService.contentHeightBelowHeader;
    this.resizeObserver = new ResizeObserver(() => {
      this.updateKnobPositions();
      this.measureChildrenBoxes();
    });
  }
  disconnectedCallback() {
    var _a, _b, _c, _d;
    (_a = this.menuItemObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    (_b = this.headerHeightSubscription) === null || _b === void 0 ? void 0 : _b.unsubscribe();
    (_c = this.contentHeightBelowHeaderSubscription) === null || _c === void 0 ? void 0 : _c.unsubscribe();
    (_d = this.resizeObserver) === null || _d === void 0 ? void 0 : _d.disconnect();
  }
  componentWillLoad() {
    verifyElementId(this.self);
  }
  updateKnobPositions() {
    var _a, _b;
    const knob = this.knobs.find((knob) => this.active.startsWith(knob.parentValue));
    if (knob == null) {
      return;
    }
    const isChildSelected = this.active.length > knob.parentValue.length + "/".length;
    const selectedChild = isChildSelected
      ? this.active.split("/")[1]
      : (_b = (_a = this.menuItems.find((item) => item.value === knob.parentValue)) === null || _a === void 0 ? void 0 : _a.children[0]) === null || _b === void 0 ? void 0 : _b.value;
    const childMenuItem = this.self.shadowRoot.querySelector(`.measurement-box .menu-item[data-path='${knob.parentValue}/${selectedChild}']`);
    if (childMenuItem == null) {
      return;
    }
    const childMenuItemBounds = childMenuItem.getBoundingClientRect();
    const childrenContainerBounds = childMenuItem.parentElement.getBoundingClientRect();
    const knobOffsetY = childMenuItemBounds.y - childrenContainerBounds.y;
    knob.element.style.transform = `translateY(${knobOffsetY}px)`;
    knob.element.style.height = `${childMenuItemBounds.height}px`;
  }
  registerMenuItemObserver() {
    this.menuItemObserver =
      createMetaComponentsSlotObserver(this.self, ":scope > ds-navigated-content-menu-item.hydrated", async (nodes) => {
        // The current implementation only support 2 levels of hierarchy
        this.menuItems = await Promise.all(nodes
          .map(async (node) => {
          return Object.assign(Object.assign({}, (await node.toMenuItemModel())), { children: await Promise.all(Array.from(node.querySelectorAll(notHidden("ds-navigated-content-menu-item"))).map((child) => child.toMenuItemModel())) });
        })
          .filter((model) => model != null));
      });
  }
  measureChildrenBoxes() {
    const childrenBoxHeights = Array.from(this.self.shadowRoot.querySelectorAll(".measurement-box")).map((element) => {
      return {
        value: element.getAttribute("data-measurementbox"),
        height: element.offsetHeight,
      };
    });
    if (diffChildHeights(childrenBoxHeights, this.childrenBoxHeights)) {
      onNextFrame(() => (this.childrenBoxHeights = childrenBoxHeights));
    }
  }
  pathForItem(item) {
    const rootItem = this.menuItems.find((someItem) => { var _a; return someItem === item || ((_a = someItem.children) === null || _a === void 0 ? void 0 : _a.includes(item)); });
    if (rootItem === item) {
      return rootItem.value;
    }
    else {
      const subItem = rootItem.children.find((subItem) => subItem === item);
      return `${rootItem.value}/${subItem.value}`;
    }
  }
  registerKnob(parentValue, element) {
    this.knobs = this.knobs.filter((knob) => knob.parentValue !== parentValue);
    this.knobs.push({ parentValue, element });
  }
  navigateToItem(item) {
    const path = this.pathForItem(item);
    this.navigate.emit({
      path,
      segments: path.split("/"),
    });
  }
  render() {
    return (h(Host, { style: {
        top: `${this.headerHeight +
          this.contentBelowHeaderHeight +
          MARGIN_TO_HEADER_PX}px`,
      } }, h("ds-card", null, h("ds-card-content", null, this.menuItems.map((item) => this.renderMenuItem(item)), h("slot", { name: "bottom" })))));
  }
  renderMenuItem(item) {
    var _a;
    const path = this.pathForItem(item);
    const isActive = this.active.startsWith(path);
    const hasChildren = ((_a = item.children) === null || _a === void 0 ? void 0 : _a.length) > 0;
    return (h("div", { class: {
        "menu-item": true,
        active: isActive,
        "has-children": hasChildren,
        "level-1": true,
      } }, h("div", { class: "menu-item-label" }, h("div", { class: "main" }, h("button", { type: "button", class: "navigate-button", id: ElementIdGenerator.createId(this.self, path), onClick: () => this.navigateToItem(item) }, h("ds-text", null, item.label), h("ds-icon", { icon: "drop-down", size: 16, class: {
        visible: hasChildren && !isActive,
        flip: isActive,
      } }))), this.renderStatusIcons(item)), this.renderItemChildren(item.children, item.value, isActive)));
  }
  renderItemChildren(children, parentValue, isExpanded) {
    if (!((children === null || children === void 0 ? void 0 : children.length) > 0)) {
      return null;
    }
    const heightInfo = this.childrenBoxHeights.find((item) => item.value === parentValue);
    return [
      this.renderItemChildrenMeasurementBox(children, parentValue),
      h("div", { class: {
          children: true,
          visible: isExpanded,
        }, style: {
          "max-height": heightInfo != null ? `${heightInfo.height}px` : "auto",
        } }, h("div", { class: "slider" }, h("div", { class: "line" }), h("div", { class: "knob", ref: (el) => this.registerKnob(parentValue, el) })), h("div", { class: "items" }, children.map((item) => this.renderChild(item, isExpanded)))),
    ];
  }
  renderItemChildrenMeasurementBox(children, parentValue) {
    return (h("div", { class: "measurement-box", "data-measurementbox": parentValue, ref: (ref) => {
        this.resizeObserver.disconnect();
        if (ref != null) {
          this.resizeObserver.observe(ref);
        }
      } }, h("div", { class: "items" }, children.map((item) => this.renderChild(item, false)))));
  }
  renderChild(item, isInteractive) {
    const path = this.pathForItem(item);
    return (h("div", { class: {
        "menu-item": true,
        "level-2": true,
      }, "data-path": path }, h("div", { class: "menu-item-label" }, h("div", { class: "main" }, h("button", { type: "button", class: "navigate-button", tabIndex: isInteractive ? 0 : -1, id: ElementIdGenerator.createId(this.self, path), onClick: () => {
        if (isInteractive) {
          this.navigateToItem(item);
        }
      } }, h("ds-text", null, item.label))), this.renderStatusIcons(item))));
  }
  renderStatusIcons(item) {
    const isError = item.error;
    const isWarning = !isError && item.warning;
    const isSuccess = !isError && !isWarning && item.success;
    return (h("div", { class: "status" }, isWarning ? (h("ds-icon", { icon: "achtung-circle", color: "gelb-1", size: 16 })) : null, isError ? (h("ds-icon", { icon: "achtung", color: "rot-1", size: 16 })) : null, isSuccess ? (h("ds-icon", { icon: "success", color: "gruen-1", size: 16 })) : null));
  }
  get self() { return getElement(this); }
};
const MARGIN_TO_HEADER_PX = 24;
function diffChildHeights(a, b) {
  if (a.length !== b.length) {
    return true;
  }
  const heightSumA = a.reduceRight((prev, curr) => prev + curr.height, 0);
  const heightSumB = b.reduceRight((prev, curr) => prev + curr.height, 0);
  if (heightSumA !== heightSumB) {
    return true;
  }
  return false;
}
DsNavigatedContentMenu.style = dsNavigatedContentMenuCss;

export { DsNavigatedContentMenu as ds_navigated_content_menu };
