// @dvag/design-system/dist/esm/dx-announcement-bar.entry.js
import { r as registerInstance, h, g as getElement } from './index-04c470ea.js';
import { a as isSlotUndefined } from './render-util-1dee6655.js';

const dxAnnouncementBarCss = ":host:host([hidden]){display:none !important}:host .announcement-content{display:flex;--dx-text-color:#2b2b2b;--dx-icon-color:#2b2b2b;padding:16px 0}:host .announcement-content .icon-container{flex-shrink:0;margin-right:8px}:host .announcement-content .text-container{flex:1}:host .announcement-content .sub-label{margin-top:8px;--dx-text-font-size-small:12px;--dx-text-line-height-small:16px;--dx-text-font-size-large:14px;--dx-text-line-height-large:20px}";

const DxAnnouncementBar = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Specifies the color of the icon.
     *
     * @value string: The background color of this component.
     *
     * @example color
     */
    this.color = "white";
  }
  renderIconContainer() {
    if (!this.icon) {
      return;
    }
    return (h("div", { class: "icon-container" }, h("dx-icon", { icon: this.icon, color: this.iconcolor, size: 24 })));
  }
  renderSubLabel() {
    if (isSlotUndefined(this.self, "sublabel")) {
      return;
    }
    return (h("div", { class: "sub-label" }, h("slot", { name: "sublabel" })));
  }
  renderContent() {
    return (h("div", { class: "content-container" }, h("slot", null), this.renderSubLabel()));
  }
  render() {
    return (h("dx-container", { color: this.color, type: "page" }, h("div", { class: "announcement-content" }, this.renderIconContainer(), this.renderContent())));
  }
  get self() { return getElement(this); }
};
DxAnnouncementBar.style = dxAnnouncementBarCss;

export { DxAnnouncementBar as dx_announcement_bar };
