// @dvag/design-system/dist/esm/dx-header-menu-link.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-04c470ea.js';
import { v as verifyElementId, E as ElementIdGenerator } from './element-id-handling-bb37154e.js';
import './value-store-f4663db1.js';

const dxHeaderMenuLinkCss = ":host{display:block}:host:host([hidden]){display:none !important}";

const DxHeaderMenuLink = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.menuInteraction = createEvent(this, "menuInteraction", 7);
    /**
     * Highlights this entry.
     *
     * @value true: Entry highlighted as selected.
     * @value false: Entry not highlighted as selected.
     * @example selected
     */
    this.selected = false;
    /**
     * @internal
     */
    this.internalopen = true;
  }
  componentWillLoad() {
    verifyElementId(this.self);
  }
  /**
   * @internal
   */
  async getDefaultWidth() {
    return this.self.shadowRoot.querySelector("dx-header-menu-button")
      .offsetWidth;
  }
  render() {
    return (h("dx-horizontal-grow-box", { open: this.internalopen }, h("dx-header-menu-button", { id: ElementIdGenerator.createId(this.self, "button"), selected: this.selected, onInteraction: () => this.menuInteraction.emit(this.connectednavigationvalue) }, h("dx-text", null, this.label))));
  }
  get self() { return getElement(this); }
};
DxHeaderMenuLink.style = dxHeaderMenuLinkCss;

export { DxHeaderMenuLink as dx_header_menu_link };
