// @dvag/design-system/dist/esm/dx-header.entry.js
import { r as registerInstance, c as createEvent, f as forceUpdate, h, H as Host, g as getElement } from './index-04c470ea.js';
import { a as assert } from './assert-7fc7463a.js';
import { B as BreakpointAware } from './breakpoint-aware-0017fabb.js';
import { E as ElementIdGenerator } from './element-id-handling-bb37154e.js';
import { o as onNextFrame, i as isSlotDefined, a as isSlotUndefined, g as getDefinedSlots } from './render-util-1dee6655.js';
import { S as SlotObserver } from './slot-observer-9196738e.js';
import { H as HeaderService, s as subscribeToHeaderHeight } from './header-service-d157d41e.js';
import './value-store-f4663db1.js';
import './query-util-02753b43.js';
import './multicast-observable-f2e53ec0.js';

function isValueInEnum(enumeration, value) {
  return Object.values(enumeration).includes(value);
}

const dxHeaderCss = ":host{--dx-header-hide-animation-duration:300ms;display:block;position:sticky;z-index:900;left:0;width:100%;transition:top 300ms ease-in-out;--dx-text-color-transition:color 300ms ease-in-out}:host:host([hidden]){display:none !important}:host .header-bar{height:56px;background-color:var(--dx-header-background-color, #ffffff);transition:background-color 300ms ease-in-out;--dx-icon-transition-duration:300ms}:host .header-bar .content{max-width:1460px;padding:0 16px;box-sizing:border-box;height:100%;display:flex;align-items:center;justify-content:space-between}@media (min-width: 1460px){:host .header-bar .content{margin:auto}}@media (min-width: 476px){:host .header-bar .content{padding:0 24px}}@media (min-width: 760px){:host .header-bar .content{padding:0 32px}}@media (min-width: 1280px){:host .header-bar .content{padding:0 40px}}:host .header-bar .content .application-logo{width:var(--dx-button-width, fit-content);width:var(--dx-button-width, -moz-fit-content);cursor:pointer;display:inline-flex;justify-content:center;align-items:center;box-sizing:border-box;white-space:nowrap;user-select:none;-webkit-user-select:none;border:none;margin:0;background-color:transparent;transition:background-color 150ms ease-in-out, border-color 150ms ease-in-out;outline-offset:2px;height:max-content;flex-shrink:0;flex-grow:0;padding:0;margin-right:12px}:host .header-bar .content .application-logo[disabled]{cursor:unset}:host .header-bar .content .application-logo.icon-right{flex-direction:row-reverse}:host .header-bar .content .application-logo:focus-visible{outline:2px solid #004763}:host .header-bar .content .application-logo.text{align-self:center}:host .header-bar .content .application-logo.image{align-self:flex-start;outline-offset:-2px}:host .header-bar .content .application-logo .logo{height:max-content}@media (min-width: 760px){:host .header-bar .content .application-logo{margin-right:24px}}:host .header-bar .content .application-name-container{flex:1;overflow:hidden;height:100%}:host .announcement-grow-box{position:relative;z-index:1002}:host .spacer{background-color:#ededed}:host .info-wrapper{background-color:var(--dx-header-info-container-color, #f7f7f7);transition:background-color 300ms ease-in-out}:host .info-wrapper .info-container{--dx-page-padding:16px;max-width:1460px;display:flex;box-sizing:border-box;gap:0;flex-direction:column-reverse;align-items:center;padding:8px var(--dx-page-padding) 0 var(--dx-page-padding)}@media (min-width: 476px){:host .info-wrapper .info-container{--dx-page-padding:24px}}@media (min-width: 760px){:host .info-wrapper .info-container{--dx-page-padding:32px}}@media (min-width: 1280px){:host .info-wrapper .info-container{--dx-page-padding:40px}}@media (min-width: 1460px){:host .info-wrapper .info-container{margin:auto}}:host .info-wrapper .info-container .breadcrumb-container{width:100%}:host .info-wrapper .info-container .representative-container{margin-left:auto;flex-shrink:0}@media (min-width: 760px){:host .info-wrapper .info-container{justify-content:space-between;flex-direction:row;gap:32px;padding-top:0;padding-left:calc(\n          var(--dx-page-padding) + 64px + 24px\n        );padding-right:var(--dx-page-padding)}:host .info-wrapper .info-container.text-logo{padding-left:var(--dx-page-padding)}:host .info-wrapper .info-container .breadcrumb-container{flex-shrink:1;width:100%;overflow:hidden}:host .info-wrapper .info-container .representative-container{flex-shrink:0;margin-left:auto;margin-top:0;margin-bottom:0}}@media (min-width: 1280px){:host .info-wrapper .info-container.text-logo{padding-left:var(--dx-page-padding)}}:host(.show-shadow){box-shadow:0 0 8px rgba(0, 0, 0, 0.16)}:host(.golden-theme){--dx-logo-fill-color:#ffffff;--dx-header-background-color:#c8aa22;--dx-header-info-container-color:#c8aa22;--dx-header-text-icon-theme-color:#ffffff;--dx-header-text-icon-theme-hover-color:#ffffff;--dx-header-text-icon-theme-active-color:#ffffff;--dx-header-menu-link-text-icon-theme-color:#ffffff;--dx-header-breadcrumb-text-icon-hover-color:#ffffff;--dx-header-breadcrumb-text-icon-active-color:#ffffff;--dx-header-breadcrumb-background-hover-color:rgba(0, 0, 0, 0.1);--dx-header-breadcrumb-background-active-color:rgba(0, 0, 0, 0.2)}:host(.golden-theme):host(:not(.sidebar-open)){--dx-header-menu-item-selected-state-line-color:#ffffff;--dx-header-menu-item-hover-color:#b4991f;--dx-header-menu-item-active-color:#a0881b;--dx-header-menu-item-text-icon-theme-color:#ffffff;--dx-notification-indicator-background-color:#c8aa22}";

const DxHeader = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.navigateHome = createEvent(this, "navigateHome", 7);
    this.lastScrollPositionY = 0;
    this.elevated = false;
    this.hideHeaderBar = false;
    this.sidebarOpen = false;
    /**
     * The theme of the header defines the color palette which is used for it's sub-elements.
     *
     * @value white: The default theme displays the header with a white background.
     * @value gold: The golden theme which is implemented for "meine.dvag".
     * @warning This property must not be used by other projects than "meine.dvag".
     */
    this.theme = "white";
    /**
     * The logo which is displayed in the left corner of the header.
     *
     * @value text: Brand Logo with the company's name. The `applicationname` will not be displayed.
     * @value image: Brand Logo without a name. The `applicationname` will be displayed next to the logo.
     */
    this.logo = "image";
    /**
     * Defines the brand appearance of header internal branding elements (e.g. the logo image).
     *
     * @value default: The default brand of the DVAG is used.
     */
    this.brand = "default";
    /**
     * Whether to show the current announcement bar or not.
     *
     * @value true: Shows an announcement in case the "announcement-bar" slot is defined.
     * @value false: Hides the announcement.
     */
    this.showannouncement = true;
    this.allow = false;
  }
  onThemeChanged() {
    if (this.theme === "white") {
      this.internalTheme = "white";
    }
    this.setupIntersectionObserver();
  }
  onShowAnnouncementChanged() {
    if (this.showannouncement) {
      // Show header bar when announcement is shown
      this.hideHeaderBar = false;
    }
  }
  componentDidLoad() {
    onNextFrame(() => {
      this.updateMenuBar();
    });
  }
  connectedCallback() {
    this.breakpointAware = new BreakpointAware(() => forceUpdate(this), true);
    const headerService = HeaderService.instance();
    const currentRegisteredHeader = headerService.header;
    assert(currentRegisteredHeader == null || currentRegisteredHeader === this.self, () => 'There is already a <dx-header> element registered.\n\n"There can be only one".\n- Connor MacLeod', this.self);
    headerService.header = this.self;
    headerService.isVisible = true;
    this.headerHeightSubscription = subscribeToHeaderHeight(() => {
      forceUpdate(this);
    });
    this.initializeSlotObserver();
    this.setupIntersectionObserver();
  }
  disconnectedCallback() {
    var _a, _b, _c, _d, _e;
    (_a = this.breakpointAware) === null || _a === void 0 ? void 0 : _a.disconnect();
    (_b = this.slotObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
    (_c = this.infoContainerSlotObserver) === null || _c === void 0 ? void 0 : _c.disconnect();
    (_d = this.intersectionObserver) === null || _d === void 0 ? void 0 : _d.disconnect();
    (_e = this.headerHeightSubscription) === null || _e === void 0 ? void 0 : _e.unsubscribe();
    HeaderService.instance().unsetHeaderElement(this.self);
  }
  onScroll() {
    const currentY = window.scrollY;
    const headerService = HeaderService.instance();
    if (headerService.isScrollInteractionEnabled) {
      this.hideHeaderBar =
        this.lastScrollPositionY >= 0 && this.lastScrollPositionY < currentY;
      headerService.isVisible = !this.hideHeaderBar;
    }
    // We want to set the elevation to true, when the scroll position exceeds the header height.
    // We want to reset the elevation when the scroll position reaches zero again.
    this.elevated =
      (this.elevated && currentY > 0) || currentY > headerService.headerHeight;
    if (!this.elevated) {
      this.internalTheme = this.theme;
    }
    this.lastScrollPositionY = currentY;
  }
  onResize() {
    this.updateMenuBar();
  }
  initializeSlotObserver() {
    this.slotObserver = new SlotObserver(this.self, ":scope > *:not([slot])", (nodes) => {
      assert(nodes.length === 0, () => "There should be no default slotted elements within a <dx-header> element. Maybe you forgot to set the 'menu' or '${value}' slot.", "\n\nAffected element:", this.self, "\n\nDefault slotted nodes:\n", ...nodes.map((node) => [node, "\n"]).flat());
    });
    this.infoContainerSlotObserver = new SlotObserver(this.self, `*[slot=${PREDEFINED_HEADER_SLOTS.BREADCRUMB}],*[slot=${PREDEFINED_HEADER_SLOTS.REPRESENTATIVE}]`, () => {
      forceUpdate(this.self);
    });
  }
  setupIntersectionObserver() {
    var _a;
    if (this.theme === "gold") {
      this.intersectionObserver = new IntersectionObserver(() => {
        this.internalTheme = this.elevated ? "white" : "gold";
      });
      this.intersectionObserver.observe(this.self);
    }
    else {
      (_a = this.intersectionObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
      this.intersectionObserver = null;
    }
  }
  updateMenuBar() {
    var _a, _b;
    if (this.menuBar == null) {
      return;
    }
    const contentWidth = (_b = (_a = this.contentElement) === null || _a === void 0 ? void 0 : _a.offsetWidth) !== null && _b !== void 0 ? _b : 0;
    const applicationInfoWidth = this.logoElement.offsetWidth + this.getMinimalApplicationNameWidth();
    this.menuBar
      .updateChildren(contentWidth -
      (applicationInfoWidth +
        MARGIN_BIAS_BETWEEN_APPLICATION_INFO_AND_MENU_BAR))
      .then();
  }
  getMinimalApplicationNameWidth() {
    var _a;
    return this.logo === "image" && ((_a = this.appname) === null || _a === void 0 ? void 0 : _a.length) > 0
      ? MIN_APPLICATION_NAME_WIDTH
      : 0;
  }
  getLogo() {
    if (this.logo === "image") {
      return "bildmarke";
    }
    else if (this.logo === "text") {
      return "wortmarke";
    }
  }
  obtainMenuBar(menuBar) {
    if (this.menuBar !== menuBar) {
      this.menuBar = menuBar;
      onNextFrame(() => {
        this.updateMenuBar();
      });
    }
  }
  renderApplicationName() {
    if (this.logo === "text") {
      return;
    }
    return (h("div", { class: "application-name-container" }, h("dx-header-application-name", { class: "application-name", appname: this.appname, onInteraction: (e) => {
        e.stopPropagation();
        this.navigateHome.emit();
      } })));
  }
  renderInfoContainer() {
    const hasBreadcrumb = isSlotDefined(this.self, PREDEFINED_HEADER_SLOTS.BREADCRUMB);
    const hasRepresentative = isSlotDefined(this.self, PREDEFINED_HEADER_SLOTS.REPRESENTATIVE);
    if (!hasBreadcrumb && !hasRepresentative) {
      return;
    }
    return (h("div", { class: {
        "info-wrapper": true,
      } }, h("div", { class: {
        "info-container": true,
        "text-logo": this.logo === "text",
      } }, hasBreadcrumb ? (h("div", { class: { "breadcrumb-container": true } }, h("slot", { name: `${PREDEFINED_HEADER_SLOTS.BREADCRUMB}` }))) : null, hasRepresentative ? (h("div", { class: "representative-container" }, h("slot", { name: `${PREDEFINED_HEADER_SLOTS.REPRESENTATIVE}` }))) : null)));
  }
  renderAnnouncement() {
    if (isSlotUndefined(this.self, "announcement-bar")) {
      return;
    }
    return [
      h("dx-simple-grow-box", { class: "announcement-grow-box", open: this.showannouncement }, h("slot", { name: "announcement-bar" })),
      this.showannouncement ? h("dx-spacer", { class: "spacer", mq1: "4v" }) : null,
    ];
  }
  getMaxTop() {
    const logoOverflow = this.logo === "image" ? 8 : 0;
    return HeaderService.instance().headerHeight + logoOverflow;
  }
  render() {
    var _a, _b, _c, _d, _e;
    const definedValueSlots = getDefinedSlots(this.self).filter((slot) => !isValueInEnum(PREDEFINED_HEADER_SLOTS, slot));
    // We need to offset the top at least 1px to actually get an intersection with the window
    const intersectionBias = 1;
    return (h(Host, { class: {
        "golden-theme": this.internalTheme === "gold",
        "show-shadow": this.elevated,
        "sidebar-open": this.sidebarOpen,
      }, role: "banner", style: {
        top: this.hideHeaderBar
          ? `-${this.getMaxTop() + intersectionBias}px`
          : "0",
        "--dx-header-height": `${(_a = HeaderService.instance().headerHeight) !== null && _a !== void 0 ? _a : 0}px`,
        "--dx-header-height-below": `${(_c = (_b = this.self.shadowRoot.querySelector(".info-container")) === null || _b === void 0 ? void 0 : _b.offsetHeight) !== null && _c !== void 0 ? _c : 0}px`,
        "--dx-header-height-top": `${(_e = (_d = this.self.shadowRoot.querySelector("dx-simple-grow-box")) === null || _d === void 0 ? void 0 : _d.offsetHeight) !== null && _e !== void 0 ? _e : 0}px`, // TODO: Search better solution after release
      } }, this.renderAnnouncement(), h("div", { class: "header-bar" }, h("div", { class: {
        content: true,
      }, ref: (ref) => (this.contentElement = ref) }, h("button", { class: {
        "application-logo": true,
        text: this.logo === "text",
        image: this.logo === "image",
      }, "aria-label": "Home", id: ElementIdGenerator.createId(this.self, "logo-control"), onClick: () => this.navigateHome.emit() }, h("dx-logo", { class: "logo", type: this.getLogo(), ref: (ref) => (this.logoElement = ref) })), this.renderApplicationName(), h("dx-header-menu-bar", { ref: (ref) => this.obtainMenuBar(ref), onSidebarOpenChange: (e) => (this.sidebarOpen = e.detail) }, h("slot", { name: "menu" }), h("slot", { slot: "navigation-header", name: "navigation-header" }), h("slot", { slot: "navigation-footer", name: "navigation-footer" }), definedValueSlots.map((slot) => {
      return h("slot", { name: slot, slot: slot });
    })))), this.renderInfoContainer()));
  }
  get self() { return getElement(this); }
  static get watchers() { return {
    "theme": ["onThemeChanged"],
    "showannouncement": ["onShowAnnouncementChanged"]
  }; }
};
const MIN_APPLICATION_NAME_WIDTH = 176;
var PREDEFINED_HEADER_SLOTS;
(function (PREDEFINED_HEADER_SLOTS) {
  PREDEFINED_HEADER_SLOTS["MENU"] = "menu";
  PREDEFINED_HEADER_SLOTS["BREADCRUMB"] = "breadcrumb";
  PREDEFINED_HEADER_SLOTS["REPRESENTATIVE"] = "representative";
})(PREDEFINED_HEADER_SLOTS || (PREDEFINED_HEADER_SLOTS = {}));
const MARGIN_BIAS_BETWEEN_APPLICATION_INFO_AND_MENU_BAR = 8;
DxHeader.style = dxHeaderCss;

export { DxHeader as dx_header };
