// @dvag/design-system/dist/esm/dx-header-user.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './index-04c470ea.js';
import { v as verifyElementId, E as ElementIdGenerator } from './element-id-handling-bb37154e.js';
import { g as generateInitialsForName } from './misc-util-d3b69f2c.js';
import './value-store-f4663db1.js';

const dxHeaderUserCss = ":host{display:block;--dx-notification-indicator-top:2px;--dx-notification-indicator-right:0px;--dx-notification-indicator-display-number-top:-2px;--dx-notification-indicator-display-number-right:-3px;--dx-notification-indicator-display-big-number-top:-2px;--dx-notification-indicator-display-big-number-right:-5.5px}:host:host([hidden]){display:none !important}";

const DxHeaderUser = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.signOut = createEvent(this, "signOut", 7);
    this.accountInteraction = createEvent(this, "accountInteraction", 7);
    this.menuInteraction = createEvent(this, "menuInteraction", 7);
    this.internalForceSidebarContentUpdate = createEvent(this, "internalForceSidebarContentUpdate", 7);
    /**
     * The number of notifications that are visible inside of the indicator on this item.
     * @value number: The number of notifications that are visible inside of the indicator. The number has to be greater than zero.
     * @value "no-number": Only a red circle without a number is displayed as an indicator.
     *
     * @example notifications
     */
    this.notifications = undefined;
    /**
     * Specifies the theme for the displayed notifications.
     * The "system" theme is used when a notification emerges from the system like a notification bell.
     * The "user" theme is used as a result of an user action, like adding something to the shopping cart.
     *
     * @value system: Red badge with white border and white text.
     * @value user: Gray badge with white border and black text.
     * @example notifications-theme
     */
    this.notificationstheme = "system";
    /**
     * @internal
     */
    this.internalactive = false;
    /**
     * @internal
     */
    this.internalopen = true;
    /**
     * Defines whether the account button "Benutzerkonto" is displayed or not.
     *
     * @value true: The account button is visible.
     * @value false: The account button is not visible.
     * @example account-available
     */
    this.accountavailable = true;
  }
  connectedCallback() {
    verifyElementId(this.self);
  }
  componentShouldUpdate(_, __, propName) {
    if (propName === "imageurl" || propName === "name") {
      this.internalForceSidebarContentUpdate.emit();
    }
  }
  /**
   * @internal
   */
  async renderSidebarContent() {
    return (h("dx-header-user-sidebar-content", { id: ElementIdGenerator.createId(this.self, "sidebar-content"), name: this.name, imageurl: this.imageurl, accountavailable: this.accountavailable, onSignOut: () => this.signOut.emit(), onAccountInteraction: () => this.accountInteraction.emit() }));
  }
  /**
   * @internal
   */
  async getDefaultWidth() {
    return 48;
  }
  render() {
    return (h("dx-horizontal-grow-box", { open: this.internalopen }, h("dx-header-menu-button", { id: ElementIdGenerator.createId(this.self, "button"), active: this.internalactive, notifications: this.notifications, notificationstheme: this.notificationstheme, onInteraction: () => {
        this.menuInteraction.emit();
      } }, h("dx-user-avatar", { class: "avatar-circle", size: "xs", imageurl: this.imageurl, initials: generateInitialsForName(this.name) }))));
  }
  get self() { return getElement(this); }
};
DxHeaderUser.style = dxHeaderUserCss;

export { DxHeaderUser as dx_header_user };
