// @dvag/design-system/dist/esm/dx-legend.entry.js
import { r as registerInstance, h, H as Host } from './index-04c470ea.js';
import { N as NewKiSyncIcon } from './new-ki-snyc-icon-d9a521a0.js';

const dxLegendCss = ":host{display:block}:host:host([hidden]){display:none !important}:host .text{display:flex;align-items:center}:host .text:not(:last-child){margin-bottom:8px}:host .required-icon{margin:0 4px}";

const DxLegend = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    /**
     * Indicates if the required hint is shown.
     * @value false: Required hint is not shown.
     * @value true: Required hint is shown.
     */
    this.requiredhint = true;
    /**
     * Indicates if the KI-Sync hint is shown.
     * @value true: KI-Sync hint is shown.
     * @value false: KI-Sync hint is not shown.
     */
    this.kisynchint = false;
  }
  renderRequiredhint() {
    if (!this.requiredhint) {
      return;
    }
    return (h("dx-text", { class: "text", type: "its", color: "headline" }, h("dx-icon", { class: "required-icon", size: 16, icon: "state-required" }), "markiert Pflichtfeld."));
  }
  renderKiSyncedHint() {
    if (!this.kisynchint) {
      return;
    }
    return (h("dx-text", { class: "text", type: "its", color: "headline" }, h(NewKiSyncIcon, null), "synchronisiert system\u00FCbergreifend."));
  }
  render() {
    return (h(Host, null, this.renderRequiredhint(), this.renderKiSyncedHint()));
  }
};
DxLegend.style = dxLegendCss;

export { DxLegend as dx_legend };
