// @dvag/design-system/dist/esm/dx-wizard-step.entry.js
import { r as registerInstance, c as createEvent, g as getElement } from './index-04c470ea.js';
import { a as assert } from './assert-7fc7463a.js';
import { C as CONTENT_SLOT_NAME } from './wizard.util-56780438.js';
import { v as verifyElementId } from './element-id-handling-bb37154e.js';
import './value-store-f4663db1.js';

const DxWizardStep = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.finishedChange = createEvent(this, "finishedChange", 7);
    this.interaction = createEvent(this, "interaction", 7);
    /**
     * Whether this step is already finished or not.
     *
     * @value true: The step is finished. If all previous steps are finished as well the wizard automatically switches to the next page.
     * @value false: The step is not finished yet.
     *
     * @example finished
     */
    this.finished = false;
  }
  onFinishedChanged() {
    this.finishedChange.emit(this.finished);
  }
  connectedCallback() {
    verifyElementId(this.self);
    assert(this.value !== CONTENT_SLOT_NAME, () => `The value "content" should not be used because it is already reserved for the "content" slot of the <dx-wizard> element.`, this.self);
  }
  /**
   * @internal
   */
  async toWizardStep() {
    return {
      label: this.label,
      value: this.value,
      finished: this.finished,
      interact: () => this.interaction.emit(),
      metaStepComponent: this.self,
    };
  }
  get self() { return getElement(this); }
  static get watchers() { return {
    "finished": ["onFinishedChanged"]
  }; }
};

export { DxWizardStep as dx_wizard_step };
