// @dvag/design-system/dist/esm/dx-yes-no.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-04c470ea.js';
import { c as assertPropertyValue } from './assert-7fc7463a.js';
import { E as ElementIdGenerator } from './element-id-handling-bb37154e.js';
import './value-store-f4663db1.js';

const dxYesNoCss = ":host{display:block}:host:host([hidden]){display:none !important}.button-container{display:flex;flex-basis:50%}.button-container .button-yes-no{width:var(--dx-button-width, fit-content);width:var(--dx-button-width, -moz-fit-content);cursor:pointer;display:inline-flex;justify-content:center;align-items:center;box-sizing:border-box;white-space:nowrap;user-select:none;-webkit-user-select:none;border:none;margin:0;background-color:transparent;transition:background-color 150ms ease-in-out, border-color 150ms ease-in-out;outline-offset:2px;width:var(--dx-button-width, fit-content);width:var(--dx-button-width, -moz-fit-content);cursor:pointer;display:inline-flex;justify-content:center;align-items:center;box-sizing:border-box;white-space:nowrap;user-select:none;-webkit-user-select:none;border:none;margin:0;background-color:transparent;transition:background-color 150ms ease-in-out, border-color 150ms ease-in-out;display:flex;flex:1;height:40px;padding:0;--dx-text-color:#575757}.button-container .button-yes-no[disabled]{cursor:unset}.button-container .button-yes-no.icon-right{flex-direction:row-reverse}.button-container .button-yes-no:focus-visible{outline:2px solid #004763}.button-container .button-yes-no[disabled]{cursor:unset}.button-container .button-yes-no.medium{height:48px}.button-container .button-yes-no:first-child{margin-right:16px}@media not all and (pointer: coarse){.button-container .button-yes-no:hover{background-color:#e5eef2;border-color:#00587c;--dx-text-color:#00587c}}.button-container .button-yes-no:active{border-color:#004763;--dx-text-color:#004763}.button-container .button-yes-no.selected{background-color:#ccdee5;--dx-text-color:#2b2b2b}.button-container .button-yes-no.selected .inner-container{border-color:#337a96;border-width:2px}@media not all and (pointer: coarse){.button-container .button-yes-no.selected:hover .inner-container{border-color:#00587c}}.button-container .button-yes-no.selected:active .inner-container{border-color:#004763}.button-container .button-yes-no .inner-container{display:flex;border:1px solid #337a96;justify-content:center;align-items:center;height:100%;flex:1;box-sizing:border-box}.button-container.disabled{opacity:0.4;pointer-events:none}";

const DxYesNo = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.valueChange = createEvent(this, "valueChange", 7);
    /**
     * The value of this component.
     * @important
     * @value yes: yes button is selected
     * @value no: no button is selected.
     * @value unset: Nothing is selected.
     * @example value
     */
    this.value = "yes";
    /**
     * Defines if the buttons are disabled or not.
     * @value true: Buttons are disabled
     * @value false: Buttons are not disabled
     * @example disabled
     */
    this.disabled = false;
    /**
     * Defines the height of the buttons.
     *
     * @value s: Buttons are small.
     * @value m: Buttons are a bit bigger.
     * @example size
     */
    this.size = "s";
  }
  setValue(value) {
    if (this.value === value) {
      this.value = "unset";
    }
    else {
      this.value = value;
    }
    this.valueChange.emit(this.value);
  }
  getButtonTextByType(value) {
    if (value === "yes") {
      return "Ja";
    }
    else if (value === "no") {
      return "Nein";
    }
  }
  renderButton(buttonValue) {
    return (h("button", { id: ElementIdGenerator.createId(this.self, `${buttonValue}-control`), disabled: this.disabled, class: {
        "button-yes-no": true,
        small: this.size === "s",
        medium: this.size === "m",
        selected: this.value === buttonValue,
      }, onClick: () => {
        this.setValue(buttonValue);
      } }, h("div", { class: "inner-container" }, h("dx-text", { type: "ps" }, this.getButtonTextByType(buttonValue)))));
  }
  render() {
    assertPropertyValue("dx-yes-no", "size", this.size, ["s", "m"]);
    return (h(Host, null, h("dx-form-wrapper", { errormessage: this.errormessage }, h("div", { class: { "button-container": true, disabled: this.disabled } }, this.renderButton("yes"), this.renderButton("no")), h("slot", { name: "error-message", slot: "error-message" }))));
  }
  get self() { return getElement(this); }
};
DxYesNo.style = dxYesNoCss;

export { DxYesNo as dx_yes_no };
