// @dvag/design-system/dist/esm/render-util-1dee6655.js
function onNextFrame(fn) {
  requestAnimationFrame(() => requestAnimationFrame(() => fn()));
}
function getSlottedElements(element) {
  return Array.from(element.querySelectorAll("[slot]"));
}
function getDefinedSlots(element) {
  return getSlottedElements(element).map((element) => element.slot);
}
function isSlotDefined(element, slotName) {
  return !!element.querySelector(`[slot="${slotName}"]`);
}
function isSlotUndefined(element, slotName) {
  return !isSlotDefined(element, slotName);
}
// Color calculations
function toHex(dec) {
  return Math.round(dec).toString(16);
}
function lighten(hex, factor) {
  const lightenColor = lightenRgb(toRgb(hex), factor);
  return rgbToHex(lightenColor);
}
function lightenRgb(rgb, factor) {
  return rgb.map((value) => 255 - factor * (255 - value));
}
function rgbToHex(rgb) {
  return "#" + rgb.map((val) => toHex(val)).join("");
}
function toRgb(hex) {
  const hexValues = hex.replace("#", "").match(/.{1,2}/g);
  return hexValues.map((hexValue) => parseInt(hexValue, 16));
}

export { isSlotUndefined as a, getDefinedSlots as g, isSlotDefined as i, lighten as l, onNextFrame as o, toHex as t };
