// @dvag/design-system/dist/collection/components/internal/ds-input-label/ds-input-label.js
import { h, Host } from "@stencil/core";
import { kiSyncIcon } from "../../functional/ki-sync-icon";
/**
 * This Component is considered internal!
 * Do not use this component on your own!
 * This component can and will be changed without a warning!
 * @internal
 */
export class DsInputLabel {
  constructor() {
    /**
     * If the label should be in an disabled state.
     */
    this.disabled = false;
    /**
     * Aligns the label to the right. USed for ds-date-input
     */
    this.alignRight = false;
  }
  render() {
    var _a;
    return (h(Host, { class: { visible: ((_a = this.label) === null || _a === void 0 ? void 0 : _a.length) > 0 || this.kisynced } }, h("div", { class: { "label-icon-container": true, "ki-synced": this.kisynced, "align-right": this.alignRight } }, h("ds-text", { class: { disabled: this.disabled } }, this.label), h("div", { class: "non-clickable-container", onClick: (e) => {
        e.preventDefault();
      } }), this.renderSyncIcon())));
  }
  renderSyncIcon() {
    if (this.kisynced) {
      return kiSyncIcon();
    }
    return null;
  }
  static get is() { return "ds-input-label"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-input-label.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-input-label.css"]
    };
  }
  static get properties() {
    return {
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "The text that will be showed on top of a input."
        },
        "attribute": "label",
        "reflect": false
      },
      "kisynced": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Defines if the sync icon should be displayed."
        },
        "attribute": "kisynced",
        "reflect": false
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "If the label should be in an disabled state."
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "alignRight": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Aligns the label to the right. USed for ds-date-input"
        },
        "attribute": "align-right",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
