// @dvag/design-system/dist/collection/new/internal/dx-notification-indicator/dx-notification-indicator.js
import { h, Host } from "@stencil/core";
import { formatNotificationIndicatorType, isValidNotificationIndicatorType, isValidNotificationIndicatorTypeNumber, } from "./notification-indicator.util";
/**
 * @internal
 */
export class DxNotificationIndicator {
  constructor() {
    /**
     * Specifies the theme for the displayed notifications.
     * The "system" theme is used when a notification emerges from the system like a notification bell.
     * The "user" theme is used as a result of an user action, like adding something to the shopping cart.
     *
     * @value system: Red badge with white border and white text.
     * @value user: Gray badge with white border and black text.
     */
    this.theme = "system";
  }
  render() {
    return (h(Host, null, h("slot", null), h("div", { class: {
        "notification-indicator": true,
        "theme-system": this.theme === "system",
        "theme-user": this.theme === "user",
        "displays-notification-number": isValidNotificationIndicatorTypeNumber(this.notifications),
        "two-digits": this.notifications > 9,
        show: isValidNotificationIndicatorType(this.notifications),
      } }, formatNotificationIndicatorType(this.notifications))));
  }
  static get is() { return "dx-notification-indicator"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-notification-indicator.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-notification-indicator.css"]
    };
  }
  static get properties() {
    return {
      "notifications": {
        "type": "any",
        "mutable": false,
        "complexType": {
          "original": "NotificationIndicatorType",
          "resolved": "\"no-number\" | number",
          "references": {
            "NotificationIndicatorType": {
              "location": "import",
              "path": "./notification-indicator.util"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": ""
        },
        "attribute": "notifications",
        "reflect": false
      },
      "theme": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "NotificationIndicatorThemeType",
          "resolved": "\"system\" | \"user\"",
          "references": {
            "NotificationIndicatorThemeType": {
              "location": "import",
              "path": "./notification-indicator.util"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "system: Red badge with white border and white text."
            }, {
              "name": "value",
              "text": "user: Gray badge with white border and black text."
            }],
          "text": "Specifies the theme for the displayed notifications.\nThe \"system\" theme is used when a notification emerges from the system like a notification bell.\nThe \"user\" theme is used as a result of an user action, like adding something to the shopping cart."
        },
        "attribute": "theme",
        "reflect": false,
        "defaultValue": "\"system\""
      }
    };
  }
}
