// @dvag/design-system/dist/collection/new/scaffold/dx-fd-outline-box/dx-fd-outline-box.js
import { h, Host } from "@stencil/core";
import { ColorPalette } from "../../../classes/color-palette";
/**
 * An outlined box with a five degree angle at the bottom which displays a content.
 *
 * @path /Scaffold/Stage
 * @name Outlined Golden Stage
 * @shortname Outlined
 * @stable
 *
 * @new
 */
export class DxFdOutlineBox {
  constructor() {
    /**
     * The color of the outline.
     *
     * @value string: Name of the color.
     * @internal
     */
    this.color = "gold";
    /**
     * @internal
     */
    this.strokewidth = 3;
  }
  render() {
    const hexColor = new ColorPalette(true).getColorCode(this.color);
    const svgLineStyles = {
      "vector-effect": "non-scaling-stroke",
      "stroke-linejoin": "miter",
      "stroke-width": `${this.strokewidth}px`,
      stroke: hexColor,
      fill: "none",
      "shape-rendering": "geometricPrecision",
    };
    return (h(Host, null, h("div", { class: "content" }, h("slot", null), h("svg", { preserveAspectRatio: "none", viewBox: "0 0 100 100" }, h("line", { x1: 0, y1: 0, x2: 0, y2: 100, style: svgLineStyles }), h("line", { x1: 100, y1: 0, x2: 100, y2: 100, style: svgLineStyles }))), h("div", { class: "triangle" }, h("div", { class: "svg-container" }, h("svg", { preserveAspectRatio: "none", viewBox: "0 0 100 100" }, h("polyline", { points: "0,-5 0,100 100,0 100,-5", style: svgLineStyles }))))));
  }
  static get is() { return "dx-fd-outline-box"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-fd-outline-box.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-fd-outline-box.css"]
    };
  }
  static get properties() {
    return {
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Name of the color."
            }, {
              "name": "internal",
              "text": undefined
            }],
          "text": "The color of the outline."
        },
        "attribute": "color",
        "reflect": false,
        "defaultValue": "\"gold\""
      },
      "strokewidth": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "internal",
              "text": undefined
            }],
          "text": ""
        },
        "attribute": "strokewidth",
        "reflect": false,
        "defaultValue": "3"
      }
    };
  }
}
