// @dvag/design-system/dist/collection/new/scaffold/dx-fd-separator/dx-fd-separator.js
import { h, Host } from "@stencil/core";
import { ColorPalette } from "../../../classes/color-palette";
/**
 * Block which displays two colors separated by a five degree angle. It automatically keeps an aspect ratio to ensure
 * that the horizontal line has an angle of five degrees.
 *
 * @path /Scaffold
 * @name Five-Degree Separator
 * @shortname Separator
 * @stable
 * @new
 * @icon minus
 */
export class DxFdSeparator {
  render() {
    const colorPalette = new ColorPalette(true);
    const fromColorHex = colorPalette.getColorCode(this.fromcolor);
    const toColorHex = colorPalette.getColorCode(this.tocolor);
    return (h(Host, null, h("svg", { viewBox: "0 0 100 100", preserveAspectRatio: "none" }, h("polygon", { points: "0,-1 0,0 0,100 100,0 100,-1", style: {
        fill: fromColorHex,
        "shape-rendering": "geometricPrecision",
      } }), h("polygon", { points: "0,100 0,101 100,101 100,100 100,0", style: {
        fill: toColorHex,
        "shape-rendering": "geometricPrecision",
      } }))));
  }
  static get is() { return "dx-fd-separator"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-fd-separator.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-fd-separator.css"]
    };
  }
  static get properties() {
    return {
      "fromcolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Name of the color from the color palette."
            }],
          "text": "The background color of the top container which should be separated from a bottom container."
        },
        "attribute": "fromcolor",
        "reflect": false
      },
      "tocolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Name of the color from the color palette."
            }],
          "text": "The background color of the bottom container which should be separated from a top container."
        },
        "attribute": "tocolor",
        "reflect": false
      }
    };
  }
}
