// @dvag/design-system/dist/collection/new/form/dx-legend/dx-legend.js
import { h, Host } from "@stencil/core";
import { NewKiSyncIcon } from "../../functional/new-ki-snyc-icon";
/**
 * An element which displays a legend for a form.
 *
 * @name Legend
 * @path /Form/Other
 * @icon info
 *
 * @stable
 */
export class DxLegend {
  constructor() {
    /**
     * Indicates if the required hint is shown.
     * @value false: Required hint is not shown.
     * @value true: Required hint is shown.
     */
    this.requiredhint = true;
    /**
     * Indicates if the KI-Sync hint is shown.
     * @value true: KI-Sync hint is shown.
     * @value false: KI-Sync hint is not shown.
     */
    this.kisynchint = false;
  }
  renderRequiredhint() {
    if (!this.requiredhint) {
      return;
    }
    return (h("dx-text", { class: "text", type: "its", color: "headline" }, h("dx-icon", { class: "required-icon", size: 16, icon: "state-required" }), "markiert Pflichtfeld."));
  }
  renderKiSyncedHint() {
    if (!this.kisynchint) {
      return;
    }
    return (h("dx-text", { class: "text", type: "its", color: "headline" }, h(NewKiSyncIcon, null), "synchronisiert system\u00FCbergreifend."));
  }
  render() {
    return (h(Host, null, this.renderRequiredhint(), this.renderKiSyncedHint()));
  }
  static get is() { return "dx-legend"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-legend.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-legend.css"]
    };
  }
  static get properties() {
    return {
      "requiredhint": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Required hint is not shown."
            }, {
              "name": "value",
              "text": "true: Required hint is shown."
            }],
          "text": "Indicates if the required hint is shown."
        },
        "attribute": "requiredhint",
        "reflect": false,
        "defaultValue": "true"
      },
      "kisynchint": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: KI-Sync hint is shown."
            }, {
              "name": "value",
              "text": "false: KI-Sync hint is not shown."
            }],
          "text": "Indicates if the KI-Sync hint is shown."
        },
        "attribute": "kisynchint",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
}
