// @dvag/design-system/dist/collection/components/scaffold/ds-user-avatar/ds-user-avatar.js
import { h, Host } from "@stencil/core";
import { assert, assertPropertyValue } from "../../../classes/assert";
/**
 * Stellt einen runden Nutzer-Avatar dar. Innerhalb des Avatars ist ein Foto des
 * Nutzers zu sehen. Wird kein Foto hinterlegt, werden die Initialen des Nutzers
 * verwendet.
 * @name Nutzer-Avatar
 * @shortname Avatar
 * @icon user-tie
 * @path Scaffold
 */
export class DsUserAvatar {
  render() {
    assert(this.imageurl != null || this.initials != null, () => "<ds-user-avatar> Es wurde weder 'imageurl' noch 'initials' definiert. Eins von beiden muss gesetzt sein.");
    assertPropertyValue("ds-user-avatar", "size", this.size, [
      "header",
      "content-s",
      "content-m",
      "content-l",
      "content-xl",
    ]);
    return (h(Host, { class: {
        "size-header": this.size === "header",
        "size-content-s": this.size === "content-s",
        "size-content-m": this.size === "content-m",
        "size-content-l": this.size === "content-l",
        "size-content-xl": this.size === "content-xl",
      } }, this.imageurl != null ? this.renderImage() : this.renderInitials()));
  }
  renderImage() {
    return (h("div", { class: "image", style: {
        "background-image": `url(${this.imageurl})`,
      } }));
  }
  renderInitials() {
    var _a;
    return (h("div", { class: "initials" }, h("ds-text", null, (_a = this.initials) !== null && _a !== void 0 ? _a : "")));
  }
  static get is() { return "ds-user-avatar"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-user-avatar.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-user-avatar.css"]
    };
  }
  static get properties() {
    return {
      "imageurl": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: URL"
            }, {
              "name": "important",
              "text": undefined
            }, {
              "name": "example",
              "text": "image"
            }],
          "text": "URL zum Foto des Benutzers"
        },
        "attribute": "imageurl",
        "reflect": false
      },
      "initials": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Initialen, z.B. `\"FG\"`"
            }, {
              "name": "important",
              "text": undefined
            }, {
              "name": "example",
              "text": "initials"
            }],
          "text": "Initialen des Nutzers"
        },
        "attribute": "initials",
        "reflect": false
      },
      "size": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "UserAvatarSize",
          "resolved": "\"content-l\" | \"content-m\" | \"content-s\" | \"content-xl\" | \"header\"",
          "references": {
            "UserAvatarSize": {
              "location": "local"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "header: F\u00FCr die Nutzung im Header verwendete Gr\u00F6\u00DFe (32px)"
            }, {
              "name": "value",
              "text": "content-s: F\u00FCr die Nutzung innerhalb von Content (48px)"
            }, {
              "name": "value",
              "text": "content-m: F\u00FCr die Nutzung innerhalb von Content (64px)"
            }, {
              "name": "value",
              "text": "content-l: F\u00FCr die Nutzung innerhalb von Content (72px)"
            }, {
              "name": "value",
              "text": "content-xl: F\u00FCr die Nutzung innerhalb von Content (144px)"
            }],
          "text": "Gr\u00F6\u00DFe des Nutzer-Avatars"
        },
        "attribute": "size",
        "reflect": false
      }
    };
  }
}
