// @dvag/design-system/dist/collection/components/form/ds-dropdown/ds-dropdown.js
import { forceUpdate, h, Host, } from "@stencil/core";
import { ElementIdGenerator, verifyElementId, } from "../../../classes/element-id-handling";
import { createMetaComponentsSlotObserver, } from "../../../classes/slot-observer";
import { findDuplicatedValues } from "../../../classes/value";
import { assert } from "../../../classes/assert";
import { onNextFrame } from "../../../classes/render-util";
import { ReadonlyTextElement } from "../../functional/base-elements";
import { hookToUpdateProperty, isActiveElement, } from "../../../classes/misc-util";
import { InputContainer } from "../../functional/input-container";
import { ErrorPopup } from "../../functional/error-popup";
import { compareObjectsEqual } from "../../../classes/compare-objects";
/**
 * Form element for selecting multiple options in a list.
 *
 * @path /Form
 * @name Dropdown
 * @icon caret-square-down
 * @stable
 */
export class DsDropdown {
  constructor() {
    this.lastEmittedChangeValue = null;
    this.lastEmittedCompleteValue = null;
    this.preventReFocus = false;
    this.selectScanningEnabled = true;
    this.options = [];
    this.inputFocused = false;
    this.listOpen = false;
    this.internalValue = null;
    this.showMouseHoverEffectOnList = true;
    /**
     * Indicates that an error exist within the input field. Colors the border in red.
     * @value true: An error exists.
     * @value false: There is no error.
     * @example error
     */
    this.error = false;
    /**
     * Whether the selection is readonly or editable.
     * If true the value of this component will be visualized as a normal text.
     * @value false: Input is editable.
     * @value true: Input is read-only.
     *
     * @example readonly
     */
    this.readonly = false;
    /**
     * Marks this input field as a required form element.
     * @value false: Not required
     * @value true: It is necessary to select an option to fulfill the requirement.
     * @example required
     */
    this.required = false;
    /**
     * Placeholder text which is displayed when no option is currently selected.
     * @value string: Placeholder text
     * @example placeholder
     */
    this.placeholder = "Bitte wählen";
    /**
     * The current value of this element. It represents the `value` property of the current selected `<dx-dropdown-option>` or null.
     * Other by selecting an option with the mouse or keyboard, it is possible to set this property programmatically.
     *
     * @important
     * @value string: The string value of the selected option.
     * @value object: The value of the selected option in case it's a complex type.
     * @value null: No option selected.
     * @example value
     */
    this.value = null;
    /**
     * Defines the text alignment.
     * @value left: Text is left aligned.
     * @value right: Text is right aligned.
     * @example text-align
     */
    this.textalign = "left";
    /**
     * Defines the position of the dropdown list.
     * @value top: The list is displayed at the top of the input field.
     * @value bottom: The list is displayed at the bottom of the input field.
     * @example popup-position
     */
    this.popupposition = "bottom";
    /**
     * Whether it is possible to select an option which is not present in the current option list.
     *
     * @value false: The user can only select an option in the list.
     * @value true: The user is able to set a value which is not present in the option list by editing a text field under
     * the selection input field.
     * @example elseoption
     */
    this.elseoption = false;
    /**
     * The label of the button which activates the `elseoption` input field. The `elseoption` property has to be `true`.
     * @value string: Button label
     * @example elseoption
     */
    this.elseoptionbuttonlabel = "Sonstige Eingabe";
    /**
     * The label of the text input field which represents the `elseoption`. The `elseoption` property has to be `true`.
     * @value string: Text input label.
     * @example elseoption
     */
    this.elseoptioninputlabel = "Sonstige Eingabe";
    /**
     * Determines if the else-option input field is visible to the user. This property automatically changes
     * on user interaction or providing a value that doesn't match any dropdown options.
     * @value false: Else-option input field is not displayed.
     * @value true: Else-option input field is displayed.
     * @example elseoption-selected
     */
    this.elseoptionselected = false;
    /**
     * Defines whether the dropdown is enabled or not.
     * @value true: Dropdown is not editable.
     * @value false: Dropdown is editable.
     * @example disabled
     */
    this.disabled = false;
    /**
     * Indicates that the form is synced with the KI.
     * @value true: Synchronized.
     * @value false: Not synchronized.
     * @example kisynced
     */
    this.kisynced = false;
  }
  valueChanged() {
    this.obtainValueChangesFromOutside();
  }
  /**
   * Sets the focus on this element.
   *
   * @example focus
   */
  async focusControl() {
    this.inputElement.focus();
  }
  componentWillLoad() {
    verifyElementId(this.self);
    this.obtainValueChangesFromOutside(true);
  }
  connectedCallback() {
    this.optionsObserver =
      createMetaComponentsSlotObserver(this.self, "ds-dropdown-option", async (nodes) => {
        await this.collectOptions(nodes);
        this.findSelectedOption();
        const duplicates = findDuplicatedValues(this.options);
        assert(duplicates.length == 0, () => `Innerhalb eines DsDropdowns sollten kein Eintrag mit denselben values existieren!\nFolgende duplizierte Values wurden gefunden: ${duplicates.join()}`);
      });
  }
  disconnectedCallback() {
    var _a;
    (_a = this.optionsObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    clearInterval(this.observeSelectElementIntervalHandle);
  }
  componentDidRender() {
    if (this.listOpen && this.selectedOption != null) {
      this.scrollIntoView(this.selectedOption);
    }
  }
  /**
   * If the value changed from outside we need to check what option to use.
   * Possible cases can be:
   * - Setting value attribute
   * - Set of possible <ds-dropdown-option> changed
   * - Native <select> supervise detected a value change
   */
  findSelectedOption() {
    if (this.self.querySelectorAll("ds-dropdown-option").length !==
      this.options.length) {
      return; // can't decide now, cause not all option nodes are initialized
    }
    const optionMatchingValue = this.options.find((option) => option.value === this.value);
    if (!this.elseoptionselected && optionMatchingValue != null) {
      // Is there an option matching the value and our internal selectedOption is unset or set to a different option?
      if (this.selectedOption == null ||
        !compareObjectsEqual(optionMatchingValue, this.selectedOption)) {
        this.selectedOption = optionMatchingValue;
      }
    }
    else if (this.elseoption && this.internalValue !== null) {
      // Is elseoption enabled and therefore we don't need to match an option and use the value as elseoption value?
      this.elseoptionselected = true;
      this.selectedOption = null;
    }
    else {
      // We couldn't figure out an existing option => no option set
      this.selectedOption = null;
      this.elseoptionselected = false;
    }
  }
  async collectOptions(nodes) {
    const previousOptions = this.options;
    this.options = await Promise.all(nodes.map((node) => node.toOption()));
    if (this.options.length > 0 && previousOptions.length > 0) {
      this.findSelectedOption();
    }
  }
  // Always keep internalValue in sync with nativeSelect value. Otherwise there could be a race condition where a selection fails.
  updateInternalValue(newValue) {
    this.internalValue = newValue;
    if (this.nativeSelect) {
      this.nativeSelect.value = JSON.stringify(this.internalValue);
    }
  }
  obtainValueChangesFromOutside(force = false) {
    this.sanitizeValue();
    if (force || this.value !== this.internalValue) {
      this.updateInternalValue(this.value);
      this.lastEmittedChangeValue = this.value;
      this.lastEmittedCompleteValue = this.value;
      this.findSelectedOption();
    }
  }
  sanitizeValue() {
    if (this.value === undefined || this.value === NULL_STRING) {
      this.value = null;
    }
  }
  onInputFocused() {
    this.inputFocused = true;
  }
  onInputBlurred() {
    onNextFrame(() => {
      if (this.preventReFocus || !isActiveElement(this.self)) {
        this.inputFocused = false;
        this.setOpenListStatus(false);
        this.emitValueChange();
        this.emitValueComplete();
      }
      else {
        this.inputElement.focus();
      }
      this.preventReFocus = false;
    });
  }
  toggleList() {
    this.setOpenListStatus(!this.listOpen);
  }
  onInputKeyUp(event) {
    if (this.handleUpAndDownNavigation(event)) {
    }
    else if (["Enter", "Escape"].includes(event.code)) {
      this.setOpenListStatus(false);
      this.emitValueComplete();
    }
  }
  onInputKeyDown(event) {
    if (["ArrowDown", "ArrowUp"].includes(event.code)) {
      event.preventDefault();
    }
  }
  onInputClicked() {
    if (!this.disabled) {
      this.toggleList();
    }
  }
  setOpenListStatus(isOpen) {
    this.listOpen = isOpen;
    forceUpdate(this);
    if (this.listOpen) {
      this.inputElement.focus();
    }
  }
  /**
   * Used by the hidden native <select> to react to changes.
   */
  selectOptionsByValue(text) {
    if (text === NULL_STRING || text === "") {
      if (this.internalValue !== null) {
        this.selectedOption = null;
        this.updateInternalValue(null);
        this.emitValueChange();
      }
    }
    else {
      const optionWithValueMatchingText = this.options.find((option) => JSON.stringify(option.value) === text);
      if (optionWithValueMatchingText != null) {
        this.selectedOption = optionWithValueMatchingText;
        this.updateInternalValue(this.selectedOption.value);
        this.elseoptionselected = false;
        this.emitValueChange();
      }
      else {
        this.updateInternalValue(this.internalValue);
      }
    }
  }
  requestElseOption() {
    this.elseoptionselected = true;
    this.selectedOption = null;
    this.updateInternalValue("");
    onNextFrame(() => {
      var _a;
      (_a = this.elseOptionInput) === null || _a === void 0 ? void 0 : _a.focusControl();
    });
  }
  selectValueFromElseOption(value) {
    this.updateInternalValue(value);
    this.emitValueChange();
  }
  handleUpAndDownNavigation(event) {
    if (["ArrowDown", "ArrowUp"].includes(event.code)) {
      const selectedIndex = this.options.findIndex((option) => { var _a; return option.value === ((_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value); });
      if (event.code == "ArrowDown") {
        if (selectedIndex === -1 || selectedIndex === this.options.length - 1) {
          this.selectOption(this.options[0]);
        }
        else {
          this.selectOption(this.options[selectedIndex + 1]);
        }
      }
      if (event.code == "ArrowUp") {
        if (selectedIndex === -1 || selectedIndex === 0) {
          this.selectOption(this.options[this.options.length - 1]);
        }
        else {
          this.selectOption(this.options[selectedIndex - 1]);
        }
      }
      this.showMouseHoverEffectOnList = false;
      this.setOpenListStatus(true);
      event.preventDefault();
      event.stopImmediatePropagation();
      return true;
    }
    return false;
  }
  emitValueChange() {
    this.value = this.internalValue;
    if (this.lastEmittedChangeValue !== this.value) {
      this.valueChange.emit(this.value);
      this.lastEmittedChangeValue = this.value;
    }
  }
  emitValueComplete() {
    if (this.lastEmittedCompleteValue !== this.value) {
      this.valueComplete.emit(this.value);
      this.lastEmittedCompleteValue = this.value;
    }
  }
  selectOption(option) {
    this.selectedOption = option;
    this.elseoptionselected = false;
    this.updateInternalValue(this.selectedOption.value);
    this.emitValueChange();
    this.scrollIntoView(this.selectedOption);
  }
  scrollIntoView(optionToView) {
    if (this.listOpen) {
      onNextFrame(() => {
        const selectedIndex = this.options.findIndex((option) => option.value === (optionToView === null || optionToView === void 0 ? void 0 : optionToView.value));
        if (selectedIndex > -1) {
          const list = this.self.shadowRoot.querySelector(".item-list-options");
          if (list == null) {
            return;
          }
          const items = Array.from(this.self.shadowRoot.querySelectorAll(".item-list .item"));
          const listHeight = list.offsetHeight;
          const itemHeight = items[0].offsetHeight;
          list.scrollTop =
            itemHeight * selectedIndex - listHeight / 2 + itemHeight / 2;
        }
      });
    }
  }
  superviseSelectElement(el) {
    clearInterval(this.observeSelectElementIntervalHandle);
    this.nativeSelect = el;
    // When the value of the select changes we want to run the code below
    hookToUpdateProperty(el, "value", () => {
      var _a;
      const areAllOptionsScanned = () => el.querySelectorAll("option").length >= this.options.length + 1;
      if (!this.elseoptionselected &&
        this.selectScanningEnabled &&
        !this.readonly &&
        JSON.stringify((_a = this.internalValue) !== null && _a !== void 0 ? _a : "null") !== el.value &&
        areAllOptionsScanned()) {
        this.selectOptionsByValue(el.value);
      }
    });
  }
  valueMatchesOption() {
    var _a, _b;
    return (((_b = (_a = this.value) === null || _a === void 0 ? void 0 : _a.toString()) === null || _b === void 0 ? void 0 : _b.length) > 0 &&
      this.options.find((option) => option.value === this.value) != null);
  }
  render() {
    var _a, _b, _c, _d;
    if (this.readonly) {
      return (h(ReadonlyTextElement, { text: (_b = (_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.label) !== null && _b !== void 0 ? _b : "", label: this.label, kisynced: this.kisynced }));
    }
    const isErrorOutline = ((_c = this.errormessage) === null || _c === void 0 ? void 0 : _c.length) > 0 || this.error;
    const requirementFulfilled = this.valueMatchesOption() || this.elseoptionselected;
    return (h(Host, null, h("div", { class: {
        input: true,
        disabled: this.disabled,
      }, ref: (el) => (this.inputElement = el), id: ElementIdGenerator.createId(this.self, "user-interaction"), tabIndex: this.disabled ? -1 : 0, onKeyDown: (event) => this.onInputKeyDown(event), onKeyUp: (event) => this.onInputKeyUp(event), onFocus: () => this.onInputFocused(), onBlur: () => this.onInputBlurred(), onClick: () => this.onInputClicked() }, " ", h("ds-input-label", { disabled: this.disabled, label: this.label, kisynced: this.kisynced }), h("dx-overlay", { position: this.popupposition, offsety: 4, disabled: !this.listOpen }, h(InputContainer, { active: this.inputFocused, error: isErrorOutline, required: this.required && !this.elseoptionselected, requirementFulfilled: requirementFulfilled, disabled: this.disabled, title: (_d = this.selectedOption) === null || _d === void 0 ? void 0 : _d.label }, this.renderNativeSelect(), this.renderCurrentValue(), this.renderOpenListButton(), h(ErrorPopup, { text: this.errormessage, show: !this.listOpen &&
        !this.elseoptionselected &&
        this.inputFocused })), this.renderItemList())), h("ds-required-hint", { required: this.required, fulfilled: requirementFulfilled }), this.renderElseOptionInput()));
  }
  renderNativeSelect() {
    // The native select can be used for test automation tools to
    // change and read values from the select box.
    let nativeOptions = [
      { label: "-", value: NULL_STRING, icon: null },
      ...this.options,
    ];
    if (this.elseoptionselected) {
      nativeOptions = [
        {
          label: this.elseoptionbuttonlabel,
          value: "$$$ELSEOPTION$$$",
          icon: null,
        },
        ...nativeOptions,
      ];
    }
    return (h("select", { id: ElementIdGenerator.createId(this.self, "control"), ref: (el) => this.superviseSelectElement(el), tabIndex: -1, onChange: (e) => {
        const value = e.target.value;
        this.selectOptionsByValue(value);
      } }, nativeOptions.map((option) => {
      return (h("option", { selected: option.value === this.internalValue ||
          (this.internalValue === null && option.value === NULL_STRING), value: JSON.stringify(option.value) }, option.label));
    })));
  }
  renderOpenListButton() {
    return (h("ds-function-icon-button", { id: ElementIdGenerator.createId(this.self, "open-list-button"), unfocusable: true, class: {
        "open-list-button": true,
      } }, h("ds-icon", { icon: "drop-down", size: 16, class: {
        rotated: this.popupposition === "top",
        disabled: this.disabled === true,
      } })));
  }
  renderItemList() {
    if (this.listOpen) {
      return (h("div", { slot: "overlay", onMouseMove: () => (this.showMouseHoverEffectOnList = true), class: {
          "list-wrapper": true,
          "list-open": this.listOpen,
          "enable-mouse-hover": this.showMouseHoverEffectOnList,
        }, tabIndex: -1 }, h("div", { class: "item-list" }, h("div", { class: "item-list-options" }, this.options.map((option) => this.renderOption(option))), this.renderElseOptionButton())));
    }
  }
  renderElseOptionButton() {
    if (!this.elseoption || this.elseoptionselected) {
      return;
    }
    return (h("div", { class: "else-option-button", onClick: () => this.requestElseOption() }, h("ds-text-button", { unfocusable: true, id: ElementIdGenerator.createId(this.self, "else-option-button"), screensize: "small" }, h("ds-text", null, this.elseoptionbuttonlabel))));
  }
  renderOption(option, interactive = true, isPlaceholder = false) {
    var _a, _b;
    return (h("div", { id: !isPlaceholder
        ? ElementIdGenerator.createId(this.self, `item-list-value-${option.value}`)
        : undefined, class: {
        item: true,
        "is-placeholder": isPlaceholder,
        right: this.textalign == "right",
        selected: option.value == ((_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value),
        disabled: this.disabled === true,
      }, onClick: () => {
        if (interactive) {
          this.selectOption(option);
          this.inputElement.blur();
        }
      } }, ((_b = option.icon) === null || _b === void 0 ? void 0 : _b.length) > 0 ? (h("div", { class: { "icon-container": true } }, h("ds-icon", { size: 24, icon: option.icon }))) : null, h("div", { class: { "labels-container": true } }, h("ds-text", { key: option.label, type: "ft", screensize: "small" }, option.label), this.renderSubLabels(option, interactive, isPlaceholder))));
  }
  renderSubLabels(option, interactive, isPlaceholder) {
    // We don't want to render the sub labels in the dropdown container.
    if (isPlaceholder || !interactive) {
      return;
    }
    const subLabels = [];
    if (option.secondaryLabel) {
      subLabels.push(h("ds-text", { type: "it", screensize: "small" }, option.secondaryLabel));
    }
    if (option.tertiaryLabel) {
      subLabels.push(h("ds-text", { type: "itk", screensize: "small" }, option.tertiaryLabel));
    }
    return subLabels;
  }
  renderCurrentValue() {
    const className = "current-value";
    if (this.elseoptionselected) {
      return (h("div", { class: className }, this.renderOption({
        label: this.elseoptionbuttonlabel,
        value: "",
      }, false)));
    }
    if (this.selectedOption != null) {
      return (h("div", { class: className }, this.renderOption(this.selectedOption, false)));
    }
    return (h("div", { class: className }, this.renderOption({
      label: this.placeholder,
      icon: this.icon,
      value: "",
    }, false, true)));
  }
  renderElseOptionInput() {
    var _a;
    if (!this.elseoptionselected) {
      return;
    }
    return (h("div", { class: "else-option-input-container" }, h("ds-text-input", { ref: (el) => (this.elseOptionInput = el), id: ElementIdGenerator.createId(this.self, "else-option-input"), label: this.elseoptioninputlabel, value: this.internalValue, disabled: this.disabled, required: this.required, error: this.error || ((_a = this.errormessage) === null || _a === void 0 ? void 0 : _a.length) > 0, errormessage: this.errormessage, onFocus: () => {
        this.selectScanningEnabled = false;
        this.preventReFocus = true;
      }, onBlur: () => {
        this.selectScanningEnabled = true;
        this.preventReFocus = false;
      }, onValueChange: (event) => {
        this.selectValueFromElseOption(event.detail);
        event.stopPropagation();
      } })));
  }
  static get is() { return "ds-dropdown"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["ds-dropdown.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["ds-dropdown.css"]
    };
  }
  static get properties() {
    return {
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "important",
              "text": undefined
            }, {
              "name": "value",
              "text": "string: Label text e.g. `First name`"
            }, {
              "name": "example",
              "text": "value"
            }],
          "text": "Label which is displayed above the selection input field."
        },
        "attribute": "label",
        "reflect": false
      },
      "errormessage": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Error text."
            }, {
              "name": "example",
              "text": "error"
            }],
          "text": "Text, which is displayed in a red error popup at the bottom of the input field."
        },
        "attribute": "errormessage",
        "reflect": false
      },
      "error": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: An error exists."
            }, {
              "name": "value",
              "text": "false: There is no error."
            }, {
              "name": "example",
              "text": "error"
            }],
          "text": "Indicates that an error exist within the input field. Colors the border in red."
        },
        "attribute": "error",
        "reflect": false,
        "defaultValue": "false"
      },
      "readonly": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Input is editable."
            }, {
              "name": "value",
              "text": "true: Input is read-only."
            }, {
              "name": "example",
              "text": "readonly"
            }],
          "text": "Whether the selection is readonly or editable.\nIf true the value of this component will be visualized as a normal text."
        },
        "attribute": "readonly",
        "reflect": false,
        "defaultValue": "false"
      },
      "required": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Not required"
            }, {
              "name": "value",
              "text": "true: It is necessary to select an option to fulfill the requirement."
            }, {
              "name": "example",
              "text": "required"
            }],
          "text": "Marks this input field as a required form element."
        },
        "attribute": "required",
        "reflect": false,
        "defaultValue": "false"
      },
      "placeholder": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Placeholder text"
            }, {
              "name": "example",
              "text": "placeholder"
            }],
          "text": "Placeholder text which is displayed when no option is currently selected."
        },
        "attribute": "placeholder",
        "reflect": false,
        "defaultValue": "\"Bitte w\u00E4hlen\""
      },
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The name of an icon."
            }, {
              "name": "example",
              "text": "icon"
            }],
          "text": "An icon which is displayed next to the placeholder text."
        },
        "attribute": "icon",
        "reflect": false
      },
      "value": {
        "type": "string",
        "mutable": true,
        "complexType": {
          "original": "StringOrObject",
          "resolved": "string | { [key: string]: any; }",
          "references": {
            "StringOrObject": {
              "location": "local"
            }
          }
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "important",
              "text": undefined
            }, {
              "name": "value",
              "text": "string: The string value of the selected option."
            }, {
              "name": "value",
              "text": "object: The value of the selected option in case it's a complex type."
            }, {
              "name": "value",
              "text": "null: No option selected."
            }, {
              "name": "example",
              "text": "value"
            }],
          "text": "The current value of this element. It represents the `value` property of the current selected `<dx-dropdown-option>` or null.\nOther by selecting an option with the mouse or keyboard, it is possible to set this property programmatically."
        },
        "attribute": "value",
        "reflect": true,
        "defaultValue": "null"
      },
      "textalign": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "left: Text is left aligned."
            }, {
              "name": "value",
              "text": "right: Text is right aligned."
            }, {
              "name": "example",
              "text": "text-align"
            }],
          "text": "Defines the text alignment."
        },
        "attribute": "textalign",
        "reflect": false,
        "defaultValue": "\"left\""
      },
      "popupposition": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "\"top\" | \"bottom\"",
          "resolved": "\"bottom\" | \"top\"",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "top: The list is displayed at the top of the input field."
            }, {
              "name": "value",
              "text": "bottom: The list is displayed at the bottom of the input field."
            }, {
              "name": "example",
              "text": "popup-position"
            }],
          "text": "Defines the position of the dropdown list."
        },
        "attribute": "popupposition",
        "reflect": false,
        "defaultValue": "\"bottom\""
      },
      "elseoption": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: The user can only select an option in the list."
            }, {
              "name": "value",
              "text": "true: The user is able to set a value which is not present in the option list by editing a text field under\nthe selection input field."
            }, {
              "name": "example",
              "text": "elseoption"
            }],
          "text": "Whether it is possible to select an option which is not present in the current option list."
        },
        "attribute": "elseoption",
        "reflect": false,
        "defaultValue": "false"
      },
      "elseoptionbuttonlabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Button label"
            }, {
              "name": "example",
              "text": "elseoption"
            }],
          "text": "The label of the button which activates the `elseoption` input field. The `elseoption` property has to be `true`."
        },
        "attribute": "elseoptionbuttonlabel",
        "reflect": false,
        "defaultValue": "\"Sonstige Eingabe\""
      },
      "elseoptioninputlabel": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Text input label."
            }, {
              "name": "example",
              "text": "elseoption"
            }],
          "text": "The label of the text input field which represents the `elseoption`. The `elseoption` property has to be `true`."
        },
        "attribute": "elseoptioninputlabel",
        "reflect": false,
        "defaultValue": "\"Sonstige Eingabe\""
      },
      "elseoptionselected": {
        "type": "boolean",
        "mutable": true,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "false: Else-option input field is not displayed."
            }, {
              "name": "value",
              "text": "true: Else-option input field is displayed."
            }, {
              "name": "example",
              "text": "elseoption-selected"
            }],
          "text": "Determines if the else-option input field is visible to the user. This property automatically changes\non user interaction or providing a value that doesn't match any dropdown options."
        },
        "attribute": "elseoptionselected",
        "reflect": true,
        "defaultValue": "false"
      },
      "disabled": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: Dropdown is not editable."
            }, {
              "name": "value",
              "text": "false: Dropdown is editable."
            }, {
              "name": "example",
              "text": "disabled"
            }],
          "text": "Defines whether the dropdown is enabled or not."
        },
        "attribute": "disabled",
        "reflect": false,
        "defaultValue": "false"
      },
      "kisynced": {
        "type": "boolean",
        "mutable": false,
        "complexType": {
          "original": "boolean",
          "resolved": "boolean",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "true: Synchronized."
            }, {
              "name": "value",
              "text": "false: Not synchronized."
            }, {
              "name": "example",
              "text": "kisynced"
            }],
          "text": "Indicates that the form is synced with the KI."
        },
        "attribute": "kisynced",
        "reflect": false,
        "defaultValue": "false"
      }
    };
  }
  static get states() {
    return {
      "options": {},
      "selectedOption": {},
      "inputFocused": {},
      "listOpen": {},
      "internalValue": {},
      "showMouseHoverEffectOnList": {}
    };
  }
  static get events() {
    return [{
        "method": "valueChange",
        "name": "valueChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "example",
              "text": "complex"
            }],
          "text": "Delegates the current selection or text input (`elseoption`) while the user is still editing."
        },
        "complexType": {
          "original": "StringOrObject",
          "resolved": "string | { [key: string]: any; }",
          "references": {
            "StringOrObject": {
              "location": "local"
            }
          }
        }
      }, {
        "method": "valueComplete",
        "name": "valueComplete",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "example",
              "text": "complex"
            }],
          "text": "Delegates the final selection or text input (`elseoption`).\nThe event is fired when the dropdown loses focus or pushing the 'Enter' key."
        },
        "complexType": {
          "original": "StringOrObject",
          "resolved": "string | { [key: string]: any; }",
          "references": {
            "StringOrObject": {
              "location": "local"
            }
          }
        }
      }];
  }
  static get methods() {
    return {
      "focusControl": {
        "complexType": {
          "signature": "() => Promise<void>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            }
          },
          "return": "Promise<void>"
        },
        "docs": {
          "text": "Sets the focus on this element.",
          "tags": [{
              "name": "example",
              "text": "focus"
            }]
        }
      }
    };
  }
  static get elementRef() { return "self"; }
  static get watchers() {
    return [{
        "propName": "value",
        "methodName": "valueChanged"
      }];
  }
}
const NULL_STRING = "null";
