// @dvag/design-system/dist/collection/new/visualization/dx-chart-sample/dx-chart-sample.js
/**
 * Data component to describe chart content.
 *
 * @name Chart-Data
 * @path /Visualization
 * @icon qrcode
 * @stable
 */
export class DxChartSample {
  onValueChange(newValue, oldValue) {
    if (newValue !== oldValue) {
      this.valueChange.emit(this.value);
    }
  }
  /**
   * Maps the component into a JS object.
   */
  async toChartSample() {
    const exposedValue = Math.max(this.value, 0) || 0;
    return {
      label: this.label,
      icon: this.icon,
      value: exposedValue,
      color: this.color,
      selectedcolor: this.selectedcolor || this.color,
    };
  }
  componentShouldUpdate() {
    if (isNaN(this.value)) {
      return false;
    }
  }
  static get is() { return "dx-chart-sample"; }
  static get encapsulation() { return "shadow"; }
  static get properties() {
    return {
      "label": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [],
          "text": "Label of the data."
        },
        "attribute": "label",
        "reflect": true
      },
      "value": {
        "type": "number",
        "mutable": false,
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: The numeric representation."
            }],
          "text": "Numeric representation of the data value."
        },
        "attribute": "value",
        "reflect": true
      },
      "icon": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: Name of the icon."
            }],
          "text": "The icon which is displayed next to the label."
        },
        "attribute": "icon",
        "reflect": true
      },
      "color": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The color code."
            }, {
              "name": "important",
              "text": undefined
            }],
          "text": "The color of the corresponding chart segment."
        },
        "attribute": "color",
        "reflect": true
      },
      "selectedcolor": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "string: The color code."
            }, {
              "name": "important",
              "text": undefined
            }],
          "text": "The color of the corresponding chart segment when selected."
        },
        "attribute": "selectedcolor",
        "reflect": true
      }
    };
  }
  static get events() {
    return [{
        "method": "valueChange",
        "name": "valueChange",
        "bubbles": true,
        "cancelable": true,
        "composed": true,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "number: Delegates the new value."
            }],
          "text": "Triggers when value of the sample changes."
        },
        "complexType": {
          "original": "number",
          "resolved": "number",
          "references": {}
        }
      }];
  }
  static get methods() {
    return {
      "toChartSample": {
        "complexType": {
          "signature": "() => Promise<NewChartSample>",
          "parameters": [],
          "references": {
            "Promise": {
              "location": "global"
            },
            "NewChartSample": {
              "location": "local"
            }
          },
          "return": "Promise<NewChartSample>"
        },
        "docs": {
          "text": "Maps the component into a JS object.",
          "tags": []
        }
      }
    };
  }
  static get watchers() {
    return [{
        "propName": "value",
        "methodName": "onValueChange"
      }];
  }
}
