// @dvag/design-system/dist/collection/new/basics/dx-text-list/dx-text-list.js
import { h, Host, } from "@stencil/core";
import { SlotObserver } from "../../../classes/slot-observer";
import { findUpperElement } from "../../../classes/misc-util";
import { ADD_BULLET_TO_TEXT_CSS_CLASS_NAME } from "../../../classes/constants";
/**
 *
 * An enumerated list that displays `<dx-text>` items as a bullet point list.
 *
 * In addition to `<dx-text>`, other `<dx-text-list>` elements can be placed,
 * to implement further indentation.
 *
 * @path /Basics
 * @name Text list
 * @icon tasks
 * @stable
 */
export class DxTextList {
  constructor() {
    this.bulletWidth = 4;
    /**
     * Specifies the style all text within this text list.
     *
     * If the `type` attribute isn't set, a default style will be set, that is based on the
     * parent components.
     *
     * @value Headline-1: Shortcut is `h1`
     * @value Headline-2: Shortcut is `h2`
     * @value Headline-3: Shortcut is `h3`
     * @value Headline-4: Shortcut is `h4`
     * @value Headline-5: Shortcut is `h5`
     *
     * @value Paragraph-Big: Shortcut is `pb`
     * @value Paragraph-Big-Italic: Shortcut is `pb-italic`
     * @value Paragraph-Big-Bold: Shortcut is `pb-bold`
     * @value Paragraph-Big-Bold-Italic: Shortcut is `pb-bold-italic`
     *
     * @value Paragraph-Standard: Shortcut is `ps`
     * @value Paragraph-Standard-Italic: Shortcut is `ps-italic`
     * @value Paragraph-Standard-Bold: Shortcut is `ps-bold`
     * @value Paragraph-Standard-Bold-Italic: Shortcut is `ps-bold-italic`
     *
     * @value Infotext-Standard: Shortcut is `it`
     * @value Infotext-Standard-Italic: Shortcut is `it-italic`
     * @value Infotext-Standard-Bold: Shortcut is `it-bold`
     * @value Infotext-Standard-Bold-Italic: Shortcut is `it-bold-italic`
     *
     * @value Infotext-Small: Shortcut is `its`
     * @value Infotext-Small-Italic: Shortcut is `its-italic`
     * @value Infotext-Small-Bold: Shortcut is `its-bold`
     * @value Infotext-Small-Bold-Italic: Shortcut is `its-bold-italic`
     *
     * @example type
     */
    this.type = "auto";
  }
  onTypeChange() {
    this.updateTextTypes();
  }
  connectedCallback() {
    this.bulletResizeObserver = new ResizeObserver(() => {
      var _a;
      this.bulletWidth = Math.round((_a = this.bulletWidthReference) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect().width);
    });
    this.updateTextTypes();
  }
  disconnectedCallback() {
    var _a, _b;
    (_a = this.slotObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    (_b = this.bulletResizeObserver) === null || _b === void 0 ? void 0 : _b.disconnect();
  }
  updateTextTypes() {
    var _a;
    (_a = this.slotObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
    this.slotObserver = new SlotObserver(this.self, "*", (nodes) => {
      nodes.forEach((node) => {
        if (node.parentElement === this.self) {
          node.classList.add(ADD_BULLET_TO_TEXT_CSS_CLASS_NAME);
          if (["dx-text", "dx-text-list"].includes(node.tagName.toLowerCase()) &&
            node["type"] ===
              "auto") {
            node["type"] =
              this.type;
          }
        }
      });
      if (this.bulletWidthReference != null) {
        this.bulletResizeObserver.observe(this.bulletWidthReference);
      }
    });
  }
  render() {
    const hasUpperTextListParent = findUpperElement(this.self, "dx-text-list") != null;
    return (h(Host, { style: {
        "--dx-text-list-bullet-width": this.bulletWidth + "px",
      }, class: {
        "has-text-list-parent": hasUpperTextListParent,
      } }, h("div", { class: {
        list: true,
      } }, h("slot", null)), h("dx-text", { ref: (ref) => (this.bulletWidthReference = ref), type: this.type, class: "whitespace-render" }, "\u2022")));
  }
  static get is() { return "dx-text-list"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() {
    return {
      "$": ["dx-text-list.scss"]
    };
  }
  static get styleUrls() {
    return {
      "$": ["dx-text-list.css"]
    };
  }
  static get properties() {
    return {
      "type": {
        "type": "string",
        "mutable": false,
        "complexType": {
          "original": "string",
          "resolved": "string",
          "references": {}
        },
        "required": false,
        "optional": false,
        "docs": {
          "tags": [{
              "name": "value",
              "text": "Headline-1: Shortcut is `h1`"
            }, {
              "name": "value",
              "text": "Headline-2: Shortcut is `h2`"
            }, {
              "name": "value",
              "text": "Headline-3: Shortcut is `h3`"
            }, {
              "name": "value",
              "text": "Headline-4: Shortcut is `h4`"
            }, {
              "name": "value",
              "text": "Headline-5: Shortcut is `h5`"
            }, {
              "name": "value",
              "text": "Paragraph-Big: Shortcut is `pb`"
            }, {
              "name": "value",
              "text": "Paragraph-Big-Italic: Shortcut is `pb-italic`"
            }, {
              "name": "value",
              "text": "Paragraph-Big-Bold: Shortcut is `pb-bold`"
            }, {
              "name": "value",
              "text": "Paragraph-Big-Bold-Italic: Shortcut is `pb-bold-italic`"
            }, {
              "name": "value",
              "text": "Paragraph-Standard: Shortcut is `ps`"
            }, {
              "name": "value",
              "text": "Paragraph-Standard-Italic: Shortcut is `ps-italic`"
            }, {
              "name": "value",
              "text": "Paragraph-Standard-Bold: Shortcut is `ps-bold`"
            }, {
              "name": "value",
              "text": "Paragraph-Standard-Bold-Italic: Shortcut is `ps-bold-italic`"
            }, {
              "name": "value",
              "text": "Infotext-Standard: Shortcut is `it`"
            }, {
              "name": "value",
              "text": "Infotext-Standard-Italic: Shortcut is `it-italic`"
            }, {
              "name": "value",
              "text": "Infotext-Standard-Bold: Shortcut is `it-bold`"
            }, {
              "name": "value",
              "text": "Infotext-Standard-Bold-Italic: Shortcut is `it-bold-italic`"
            }, {
              "name": "value",
              "text": "Infotext-Small: Shortcut is `its`"
            }, {
              "name": "value",
              "text": "Infotext-Small-Italic: Shortcut is `its-italic`"
            }, {
              "name": "value",
              "text": "Infotext-Small-Bold: Shortcut is `its-bold`"
            }, {
              "name": "value",
              "text": "Infotext-Small-Bold-Italic: Shortcut is `its-bold-italic`"
            }, {
              "name": "example",
              "text": "type"
            }],
          "text": "Specifies the style all text within this text list.\n\nIf the `type` attribute isn't set, a default style will be set, that is based on the\nparent components."
        },
        "attribute": "type",
        "reflect": false,
        "defaultValue": "\"auto\""
      }
    };
  }
  static get states() {
    return {
      "bulletWidth": {}
    };
  }
  static get elementRef() { return "self"; }
  static get watchers() {
    return [{
        "propName": "type",
        "methodName": "onTypeChange"
      }];
  }
}
